{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8679283f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install pandas numpy scikit-learn matplotlib seaborn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b0f410e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install --upgrade numpy pandas\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e784990",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30a931ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:3: SyntaxWarning: invalid escape sequence '\\|'\n",
      "<>:3: SyntaxWarning: invalid escape sequence '\\|'\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/553985258.py:3: SyntaxWarning: invalid escape sequence '\\|'\n",
      "  sep=\"\\|\\|\",\n"
     ]
    }
   ],
   "source": [
    "train_var = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_variants\")\n",
    "train_text = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_text\",\n",
    "                          sep=\"\\|\\|\",\n",
    "                            engine=\"python\",\n",
    "                              names=[\"ID\",\"Text\"],\n",
    "                                skiprows=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "044e2fe9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coloumns are :  Index(['ID', 'Gene', 'Variation', 'Class'], dtype='object')\n",
      "shape of the data is :  (3321, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class\n",
       "0   0  FAM58A  Truncating Mutations      1\n",
       "1   1     CBL                 W802*      2\n",
       "2   2     CBL                 Q249E      2"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"coloumns are : \",train_var.columns)\n",
    "print(\"shape of the data is : \",train_var.shape)\n",
    "\n",
    "train_var.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53ad6416",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coloumns are :  Index(['ID', 'Text'], dtype='object')\n",
      "(3321, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"coloumns are : \",train_text.columns)\n",
    "print(train_text.shape)\n",
    "\n",
    "train_text.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cb74043a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  FAM58A  Truncating Mutations      1   \n",
       "1   1     CBL                 W802*      2   \n",
       "2   2     CBL                 Q249E      2   \n",
       "\n",
       "                                                Text  \n",
       "0  Cyclin-dependent kinases (CDKs) regulate a var...  \n",
       "1   Abstract Background  Non-small cell lung canc...  \n",
       "2   Abstract Background  Non-small cell lung canc...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(train_var,train_text,on=\"ID\")\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "db9dd8d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Cyclin-dependent kinases (CDKs) regulate a variety of fundamental cellular processes. CDK10 stands out as one of the last orphan CDKs for which no activating cyclin has been identified and no kinase activity revealed. Previous work has shown that CDK10 silencing increases ETS2 (v-ets erythroblastosis virus E26 oncogene homolog 2)-driven activation of the MAPK pathway, which confers tamoxifen resistance to breast cancer cells. The precise mechanisms by which CDK10 modulates ETS2 activity, and more generally the functions of CDK10, remain elusive. Here we demonstrate that CDK10 is a cyclin-dependent kinase by identifying cyclin M as an activating cyclin. Cyclin M, an orphan cyclin, is the product of FAM58A, whose mutations cause STAR syndrome, a human developmental anomaly whose features include toe syndactyly, telecanthus, and anogenital and renal malformations. We show that STAR syndrome-associated cyclin M mutants are unable to interact with CDK10. Cyclin M silencing phenocopies CDK10 silencing in increasing c-Raf and in conferring tamoxifen resistance to breast cancer cells. CDK10/cyclin M phosphorylates ETS2 in vitro, and in cells it positively controls ETS2 degradation by the proteasome. ETS2 protein levels are increased in cells derived from a STAR patient, and this increase is attributable to decreased cyclin M levels. Altogether, our results reveal an additional regulatory mechanism for ETS2, which plays key roles in cancer and development. They also shed light on the molecular mechanisms underlying STAR syndrome.Cyclin-dependent kinases (CDKs) play a pivotal role in the control of a number of fundamental cellular processes (1). The human genome contains 21 genes encoding proteins that can be considered as members of the CDK family owing to their sequence similarity with bona fide CDKs, those known to be activated by cyclins (2). Although discovered almost 20 y ago (3, 4), CDK10 remains one of the two CDKs without an identified cyclin partner. This knowledge gap has largely impeded the exploration of its biological functions. CDK10 can act as a positive cell cycle regulator in some cells (5, 6) or as a tumor suppressor in others (7, 8). CDK10 interacts with the ETS2 (v-ets erythroblastosis virus E26 oncogene homolog 2) transcription factor and inhibits its transcriptional activity through an unknown mechanism (9). CDK10 knockdown derepresses ETS2, which increases the expression of the c-Raf protein kinase, activates the MAPK pathway, and induces resistance of MCF7 cells to tamoxifen (6).Here, we deorphanize CDK10 by identifying cyclin M, the product of FAM58A, as a binding partner. Mutations in this gene that predict absence or truncation of cyclin M are associated with STAR syndrome, whose features include toe syndactyly, telecanthus, and anogenital and renal malformations in heterozygous females (10). However, both the functions of cyclin M and the pathogenesis of STAR syndrome remain unknown. We show that a recombinant CDK10/cyclin M heterodimer is an active protein kinase that phosphorylates ETS2 in vitro. Cyclin M silencing phenocopies CDK10 silencing in increasing c-Raf and phospho-ERK expression levels and in inducing tamoxifen resistance in estrogen receptor (ER)+ breast cancer cells. We show that CDK10/cyclin M positively controls ETS2 degradation by the proteasome, through the phosphorylation of two neighboring serines. Finally, we detect an increased ETS2 expression level in cells derived from a STAR patient, and we demonstrate that it is attributable to the decreased cyclin M expression level observed in these cells.Previous SectionNext SectionResultsA yeast two-hybrid (Y2H) screen unveiled an interaction signal between CDK10 and a mouse protein whose C-terminal half presents a strong sequence homology with the human FAM58A gene product [whose proposed name is cyclin M (11)]. We thus performed Y2H mating assays to determine whether human CDK10 interacts with human cyclin M (Fig. 1 A–C). The longest CDK10 isoform (P1) expressed as a bait protein produced a strong interaction phenotype with full-length cyclin M (expressed as a prey protein) but no detectable phenotype with cyclin D1, p21 (CIP1), and Cdi1 (KAP), which are known binding partners of other CDKs (Fig. 1B). CDK1 and CDK3 also produced Y2H signals with cyclin M, albeit notably weaker than that observed with CDK10 (Fig. 1B). An interaction phenotype was also observed between full-length cyclin M and CDK10 proteins expressed as bait and prey, respectively (Fig. S1A). We then tested different isoforms of CDK10 and cyclin M originating from alternative gene splicing, and two truncated cyclin M proteins corresponding to the hypothetical products of two mutated FAM58A genes found in STAR syndrome patients (10). None of these shorter isoforms produced interaction phenotypes (Fig. 1 A and C and Fig. S1A).Fig. 1.In a new window Download PPTFig. 1.CDK10 and cyclin M form an interaction complex. (A) Schematic representation of the different protein isoforms analyzed by Y2H assays. Amino acid numbers are indicated. Black boxes indicate internal deletions. The red box indicates a differing amino acid sequence compared with CDK10 P1. (B) Y2H assay between a set of CDK proteins expressed as baits (in fusion to the LexA DNA binding domain) and CDK interacting proteins expressed as preys (in fusion to the B42 transcriptional activator). pEG202 and pJG4-5 are the empty bait and prey plasmids expressing LexA and B42, respectively. lacZ was used as a reporter gene, and blue yeast are indicative of a Y2H interaction phenotype. (C) Y2H assay between the different CDK10 and cyclin M isoforms. The amino-terminal region of ETS2, known to interact with CDK10 (9), was also assayed. (D) Western blot analysis of Myc-CDK10 (wt or kd) and CycM-V5-6His expression levels in transfected HEK293 cells. (E) Western blot analysis of Myc-CDK10 (wt or kd) immunoprecipitates obtained using the anti-Myc antibody. “Inputs” correspond to 10 μg total lysates obtained from HEK293 cells coexpressing Myc-CDK10 (wt or kd) and CycM-V5-6His. (F) Western blot analysis of immunoprecipitates obtained using the anti-CDK10 antibody or a control goat antibody, from human breast cancer MCF7 cells. “Input” corresponds to 30 μg MCF7 total cell lysates. The lower band of the doublet observed on the upper panel comigrates with the exogenously expressed untagged CDK10 and thus corresponds to endogenous CDK10. The upper band of the doublet corresponds to a nonspecific signal, as demonstrated by it insensitivity to either overexpression of CDK10 (as seen on the left lane) or silencing of CDK10 (Fig. S2B). Another experiment with a longer gel migration is shown in Fig. S1D.Next we examined the ability of CDK10 and cyclin M to interact when expressed in human cells (Fig. 1 D and E). We tested wild-type CDK10 (wt) and a kinase dead (kd) mutant bearing a D181A amino acid substitution that abolishes ATP binding (12). We expressed cyclin M-V5-6His and/or Myc-CDK10 (wt or kd) in a human embryonic kidney cell line (HEK293). The expression level of cyclin M-V5-6His was significantly increased upon coexpression with Myc-CDK10 (wt or kd) and, to a lesser extent, that of Myc-CDK10 (wt or kd) was increased upon coexpression with cyclin M-V5-6His (Fig. 1D). We then immunoprecipitated Myc-CDK10 proteins and detected the presence of cyclin M in the CDK10 (wt) and (kd) immunoprecipitates only when these proteins were coexpressed pair-wise (Fig. 1E). We confirmed these observations by detecting the presence of Myc-CDK10 in cyclin M-V5-6His immunoprecipitates (Fig. S1B). These experiments confirmed the lack of robust interaction between the CDK10.P2 isoform and cyclin M (Fig. S1C). To detect the interaction between endogenous proteins, we performed immunoprecipitations on nontransfected MCF7 cells derived from a human breast cancer. CDK10 and cyclin M antibodies detected their cognate endogenous proteins by Western blotting. We readily detected cyclin M in immunoprecipitates obtained with the CDK10 antibody but not with a control antibody (Fig. 1F). These results confirm the physical interaction between CDK10 and cyclin M in human cells.To unveil a hypothesized CDK10/cyclin M protein kinase activity, we produced GST-CDK10 and StrepII-cyclin M fusion proteins in insect cells, either individually or in combination. We observed that GST-CDK10 and StrepII-cyclin M copurified, thus confirming their interaction in yet another cellular model (Fig. 2A). We then performed in vitro kinase assays with purified proteins, using histone H1 as a generic substrate. Histone H1 phosphorylation was detected only from lysates of cells coexpressing GST-CDK10 and StrepII-cyclin M. No phosphorylation was detected when GST-CDK10 or StrepII-cyclin M were expressed alone, or when StrepII-cyclin M was coexpressed with GST-CDK10(kd) (Fig. 2A). Next we investigated whether ETS2, which is known to interact with CDK10 (9) (Fig. 1C), is a phosphorylation substrate of CDK10/cyclin M. We detected strong phosphorylation of ETS2 by the GST-CDK10/StrepII-cyclin M purified heterodimer, whereas no phosphorylation was detected using GST-CDK10 alone or GST-CDK10(kd)/StrepII-cyclin M heterodimer (Fig. 2B).Fig. 2.In a new window Download PPTFig. 2.CDK10 is a cyclin M-dependent protein kinase. (A) In vitro protein kinase assay on histone H1. Lysates from insect cells expressing different proteins were purified on a glutathione Sepharose matrix to capture GST-CDK10(wt or kd) fusion proteins alone, or in complex with STR-CycM fusion protein. Purified protein expression levels were analyzed by Western blots (Top and Upper Middle). The kinase activity was determined by autoradiography of histone H1, whose added amounts were visualized by Coomassie staining (Lower Middle and Bottom). (B) Same as in A, using purified recombinant 6His-ETS2 as a substrate.CDK10 silencing has been shown to increase ETS2-driven c-RAF transcription and to activate the MAPK pathway (6). We investigated whether cyclin M is also involved in this regulatory pathway. To aim at a highly specific silencing, we used siRNA pools (mix of four different siRNAs) at low final concentration (10 nM). Both CDK10 and cyclin M siRNA pools silenced the expression of their cognate targets (Fig. 3 A and C and Fig. S2) and, interestingly, the cyclin M siRNA pool also caused a marked decrease in CDK10 protein level (Fig. 3A and Fig. S2B). These results, and those shown in Fig. 1D, suggest that cyclin M binding stabilizes CDK10. Cyclin M silencing induced an increase in c-Raf protein and mRNA levels (Fig. 3 B and C) and in phosphorylated ERK1 and ERK2 protein levels (Fig. S3B), similarly to CDK10 silencing. As expected from these effects (6), CDK10 and cyclin M silencing both decreased the sensitivity of ER+ MCF7 cells to tamoxifen, to a similar extent. The combined silencing of both genes did not result in a higher resistance to the drug (Fig. S3C). Altogether, these observations demonstrate a functional interaction between cyclin M and CDK10, which negatively controls ETS2.Fig. 3.In a new window Download PPTFig. 3.Cyclin M silencing up-regulates c-Raf expression. (A) Western blot analysis of endogenous CDK10 and cyclin M expression levels in MCF7 cells, in response to siRNA-mediated gene silencing. (B) Western blot analysis of endogenous c-Raf expression levels in MCF7 cells, in response to CDK10 or cyclin M silencing. A quantification is shown in Fig. S3A. (C) Quantitative RT-PCR analysis of CDK10, cyclin M, and c-Raf mRNA levels, in response to CDK10 (Upper) or cyclin M (Lower) silencing. **P ≤ 0.01; ***P ≤ 0.001.We then wished to explore the mechanism by which CDK10/cyclin M controls ETS2. ETS2 is a short-lived protein degraded by the proteasome (13). A straightforward hypothesis is that CDK10/cyclin M positively controls ETS2 degradation. We thus examined the impact of CDK10 or cyclin M silencing on ETS2 expression levels. The silencing of CDK10 and that of cyclin M caused an increase in the expression levels of an exogenously expressed Flag-ETS2 protein (Fig. S4A), as well as of the endogenous ETS2 protein (Fig. 4A). This increase is not attributable to increased ETS2 mRNA levels, which marginally fluctuated in response to CDK10 or cyclin M silencing (Fig. S4B). We then examined the expression levels of the Flag-tagged ETS2 protein when expressed alone or in combination with Myc-CDK10 or -CDK10(kd), with or without cyclin M-V5-6His. Flag-ETS2 was readily detected when expressed alone or, to a lesser extent, when coexpressed with CDK10(kd). However, its expression level was dramatically decreased when coexpressed with CDK10 alone, or with CDK10 and cyclin M (Fig. 4B). These observations suggest that endogenous cyclin M levels are in excess compared with those of CDK10 in MCF7 cells, and they show that the major decrease in ETS2 levels observed upon CDK10 coexpression involves CDK10 kinase activity. Treatment of cells coexpressing Flag-ETS2, CDK10, and cyclin M with the proteasome inhibitor MG132 largely rescued Flag-ETS2 expression levels (Fig. 4B).Fig. 4.In a new window Download PPTFig. 4.CDK10/cyclin M controls ETS2 stability in human cancer derived cells. (A) Western blot analysis of endogenous ETS2 expression levels in MCF7 cells, in response to siRNA-mediated CDK10 and/or cyclin M silencing. A quantification is shown in Fig. S4B. (B) Western blot analysis of exogenously expressed Flag-ETS2 protein levels in MCF7 cells cotransfected with empty vectors or coexpressing Myc-CDK10 (wt or kd), or Myc-CDK10/CycM-V5-6His. The latter cells were treated for 16 h with the MG132 proteasome inhibitor. Proper expression of CDK10 and cyclin M tagged proteins was verified by Western blot analysis. (C and D) Western blot analysis of expression levels of exogenously expressed Flag-ETS2 wild-type or mutant proteins in MCF7 cells, in the absence of (C) or in response to (D) Myc-CDK10/CycM-V5-6His expression. Quantifications are shown in Fig. S4 C and D.A mass spectrometry analysis of recombinant ETS2 phosphorylated by CDK10/cyclin M in vitro revealed the existence of multiple phosphorylated residues, among which are two neighboring phospho-serines (at positions 220 and 225) that may form a phosphodegron (14) (Figs. S5–S8). To confirm this finding, we compared the phosphorylation level of recombinant ETS2wt with that of ETS2SASA protein, a mutant bearing alanine substitutions of these two serines. As expected from the existence of multiple phosphorylation sites, we detected a small but reproducible, significant decrease of phosphorylation level of ETS2SASA compared with ETS2wt (Fig. S9), thus confirming that Ser220/Ser225 are phosphorylated by CDK10/cyclin M. To establish a direct link between ETS2 phosphorylation by CDK10/cyclin M and degradation, we examined the expression levels of Flag-ETS2SASA. In the absence of CDK10/cyclin M coexpression, it did not differ significantly from that of Flag-ETS2. This is contrary to that of Flag-ETS2DBM, bearing a deletion of the N-terminal destruction (D-) box that was previously shown to be involved in APC-Cdh1–mediated degradation of ETS2 (13) (Fig. 4C). However, contrary to Flag-ETS2 wild type, the expression level of Flag-ETS2SASA remained insensitive to CDK10/cyclin M coexpression (Fig. 4D). Altogether, these results suggest that CDK10/cyclin M directly controls ETS2 degradation through the phosphorylation of these two serines.Finally, we studied a lymphoblastoid cell line derived from a patient with STAR syndrome, bearing FAM58A mutation c.555+1G>A, predicted to result in aberrant splicing (10). In accordance with incomplete skewing of X chromosome inactivation previously found in this patient, we detected a decreased expression level of cyclin M protein in the STAR cell line, compared with a control lymphoblastoid cell line. In line with our preceding observations, we detected an increased expression level of ETS2 protein in the STAR cell line compared with the control (Fig. 5A and Fig. S10A). We then examined by quantitative RT-PCR the mRNA expression levels of the corresponding genes. The STAR cell line showed a decreased expression level of cyclin M mRNA but an expression level of ETS2 mRNA similar to that of the control cell line (Fig. 5B). To demonstrate that the increase in ETS2 protein expression is indeed a result of the decreased cyclin M expression observed in the STAR patient-derived cell line, we expressed cyclin M-V5-6His in this cell line. This expression caused a decrease in ETS2 protein levels (Fig. 5C).Fig. 5.In a new window Download PPTFig. 5.Decreased cyclin M expression in STAR patient-derived cells results in increased ETS2 protein level. (A) Western blot analysis of cyclin M and ETS2 protein levels in a STAR patient-derived lymphoblastoid cell line and in a control lymphoblastoid cell line, derived from a healthy individual. A quantification is shown in Fig. S10A. (B) Quantitative RT-PCR analysis of cyclin M and ETS2 mRNA levels in the same cells. ***P ≤ 0.001. (C) Western blot analysis of ETS2 protein levels in the STAR patient-derived lymphoblastoid cell line transfected with an empty vector or a vector directing the expression of cyclin M-V5-6His. Another Western blot revealing endogenously and exogenously expressed cyclin M levels is shown in Fig. S10B. A quantification of ETS2 protein levels is shown in Fig. S10C.Previous SectionNext SectionDiscussionIn this work, we unveil the interaction between CDK10, the last orphan CDK discovered in the pregenomic era (2), and cyclin M, the only cyclin associated with a human genetic disease so far, and whose functions remain unknown (10). The closest paralogs of CDK10 within the CDK family are the CDK11 proteins, which interact with L-type cyclins (15). Interestingly, the closest paralog of these cyclins within the cyclin family is cyclin M (Fig. S11). The fact that none of the shorter CDK10 isoforms interact robustly with cyclin M suggests that alternative splicing of the CDK10 gene (16, 17) plays an important role in regulating CDK10 functions.The functional relevance of the interaction between CDK10 and cyclin M is supported by different observations. Both proteins seem to enhance each other’s stability, as judged from their increased expression levels when their partner is exogenously coexpressed (Fig. 1D) and from the much reduced endogenous CDK10 expression level observed in response to cyclin M silencing (Fig. 3A and Fig. S2B). CDK10 is subject to ubiquitin-mediated degradation (18). Our observations suggest that cyclin M protects CDK10 from such degradation and that it is the only cyclin partner of CDK10, at least in MCF7 cells. They also suggest that cyclin M stability is enhanced upon binding to CDK10, independently from its kinase activity, as seen for cyclin C and CDK8 (19). We uncover a cyclin M-dependent CDK10 protein kinase activity in vitro, thus demonstrating that this protein, which was named a CDK on the sole basis of its amino acid sequence, is indeed a genuine cyclin-dependent kinase. Our Y2H assays reveal that truncated cyclin M proteins corresponding to the hypothetical products of two STAR syndrome-associated FAM58A mutations do not produce an interaction phenotype with CDK10. Hence, regardless of whether these mutated mRNAs undergo nonsense-mediated decay (as suggested from the decreased cyclin M mRNA levels in STAR cells, shown in Fig. 5B) or give rise to truncated cyclin M proteins, females affected by the STAR syndrome must exhibit compromised CDK10/cyclin M kinase activity at least in some tissues and during specific developmental stages.We show that ETS2, a known interactor of CDK10, is a phosphorylation substrate of CDK10/cyclin M in vitro and that CDK10/cyclin M kinase activity positively controls ETS2 degradation by the proteasome. This control seems to be exerted through a very fine mechanism, as judged from the sensitivity of ETS2 levels to partially decreased CDK10 and cyclin M levels, achieved in MCF7 cells and observed in STAR cells, respectively. These findings offer a straightforward explanation for the already reported up-regulation of ETS2-driven transcription of c-RAF in response to CDK10 silencing (6). We bring evidence that CDK10/cyclin M directly controls ETS2 degradation through the phosphorylation of two neighboring serines, which may form a noncanonical β-TRCP phosphodegron (DSMCPAS) (14). Because none of these two serines precede a proline, they do not conform to usual CDK phosphorylation sites. However, multiple so-called transcriptional CDKs (CDK7, -8, -9, and -11) (to which CDK10 may belong; Fig. S11) have been shown to phosphorylate a variety of motifs in a non–proline-directed fashion, especially in the context of molecular docking with the substrate (20). Here, it can be hypothesized that the high-affinity interaction between CDK10 and the Pointed domain of ETS2 (6, 9) (Fig. 1C) would allow docking-mediated phosphorylation of atypical sites. The control of ETS2 degradation involves a number of players, including APC-Cdh1 (13) and the cullin-RING ligase CRL4 (21). The formal identification of the ubiquitin ligase involved in the CDK10/cyclin M pathway and the elucidation of its concerted action with the other ubiquitin ligases to regulate ETS2 degradation will require further studies.Our results present a number of significant biological and medical implications. First, they shed light on the regulation of ETS2, which plays an important role in development (22) and is frequently deregulated in many cancers (23). Second, our results contribute to the understanding of the molecular mechanisms causing tamoxifen resistance associated with reduced CDK10 expression levels, and they suggest that, like CDK10 (6), cyclin M could also be a predictive clinical marker of hormone therapy response of ERα-positive breast cancer patients. Third, our findings offer an interesting hypothesis on the molecular mechanisms underlying STAR syndrome. Ets2 transgenic mice showing a less than twofold overexpression of Ets2 present severe cranial abnormalities (24), and those observed in STAR patients could thus be caused at least in part by increased ETS2 protein levels. Another expected consequence of enhanced ETS2 expression levels would be a decreased risk to develop certain types of cancers and an increased risk to develop others. Studies on various mouse models (including models of Down syndrome, in which three copies of ETS2 exist) have revealed that ETS2 dosage can repress or promote tumor growth and, hence, that ETS2 exerts noncell autonomous functions in cancer (25). Intringuingly, one of the very few STAR patients identified so far has been diagnosed with a nephroblastoma (26). Finally, our findings will facilitate the general exploration of the biological functions of CDK10 and, in particular, its role in the control of cell division. Previous studies have suggested either a positive role in cell cycle control (5, 6) or a tumor-suppressive activity in some cancers (7, 8). The severe growth retardation exhibited by STAR patients strongly suggests that CDK10/cyclin M plays an important role in the control of cell proliferation.Previous SectionNext SectionMaterials and MethodsCloning of CDK10 and cyclin M cDNAs, plasmid constructions, tamoxifen response analysis, quantitative RT-PCR, mass spectrometry experiments, and antibody production are detailed in SI Materials and Methods.Yeast Two-Hybrid Interaction Assays. We performed yeast interaction mating assays as previously described (27).Mammalian Cell Cultures and Transfections. We grew human HEK293 and MCF7 cells in DMEM supplemented with 10% (vol/vol) FBS (Invitrogen), and we grew lymphoblastoid cells in RPMI 1640 GlutaMAX supplemented with 15% (vol/vol) FBS. We transfected HEK293 and MCF7 cells using Lipofectamine 2000 (Invitrogen) for plasmids, Lipofectamine RNAiMAX (Invitrogen) for siRNAs, and Jetprime (Polyplus) for plasmids/siRNAs combinations according to the manufacturers’ instructions. We transfected lymphoblastoid cells by electroporation (Neon, Invitrogen). For ETS2 stability studies we treated MCF7 cells 32 h after transfection with 10 μM MG132 (Fisher Scientific) for 16 h.Coimmunoprecipitation and Western Blot Experiments. We collected cells by scraping in PBS (or centrifugation for lymphoblastoid cells) and lysed them by sonication in a lysis buffer containing 60 mM β-glycerophosphate, 15 mM p-nitrophenylphosphate, 25 mM 3-(N-morpholino)propanesulfonic acid (Mops) (pH 7.2), 15 mM EGTA, 15 mM MgCl2, 1 mM Na vanadate, 1 mM NaF, 1mM phenylphosphate, 0.1% Nonidet P-40, and a protease inhibitor mixture (Roche). We spun the lysates 15 min at 20,000 × g at 4 °C, collected the supernatants, and determined the protein content using a Bradford assay. We performed the immunoprecipitation experiments on 500 μg of total proteins, in lysis buffer. We precleared the lysates with 20 μL of protein A or G-agarose beads, incubated 1 h 4 °C on a rotating wheel. We added 5 μg of antibody to the supernatants, incubated 1 h 4 °C on a rotating wheel, added 20 μL of protein A or G-agarose beads, and incubated 1 h 4 °C on a rotating wheel. We collected the beads by centrifugation 30 s at 18,000 × g at 4 °C and washed three times in a bead buffer containing 50 mM Tris (pH 7.4), 5 mM NaF, 250 mM NaCl, 5 mM EDTA, 5 mM EGTA, 0.1% Nonidet P-40, and a protease inhibitor coktail (Roche). We directly added sample buffer to the washed pellets, heat-denatured the proteins, and ran the samples on 10% Bis-Tris SDS/PAGE. We transferred the proteins onto Hybond nitrocellulose membranes and processed the blots according to standard procedures. For Western blot experiments, we used the following primary antibodies: anti-Myc (Abcam ab9106, 1:2,000), anti-V5 (Invitrogen R960, 1:5,000), anti-tubulin (Santa Cruz Biotechnology B-7, 1:500), anti-CDK10 (Covalab pab0847p, 1:500 or Santa Cruz Biotechnology C-19, 1:500), anti-CycM (home-made, dilution 1:500 or Covalab pab0882-P, dilution 1:500), anti-Raf1 (Santa Cruz Biotechnology C-20, 1:1,000), anti-ETS2 (Santa Cruz Biotechnology C-20, 1:1,000), anti-Flag (Sigma F7425, 1:1,000), and anti-actin (Sigma A5060, 1:5,000). We used HRP-coupled anti-goat (Santa Cruz Biotechnology SC-2033, dilution 1:2,000), anti-mouse (Bio-Rad 170–6516, dilution 1:3,000) or anti-rabbit (Bio-Rad 172–1019, 1:5,000) as secondary antibodies. We revealed the blots by enhanced chemiluminescence (SuperSignal West Femto, Thermo Scientific).Production and Purification of Recombinant Proteins.GST-CDK10(kd)/StrepII-CycM. We generated recombinant bacmids in DH10Bac Escherichia coli and baculoviruses in Sf9 cells using the Bac-to-Bac system, as described by the provider (Invitrogen). We infected Sf9 cells with GST-CDK10- (or GST-CDK10kd)-producing viruses, or coinfected the cells with StrepII-CycM–producing viruses, and we collected the cells 72 h after infection. To purify GST-fusion proteins, we spun 250 mL cells and resuspended the pellet in 40 mL lysis buffer (PBS, 250 mM NaCl, 0.5% Nonidet P-40, 50 mM NaF, 10 mM β-glycerophosphate, and 0.3 mM Na-vanadate) containing a protease inhibitor mixture (Roche). We lysed the cells by sonication, spun the lysate 30 min at 15,000 × g, collected the soluble fraction, and added it to a 1-mL glutathione-Sepharose matrix. We incubated 1 h at 4 °C, washed four times with lysis buffer, one time with kinase buffer A (see below), and finally resuspended the beads in 100 μL kinase buffer A containing 10% (vol/vol) glycerol for storage.6His-ETS2. We transformed Origami2 DE3 (Novagen) with the 6His-ETS2 expression vector. We induced expression with 0.2 mM isopropyl-β-d-1-thiogalactopyranoside for 3 h at 22 °C. To purify 6His-ETS2, we spun 50 mL cells and resuspended the pellet in 2 mL lysis buffer (PBS, 300 mM NaCl, 10 mM Imidazole, 1 mM DTT, and 0.1% Nonidet P-40) containing a protease inhibitor mixture without EDTA (Roche). We lysed the cells at 1.6 bar using a cell disruptor and spun the lysate 10 min at 20,000 × g. We collected the soluble fraction and added it to 200 μL Cobalt beads (Thermo Scientific). After 1 h incubation at 4 °C on a rotating wheel, we washed four times with lysis buffer. To elute, we incubated beads 30 min with elution buffer (PBS, 250 mM imidazole, pH 7.6) containing the protease inhibitor mixture, spun 30 s at 10,000 × g, and collected the eluted protein.Protein Kinase Assays. We mixed glutathione-Sepharose beads (harboring GST-CDK10 wt or kd, either monomeric or complexed with StrepII-CycM), 22.7 μM BSA, 15 mM DTT, 100 μM ATP, 5 μCi ATP[γ-32P], 7.75 μM histone H1, or 1 μM 6His-ETS2 and added kinase buffer A (25 mM Tris·HCl, 10 mM MgCl2, 1 mM EGTA, 1 mM DTT, and 3.7 μM heparin, pH 7.5) up to a total volume of 30 μL. We incubated the reactions 30 min at 30 °C, added Laemli sample buffer, heat-denatured the samples, and ran 10% Bis-Tris SDS/PAGE. We cut gel slices to detect GST-CDK10 and StrepII-CycM by Western blotting. We stained the gel slices containing the substrate with Coomassie (R-250, Bio-Rad), dried them, and detected the incorporated radioactivity by autoradiography. We identified four unrelated girls with anogenital and renal malformations, dysmorphic facial features, normal intellect and syndactyly of toes. A similar combination of features had been reported previously in a mother–daughter pair1 (Table 1 and Supplementary Note online). These authors noted clinical overlap with Townes-Brocks syndrome but suggested that the phenotype represented a separate autosomal dominant entity (MIM601446). Here we define the cardinal features of this syndrome as a characteristic facial appearance with apparent telecanthus and broad tripartite nasal tip, variable syndactyly of toes 2–5, hypoplastic labia, anal atresia and urogenital malformations (Fig. 1a–h). We also observed a variety of other features (Table 1).  Figure 1: Clinical and molecular characterization of STAR syndrome.  Figure 1 : Clinical and molecular characterization of STAR syndrome. (a–f) Facial appearances of cases 1–3 (apparent telecanthus, dysplastic ears and thin upper lips; a,c,e), and toe syndactyly 2–5, 3–5 or 4–5 (b,d,f) in these cases illustrate recognizable features of STAR syndrome (specific parental consent has been obtained for publication of these photographs). Anal atresia and hypoplastic labia are not shown. (g,h) X-ray films of the feet of case 2 showing only four rays on the left and delta-shaped 4th and 5th metatarsals on the right (h; compare to clinical picture in d). (i) Array-CGH data. Log2 ratio represents copy number loss of six probes spanning between 37.9 and 50.7 kb, with one probe positioned within FAM58A. The deletion does not remove parts of other functional genes. (j) Schematic structure of FAM58A and position of the mutations. FAM58A has five coding exons (boxes). The cyclin domain (green) is encoded by exons 2–4. The horizontal arrow indicates the deletion extending 5' in case 1, which includes exons 1 and 2, whereas the horizontal line below exon 5 indicates the deletion found in case 3, which removes exon 5 and some 3' sequence. The pink horizontal bars above the boxes indicate the amplicons used for qPCR and sequencing (one alternative exon 5 amplicon is not indicated because of space constraints). The mutation 201dupT (case 4) results in an immediate stop codon, and the 555+1G>A and 555-1G>A splice mutations in cases 2, 5 and 6 are predicted to be deleterious because they alter the conserved splice donor and acceptor site of intron 4, respectively.  Full size image (97 KB)  Table 1: Clinical features in STAR syndrome cases  Table 1 - Clinical features in STAR syndrome cases  Full table  On the basis of the phenotypic overlap with Townes-Brocks, Okihiro and Feingold syndromes, we analyzed SALL1 (ref. 2), SALL4 (ref. 3) and MYCN4 but found no mutations in any of these genes (Supplementary Methods online). Next, we carried out genome-wide high-resolution oligonucleotide array comparative genomic hybridization (CGH)5 analysis (Supplementary Methods) of genomic DNA from the most severely affected individual (case 1, with lower lid coloboma, epilepsy and syringomyelia) and identified a heterozygous deletion of 37.9–50.7 kb on Xq28, which removed exons 1 and 2 of FAM58A (Fig. 1i,j). Using real-time PCR, we confirmed the deletion in the child and excluded it in her unaffected parents (Supplementary Fig. 1a online, Supplementary Methods and Supplementary Table 1 online). Through CGH with a customized oligonucleotide array enriched in probes for Xq28, followed by breakpoint cloning, we defined the exact deletion size as 40,068 bp (g.152,514,164_152,554,231del(chromosome X, NCBI Build 36.2); Fig. 1j and Supplementary Figs. 2,3 online). The deletion removes the coding regions of exons 1 and 2 as well as intron 1 (2,774 bp), 492 bp of intron 2, and 36,608 bp of 5' sequence, including the 5' UTR and the entire KRT18P48 pseudogene (NCBI gene ID 340598). Paternity was proven using routine methods. We did not find deletions overlapping FAM58A in the available copy number variation (CNV) databases.  Subsequently, we carried out qPCR analysis of the three other affected individuals (cases 2, 3 and 4) and the mother-daughter pair from the literature (cases 5 and 6). In case 3, we detected a de novo heterozygous deletion of 1.1–10.3 kb overlapping exon 5 (Supplementary Fig. 1b online). Using Xq28-targeted array CGH and breakpoint cloning, we identified a deletion of 4,249 bp (g.152,504,123_152,508,371del(chromosome X, NCBI Build 36.2); Fig. 1j and Supplementary Figs. 2,3), which removed 1,265 bp of intron 4, all of exon 5, including the 3' UTR, and 2,454 bp of 3' sequence.  We found heterozygous FAM58A point mutations in the remaining cases (Fig. 1j, Supplementary Fig. 2, Supplementary Methods and Supplementary Table 1). In case 2, we identified the mutation 555+1G>A, affecting the splice donor site of intron 4. In case 4, we identified the frameshift mutation 201dupT, which immediately results in a premature stop codon N68XfsX1. In cases 5 and 6, we detected the mutation 556-1G>A, which alters the splice acceptor site of intron 4. We validated the point mutations and deletions by independent rounds of PCR and sequencing or by qPCR. We confirmed paternity and de novo status of the point mutations and deletions in all sporadic cases. None of the mutations were seen in the DNA of 60 unaffected female controls, and no larger deletions involving FAM58A were found in 93 unrelated array-CGH investigations.  By analyzing X-chromosome inactivation (Supplementary Methods and Supplementary Fig. 4 online), we found complete skewing of X inactivation in cases 1 and 3–6 and almost complete skewing in case 2, suggesting that cells carrying the mutation on the active X chromosome have a growth disadvantage during fetal development. Using RT-PCR on RNA from lymphoblastoid cells of case 2 (Supplementary Fig. 2), we did not find any aberrant splice products as additional evidence that the mutated allele is inactivated. Furthermore, FAM58A is subjected to X inactivation6. In cases 1 and 3, the parental origin of the deletions could not be determined, as a result of lack of informative SNPs. Case 5, the mother of case 6, gave birth to two boys, both clinically unaffected (samples not available). We cannot exclude that the condition is lethal in males. No fetal losses were reported from any of the families.  The function of FAM58A is unknown. The gene consists of five coding exons, and the 642-bp coding region encodes a protein of 214 amino acids. GenBank lists a mRNA length of 1,257 bp for the reference sequence (NM_152274.2). Expression of the gene (by EST data) was found in 27 of 48 adult tissues including kidney, colon, cervix and uterus, but not heart (NCBI expression viewer, UniGene Hs.496943). Expression was also noted in 24 of 26 listed tumor tissues as well as in embryo and fetus. Genes homologous to FAM58A (NCBI HomoloGene: 13362) are found on the X chromosome in the chimpanzee and the dog. The zebrafish has a similar gene on chromosome 23. However, in the mouse and rat, there are no true homologs. These species have similar but intronless genes on chromosomes 11 (mouse) and 10 (rat), most likely arising from a retrotransposon insertion event. On the murine X chromosome, the flanking genes Atp2b3 and Dusp9 are conserved, but only remnants of the FAM58A sequence can be detected.  FAM58A contains a cyclin-box-fold domain, a protein-binding domain found in cyclins with a role in cell cycle and transcription control. No human phenotype resulting from a cyclin gene mutation has yet been reported. Homozygous knockout mice for Ccnd1 (encoding cyclin D1) are viable but small and have reduced lifespan. They also have dystrophic changes of the retina, likely as a result of decreased cell proliferation and degeneration of photoreceptor cells during embryogenesis7, 8.  Cyclin D1 colocalizes with SALL4 in the nucleus, and both proteins cooperatively mediate transcriptional repression9. As the phenotype of our cases overlaps considerably with that of Townes-Brocks syndrome caused by SALL1 mutations1, we carried out co-immunoprecipitation to find out if SALL1 or SALL4 would interact with FAM58A in a manner similar to that observed for SALL4 and cyclin D1. We found that FAM58A interacts with SALL1 but not with SALL4 (Supplementary Fig. 5 online), supporting the hypothesis that FAM58A and SALL1 participate in the same developmental pathway.  How do FAM58A mutations lead to STAR syndrome? Growth retardation (all cases; Table 1) and retinal abnormalities (three cases) are reminiscent of the reduced body size and retinal anomalies in cyclin D1 knockout mice7, 8. Therefore, a proliferation defect might be partly responsible for STAR syndrome. To address this question, we carried out a knockdown of FAM58A mRNA followed by a proliferation assay. Transfection of HEK293 cells with three different FAM58A-specific RNAi oligonucleotides resulted in a significant reduction of both FAM58A mRNA expression and proliferation of transfected cells (Supplementary Methods and Supplementary Fig. 6 online), supporting the link between FAM58A and cell proliferation.  We found that loss-of-function mutations of FAM58A result in a rather homogeneous clinical phenotype. The additional anomalies in case 1 are likely to result from an effect of the 40-kb deletion on expression of a neighboring gene, possibly ATP2B3 or DUSP9. However, we cannot exclude that the homogeneous phenotype results from an ascertainment bias and that FAM58A mutations, including missense changes, could result in a broader spectrum of malformations. The genes causing the overlapping phenotypes of STAR syndrome and Townes-Brocks syndrome seem to act in the same pathway. Of note, MYCN, a gene mutated in Feingold syndrome, is a direct regulator of cyclin D2 (refs. 10,11); thus, it is worth exploring whether the phenotypic similarities between Feingold and STAR syndrome might be explained by direct regulation of FAM58A by MYCN.  FAM58A is located approximately 0.56 Mb centromeric to MECP2 on Xq28. Duplications overlapping both MECP2 and FAM58A have been described and are not associated with a clinical phenotype in females12, but no deletions overlapping both MECP2 and FAM58A have been observed to date13. Although other genes between FAM58A and MECP2 have been implicated in brain development, FAM58A and MECP2 are the only genes in this region known to result in X-linked dominant phenotypes; thus, deletion of both genes on the same allele might be lethal in both males and females.\""
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Text'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04e51085",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9b505487",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>fam58a</td>\n",
       "      <td>truncating mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>cyclin-dependent kinases (cdks) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cbl</td>\n",
       "      <td>w802*</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>cbl</td>\n",
       "      <td>q249e</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  fam58a  truncating mutations      1   \n",
       "1   1     cbl                 w802*      2   \n",
       "2   2     cbl                 q249e      2   \n",
       "\n",
       "                                                Text  \n",
       "0  cyclin-dependent kinases (cdks) regulate a var...  \n",
       "1   abstract background  non-small cell lung canc...  \n",
       "2   abstract background  non-small cell lung canc...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Text'] = df['Text'].str.lower()\n",
    "df['Gene'] = df['Gene'].str.lower()\n",
    "df['Variation'] = df['Variation'].str.lower()\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fd21e3ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>fam58a</td>\n",
       "      <td>truncating_mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>cyclin-dependent kinases (cdks) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cbl</td>\n",
       "      <td>w802*</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>cbl</td>\n",
       "      <td>q249e</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  fam58a  truncating_mutations      1   \n",
       "1   1     cbl                 w802*      2   \n",
       "2   2     cbl                 q249e      2   \n",
       "\n",
       "                                                Text  \n",
       "0  cyclin-dependent kinases (cdks) regulate a var...  \n",
       "1   abstract background  non-small cell lung canc...  \n",
       "2   abstract background  non-small cell lung canc...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# also replace space with underscore in gene and variation column\n",
    "df['Gene'] = df['Gene'].str.replace(' ', '_')\n",
    "df['Variation'] = df['Variation'].str.replace(' ', '_')\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "927e2a21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "Text\n",
      "<class 'str'>      3316\n",
      "<class 'float'>       5\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['Text'].isnull().sum())  # কতগুলো NaN আছে\n",
    "print(df['Text'].apply(type).value_counts())  # কোন টাইপের ডেটা আছে\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3c180c89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop na the missing Values\n",
    "\n",
    "df.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f845a9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "Text\n",
      "<class 'str'>    3316\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['Text'].isnull().sum())  # কতগুলো NaN আছে\n",
    "print(df['Text'].apply(type).value_counts())  # কোন টাইপের ডেটা আছে\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c2f1b90",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c69051b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: nltk in ./venv/lib/python3.12/site-packages (3.9.1)\n",
      "Requirement already satisfied: click in ./venv/lib/python3.12/site-packages (from nltk) (8.3.0)\n",
      "Requirement already satisfied: joblib in ./venv/lib/python3.12/site-packages (from nltk) (1.5.2)\n",
      "Requirement already satisfied: regex>=2021.8.3 in ./venv/lib/python3.12/site-packages (from nltk) (2025.9.18)\n",
      "Requirement already satisfied: tqdm in ./venv/lib/python3.12/site-packages (from nltk) (4.67.1)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Error loading stopwords: <urlopen error [SSL:\n",
      "[nltk_data]     CERTIFICATE_VERIFY_FAILED] certificate verify failed:\n",
      "[nltk_data]     unable to get local issuer certificate (_ssl.c:1000)>\n"
     ]
    }
   ],
   "source": [
    "# nltlk install stopwords\n",
    "!pip install nltk\n",
    "import nltk\n",
    "nltk.download('stopwords')\n",
    "from nltk.corpus import stopwords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e4ba63ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:7: SyntaxWarning: invalid escape sequence '\\['\n",
      "<>:9: SyntaxWarning: invalid escape sequence '\\w'\n",
      "<>:12: SyntaxWarning: invalid escape sequence '\\S'\n",
      "<>:7: SyntaxWarning: invalid escape sequence '\\['\n",
      "<>:9: SyntaxWarning: invalid escape sequence '\\w'\n",
      "<>:12: SyntaxWarning: invalid escape sequence '\\S'\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/189891491.py:7: SyntaxWarning: invalid escape sequence '\\['\n",
      "  text = re.sub('\\[.*?\\]', '', text)\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/189891491.py:9: SyntaxWarning: invalid escape sequence '\\w'\n",
      "  text = re.sub('\\w*\\d\\w*', '', text)\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/189891491.py:12: SyntaxWarning: invalid escape sequence '\\S'\n",
      "  text = re.sub('http\\S+|www\\S+|https\\S+', '', text, flags=re.MULTILINE)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "      <th>Cleaned_Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>fam58a</td>\n",
       "      <td>truncating_mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>cyclin-dependent kinases (cdks) regulate a var...</td>\n",
       "      <td>cyclindependent kinases cdks regulate variety ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cbl</td>\n",
       "      <td>w802*</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "      <td>abstract background nonsmall cell lung cancer ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>cbl</td>\n",
       "      <td>q249e</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "      <td>abstract background nonsmall cell lung cancer ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  fam58a  truncating_mutations      1   \n",
       "1   1     cbl                 w802*      2   \n",
       "2   2     cbl                 q249e      2   \n",
       "\n",
       "                                                Text  \\\n",
       "0  cyclin-dependent kinases (cdks) regulate a var...   \n",
       "1   abstract background  non-small cell lung canc...   \n",
       "2   abstract background  non-small cell lung canc...   \n",
       "\n",
       "                                        Cleaned_Text  \n",
       "0  cyclindependent kinases cdks regulate variety ...  \n",
       "1  abstract background nonsmall cell lung cancer ...  \n",
       "2  abstract background nonsmall cell lung cancer ...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove stop words, punctuations, special characters, numbers etc  \n",
    "import re\n",
    "import string   \n",
    "from nltk.corpus import stopwords\n",
    "stop_words = set(stopwords.words('english'))\n",
    "def clean_text(text):\n",
    "    text = re.sub('\\[.*?\\]', '', text)\n",
    "    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)\n",
    "    text = re.sub('\\w*\\d\\w*', '', text)\n",
    "    text = re.sub('[''\"\"…]', '', text)\n",
    "    text = re.sub('\\n', '', text)\n",
    "    text = re.sub('http\\S+|www\\S+|https\\S+', '', text, flags=re.MULTILINE)\n",
    "    text = re.sub(r'\\S*@\\S*\\s?', '', text)\n",
    "    text = re.sub(r'\\s+', ' ', text).strip()\n",
    "    text = ' '.join(word for word in text.split() if word not in stop_words)\n",
    "    return text\n",
    "#df['Text'] = df['Text'].apply(clean_text)\n",
    "df['Cleaned_Text'] = df['Text'].apply(clean_text)\n",
    "\n",
    "df.head(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "48344b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lemmatiztion\n",
    "# lemmatization\n",
    "import nltk\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from nltk.corpus import wordnet\n",
    "\n",
    "lemmatizer = WordNetLemmatizer()\n",
    "\n",
    "def lemma(text):\n",
    "    y = []\n",
    "    for i in text.split():\n",
    "        y.append(lemmatizer.lemmatize(i))  \n",
    "    return \" \".join(y)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4fa10d75",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Cleaned_Text'] = df['Cleaned_Text'].apply(lemma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a5cca057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "      <th>Cleaned_Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>fam58a</td>\n",
       "      <td>truncating_mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>cyclin-dependent kinases (cdks) regulate a var...</td>\n",
       "      <td>cyclindependent kinase cdks regulate variety f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cbl</td>\n",
       "      <td>w802*</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "      <td>abstract background nonsmall cell lung cancer ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>cbl</td>\n",
       "      <td>q249e</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "      <td>abstract background nonsmall cell lung cancer ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>cbl</td>\n",
       "      <td>n454d</td>\n",
       "      <td>3</td>\n",
       "      <td>recent evidence has demonstrated that acquired...</td>\n",
       "      <td>recent evidence demonstrated acquired uniparen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>cbl</td>\n",
       "      <td>l399v</td>\n",
       "      <td>4</td>\n",
       "      <td>oncogenic mutations in the monomeric casitas b...</td>\n",
       "      <td>oncogenic mutation monomeric casitas blineage ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  fam58a  truncating_mutations      1   \n",
       "1   1     cbl                 w802*      2   \n",
       "2   2     cbl                 q249e      2   \n",
       "3   3     cbl                 n454d      3   \n",
       "4   4     cbl                 l399v      4   \n",
       "\n",
       "                                                Text  \\\n",
       "0  cyclin-dependent kinases (cdks) regulate a var...   \n",
       "1   abstract background  non-small cell lung canc...   \n",
       "2   abstract background  non-small cell lung canc...   \n",
       "3  recent evidence has demonstrated that acquired...   \n",
       "4  oncogenic mutations in the monomeric casitas b...   \n",
       "\n",
       "                                        Cleaned_Text  \n",
       "0  cyclindependent kinase cdks regulate variety f...  \n",
       "1  abstract background nonsmall cell lung cancer ...  \n",
       "2  abstract background nonsmall cell lung cancer ...  \n",
       "3  recent evidence demonstrated acquired uniparen...  \n",
       "4  oncogenic mutation monomeric casitas blineage ...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0e4cbd41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"cyclin-dependent kinases (cdks) regulate a variety of fundamental cellular processes. cdk10 stands out as one of the last orphan cdks for which no activating cyclin has been identified and no kinase activity revealed. previous work has shown that cdk10 silencing increases ets2 (v-ets erythroblastosis virus e26 oncogene homolog 2)-driven activation of the mapk pathway, which confers tamoxifen resistance to breast cancer cells. the precise mechanisms by which cdk10 modulates ets2 activity, and more generally the functions of cdk10, remain elusive. here we demonstrate that cdk10 is a cyclin-dependent kinase by identifying cyclin m as an activating cyclin. cyclin m, an orphan cyclin, is the product of fam58a, whose mutations cause star syndrome, a human developmental anomaly whose features include toe syndactyly, telecanthus, and anogenital and renal malformations. we show that star syndrome-associated cyclin m mutants are unable to interact with cdk10. cyclin m silencing phenocopies cdk10 silencing in increasing c-raf and in conferring tamoxifen resistance to breast cancer cells. cdk10/cyclin m phosphorylates ets2 in vitro, and in cells it positively controls ets2 degradation by the proteasome. ets2 protein levels are increased in cells derived from a star patient, and this increase is attributable to decreased cyclin m levels. altogether, our results reveal an additional regulatory mechanism for ets2, which plays key roles in cancer and development. they also shed light on the molecular mechanisms underlying star syndrome.cyclin-dependent kinases (cdks) play a pivotal role in the control of a number of fundamental cellular processes (1). the human genome contains 21 genes encoding proteins that can be considered as members of the cdk family owing to their sequence similarity with bona fide cdks, those known to be activated by cyclins (2). although discovered almost 20 y ago (3, 4), cdk10 remains one of the two cdks without an identified cyclin partner. this knowledge gap has largely impeded the exploration of its biological functions. cdk10 can act as a positive cell cycle regulator in some cells (5, 6) or as a tumor suppressor in others (7, 8). cdk10 interacts with the ets2 (v-ets erythroblastosis virus e26 oncogene homolog 2) transcription factor and inhibits its transcriptional activity through an unknown mechanism (9). cdk10 knockdown derepresses ets2, which increases the expression of the c-raf protein kinase, activates the mapk pathway, and induces resistance of mcf7 cells to tamoxifen (6).here, we deorphanize cdk10 by identifying cyclin m, the product of fam58a, as a binding partner. mutations in this gene that predict absence or truncation of cyclin m are associated with star syndrome, whose features include toe syndactyly, telecanthus, and anogenital and renal malformations in heterozygous females (10). however, both the functions of cyclin m and the pathogenesis of star syndrome remain unknown. we show that a recombinant cdk10/cyclin m heterodimer is an active protein kinase that phosphorylates ets2 in vitro. cyclin m silencing phenocopies cdk10 silencing in increasing c-raf and phospho-erk expression levels and in inducing tamoxifen resistance in estrogen receptor (er)+ breast cancer cells. we show that cdk10/cyclin m positively controls ets2 degradation by the proteasome, through the phosphorylation of two neighboring serines. finally, we detect an increased ets2 expression level in cells derived from a star patient, and we demonstrate that it is attributable to the decreased cyclin m expression level observed in these cells.previous sectionnext sectionresultsa yeast two-hybrid (y2h) screen unveiled an interaction signal between cdk10 and a mouse protein whose c-terminal half presents a strong sequence homology with the human fam58a gene product [whose proposed name is cyclin m (11)]. we thus performed y2h mating assays to determine whether human cdk10 interacts with human cyclin m (fig. 1 a–c). the longest cdk10 isoform (p1) expressed as a bait protein produced a strong interaction phenotype with full-length cyclin m (expressed as a prey protein) but no detectable phenotype with cyclin d1, p21 (cip1), and cdi1 (kap), which are known binding partners of other cdks (fig. 1b). cdk1 and cdk3 also produced y2h signals with cyclin m, albeit notably weaker than that observed with cdk10 (fig. 1b). an interaction phenotype was also observed between full-length cyclin m and cdk10 proteins expressed as bait and prey, respectively (fig. s1a). we then tested different isoforms of cdk10 and cyclin m originating from alternative gene splicing, and two truncated cyclin m proteins corresponding to the hypothetical products of two mutated fam58a genes found in star syndrome patients (10). none of these shorter isoforms produced interaction phenotypes (fig. 1 a and c and fig. s1a).fig. 1.in a new window download pptfig. 1.cdk10 and cyclin m form an interaction complex. (a) schematic representation of the different protein isoforms analyzed by y2h assays. amino acid numbers are indicated. black boxes indicate internal deletions. the red box indicates a differing amino acid sequence compared with cdk10 p1. (b) y2h assay between a set of cdk proteins expressed as baits (in fusion to the lexa dna binding domain) and cdk interacting proteins expressed as preys (in fusion to the b42 transcriptional activator). peg202 and pjg4-5 are the empty bait and prey plasmids expressing lexa and b42, respectively. lacz was used as a reporter gene, and blue yeast are indicative of a y2h interaction phenotype. (c) y2h assay between the different cdk10 and cyclin m isoforms. the amino-terminal region of ets2, known to interact with cdk10 (9), was also assayed. (d) western blot analysis of myc-cdk10 (wt or kd) and cycm-v5-6his expression levels in transfected hek293 cells. (e) western blot analysis of myc-cdk10 (wt or kd) immunoprecipitates obtained using the anti-myc antibody. “inputs” correspond to 10 μg total lysates obtained from hek293 cells coexpressing myc-cdk10 (wt or kd) and cycm-v5-6his. (f) western blot analysis of immunoprecipitates obtained using the anti-cdk10 antibody or a control goat antibody, from human breast cancer mcf7 cells. “input” corresponds to 30 μg mcf7 total cell lysates. the lower band of the doublet observed on the upper panel comigrates with the exogenously expressed untagged cdk10 and thus corresponds to endogenous cdk10. the upper band of the doublet corresponds to a nonspecific signal, as demonstrated by it insensitivity to either overexpression of cdk10 (as seen on the left lane) or silencing of cdk10 (fig. s2b). another experiment with a longer gel migration is shown in fig. s1d.next we examined the ability of cdk10 and cyclin m to interact when expressed in human cells (fig. 1 d and e). we tested wild-type cdk10 (wt) and a kinase dead (kd) mutant bearing a d181a amino acid substitution that abolishes atp binding (12). we expressed cyclin m-v5-6his and/or myc-cdk10 (wt or kd) in a human embryonic kidney cell line (hek293). the expression level of cyclin m-v5-6his was significantly increased upon coexpression with myc-cdk10 (wt or kd) and, to a lesser extent, that of myc-cdk10 (wt or kd) was increased upon coexpression with cyclin m-v5-6his (fig. 1d). we then immunoprecipitated myc-cdk10 proteins and detected the presence of cyclin m in the cdk10 (wt) and (kd) immunoprecipitates only when these proteins were coexpressed pair-wise (fig. 1e). we confirmed these observations by detecting the presence of myc-cdk10 in cyclin m-v5-6his immunoprecipitates (fig. s1b). these experiments confirmed the lack of robust interaction between the cdk10.p2 isoform and cyclin m (fig. s1c). to detect the interaction between endogenous proteins, we performed immunoprecipitations on nontransfected mcf7 cells derived from a human breast cancer. cdk10 and cyclin m antibodies detected their cognate endogenous proteins by western blotting. we readily detected cyclin m in immunoprecipitates obtained with the cdk10 antibody but not with a control antibody (fig. 1f). these results confirm the physical interaction between cdk10 and cyclin m in human cells.to unveil a hypothesized cdk10/cyclin m protein kinase activity, we produced gst-cdk10 and strepii-cyclin m fusion proteins in insect cells, either individually or in combination. we observed that gst-cdk10 and strepii-cyclin m copurified, thus confirming their interaction in yet another cellular model (fig. 2a). we then performed in vitro kinase assays with purified proteins, using histone h1 as a generic substrate. histone h1 phosphorylation was detected only from lysates of cells coexpressing gst-cdk10 and strepii-cyclin m. no phosphorylation was detected when gst-cdk10 or strepii-cyclin m were expressed alone, or when strepii-cyclin m was coexpressed with gst-cdk10(kd) (fig. 2a). next we investigated whether ets2, which is known to interact with cdk10 (9) (fig. 1c), is a phosphorylation substrate of cdk10/cyclin m. we detected strong phosphorylation of ets2 by the gst-cdk10/strepii-cyclin m purified heterodimer, whereas no phosphorylation was detected using gst-cdk10 alone or gst-cdk10(kd)/strepii-cyclin m heterodimer (fig. 2b).fig. 2.in a new window download pptfig. 2.cdk10 is a cyclin m-dependent protein kinase. (a) in vitro protein kinase assay on histone h1. lysates from insect cells expressing different proteins were purified on a glutathione sepharose matrix to capture gst-cdk10(wt or kd) fusion proteins alone, or in complex with str-cycm fusion protein. purified protein expression levels were analyzed by western blots (top and upper middle). the kinase activity was determined by autoradiography of histone h1, whose added amounts were visualized by coomassie staining (lower middle and bottom). (b) same as in a, using purified recombinant 6his-ets2 as a substrate.cdk10 silencing has been shown to increase ets2-driven c-raf transcription and to activate the mapk pathway (6). we investigated whether cyclin m is also involved in this regulatory pathway. to aim at a highly specific silencing, we used sirna pools (mix of four different sirnas) at low final concentration (10 nm). both cdk10 and cyclin m sirna pools silenced the expression of their cognate targets (fig. 3 a and c and fig. s2) and, interestingly, the cyclin m sirna pool also caused a marked decrease in cdk10 protein level (fig. 3a and fig. s2b). these results, and those shown in fig. 1d, suggest that cyclin m binding stabilizes cdk10. cyclin m silencing induced an increase in c-raf protein and mrna levels (fig. 3 b and c) and in phosphorylated erk1 and erk2 protein levels (fig. s3b), similarly to cdk10 silencing. as expected from these effects (6), cdk10 and cyclin m silencing both decreased the sensitivity of er+ mcf7 cells to tamoxifen, to a similar extent. the combined silencing of both genes did not result in a higher resistance to the drug (fig. s3c). altogether, these observations demonstrate a functional interaction between cyclin m and cdk10, which negatively controls ets2.fig. 3.in a new window download pptfig. 3.cyclin m silencing up-regulates c-raf expression. (a) western blot analysis of endogenous cdk10 and cyclin m expression levels in mcf7 cells, in response to sirna-mediated gene silencing. (b) western blot analysis of endogenous c-raf expression levels in mcf7 cells, in response to cdk10 or cyclin m silencing. a quantification is shown in fig. s3a. (c) quantitative rt-pcr analysis of cdk10, cyclin m, and c-raf mrna levels, in response to cdk10 (upper) or cyclin m (lower) silencing. **p ≤ 0.01; ***p ≤ 0.001.we then wished to explore the mechanism by which cdk10/cyclin m controls ets2. ets2 is a short-lived protein degraded by the proteasome (13). a straightforward hypothesis is that cdk10/cyclin m positively controls ets2 degradation. we thus examined the impact of cdk10 or cyclin m silencing on ets2 expression levels. the silencing of cdk10 and that of cyclin m caused an increase in the expression levels of an exogenously expressed flag-ets2 protein (fig. s4a), as well as of the endogenous ets2 protein (fig. 4a). this increase is not attributable to increased ets2 mrna levels, which marginally fluctuated in response to cdk10 or cyclin m silencing (fig. s4b). we then examined the expression levels of the flag-tagged ets2 protein when expressed alone or in combination with myc-cdk10 or -cdk10(kd), with or without cyclin m-v5-6his. flag-ets2 was readily detected when expressed alone or, to a lesser extent, when coexpressed with cdk10(kd). however, its expression level was dramatically decreased when coexpressed with cdk10 alone, or with cdk10 and cyclin m (fig. 4b). these observations suggest that endogenous cyclin m levels are in excess compared with those of cdk10 in mcf7 cells, and they show that the major decrease in ets2 levels observed upon cdk10 coexpression involves cdk10 kinase activity. treatment of cells coexpressing flag-ets2, cdk10, and cyclin m with the proteasome inhibitor mg132 largely rescued flag-ets2 expression levels (fig. 4b).fig. 4.in a new window download pptfig. 4.cdk10/cyclin m controls ets2 stability in human cancer derived cells. (a) western blot analysis of endogenous ets2 expression levels in mcf7 cells, in response to sirna-mediated cdk10 and/or cyclin m silencing. a quantification is shown in fig. s4b. (b) western blot analysis of exogenously expressed flag-ets2 protein levels in mcf7 cells cotransfected with empty vectors or coexpressing myc-cdk10 (wt or kd), or myc-cdk10/cycm-v5-6his. the latter cells were treated for 16 h with the mg132 proteasome inhibitor. proper expression of cdk10 and cyclin m tagged proteins was verified by western blot analysis. (c and d) western blot analysis of expression levels of exogenously expressed flag-ets2 wild-type or mutant proteins in mcf7 cells, in the absence of (c) or in response to (d) myc-cdk10/cycm-v5-6his expression. quantifications are shown in fig. s4 c and d.a mass spectrometry analysis of recombinant ets2 phosphorylated by cdk10/cyclin m in vitro revealed the existence of multiple phosphorylated residues, among which are two neighboring phospho-serines (at positions 220 and 225) that may form a phosphodegron (14) (figs. s5–s8). to confirm this finding, we compared the phosphorylation level of recombinant ets2wt with that of ets2sasa protein, a mutant bearing alanine substitutions of these two serines. as expected from the existence of multiple phosphorylation sites, we detected a small but reproducible, significant decrease of phosphorylation level of ets2sasa compared with ets2wt (fig. s9), thus confirming that ser220/ser225 are phosphorylated by cdk10/cyclin m. to establish a direct link between ets2 phosphorylation by cdk10/cyclin m and degradation, we examined the expression levels of flag-ets2sasa. in the absence of cdk10/cyclin m coexpression, it did not differ significantly from that of flag-ets2. this is contrary to that of flag-ets2dbm, bearing a deletion of the n-terminal destruction (d-) box that was previously shown to be involved in apc-cdh1–mediated degradation of ets2 (13) (fig. 4c). however, contrary to flag-ets2 wild type, the expression level of flag-ets2sasa remained insensitive to cdk10/cyclin m coexpression (fig. 4d). altogether, these results suggest that cdk10/cyclin m directly controls ets2 degradation through the phosphorylation of these two serines.finally, we studied a lymphoblastoid cell line derived from a patient with star syndrome, bearing fam58a mutation c.555+1g>a, predicted to result in aberrant splicing (10). in accordance with incomplete skewing of x chromosome inactivation previously found in this patient, we detected a decreased expression level of cyclin m protein in the star cell line, compared with a control lymphoblastoid cell line. in line with our preceding observations, we detected an increased expression level of ets2 protein in the star cell line compared with the control (fig. 5a and fig. s10a). we then examined by quantitative rt-pcr the mrna expression levels of the corresponding genes. the star cell line showed a decreased expression level of cyclin m mrna but an expression level of ets2 mrna similar to that of the control cell line (fig. 5b). to demonstrate that the increase in ets2 protein expression is indeed a result of the decreased cyclin m expression observed in the star patient-derived cell line, we expressed cyclin m-v5-6his in this cell line. this expression caused a decrease in ets2 protein levels (fig. 5c).fig. 5.in a new window download pptfig. 5.decreased cyclin m expression in star patient-derived cells results in increased ets2 protein level. (a) western blot analysis of cyclin m and ets2 protein levels in a star patient-derived lymphoblastoid cell line and in a control lymphoblastoid cell line, derived from a healthy individual. a quantification is shown in fig. s10a. (b) quantitative rt-pcr analysis of cyclin m and ets2 mrna levels in the same cells. ***p ≤ 0.001. (c) western blot analysis of ets2 protein levels in the star patient-derived lymphoblastoid cell line transfected with an empty vector or a vector directing the expression of cyclin m-v5-6his. another western blot revealing endogenously and exogenously expressed cyclin m levels is shown in fig. s10b. a quantification of ets2 protein levels is shown in fig. s10c.previous sectionnext sectiondiscussionin this work, we unveil the interaction between cdk10, the last orphan cdk discovered in the pregenomic era (2), and cyclin m, the only cyclin associated with a human genetic disease so far, and whose functions remain unknown (10). the closest paralogs of cdk10 within the cdk family are the cdk11 proteins, which interact with l-type cyclins (15). interestingly, the closest paralog of these cyclins within the cyclin family is cyclin m (fig. s11). the fact that none of the shorter cdk10 isoforms interact robustly with cyclin m suggests that alternative splicing of the cdk10 gene (16, 17) plays an important role in regulating cdk10 functions.the functional relevance of the interaction between cdk10 and cyclin m is supported by different observations. both proteins seem to enhance each other’s stability, as judged from their increased expression levels when their partner is exogenously coexpressed (fig. 1d) and from the much reduced endogenous cdk10 expression level observed in response to cyclin m silencing (fig. 3a and fig. s2b). cdk10 is subject to ubiquitin-mediated degradation (18). our observations suggest that cyclin m protects cdk10 from such degradation and that it is the only cyclin partner of cdk10, at least in mcf7 cells. they also suggest that cyclin m stability is enhanced upon binding to cdk10, independently from its kinase activity, as seen for cyclin c and cdk8 (19). we uncover a cyclin m-dependent cdk10 protein kinase activity in vitro, thus demonstrating that this protein, which was named a cdk on the sole basis of its amino acid sequence, is indeed a genuine cyclin-dependent kinase. our y2h assays reveal that truncated cyclin m proteins corresponding to the hypothetical products of two star syndrome-associated fam58a mutations do not produce an interaction phenotype with cdk10. hence, regardless of whether these mutated mrnas undergo nonsense-mediated decay (as suggested from the decreased cyclin m mrna levels in star cells, shown in fig. 5b) or give rise to truncated cyclin m proteins, females affected by the star syndrome must exhibit compromised cdk10/cyclin m kinase activity at least in some tissues and during specific developmental stages.we show that ets2, a known interactor of cdk10, is a phosphorylation substrate of cdk10/cyclin m in vitro and that cdk10/cyclin m kinase activity positively controls ets2 degradation by the proteasome. this control seems to be exerted through a very fine mechanism, as judged from the sensitivity of ets2 levels to partially decreased cdk10 and cyclin m levels, achieved in mcf7 cells and observed in star cells, respectively. these findings offer a straightforward explanation for the already reported up-regulation of ets2-driven transcription of c-raf in response to cdk10 silencing (6). we bring evidence that cdk10/cyclin m directly controls ets2 degradation through the phosphorylation of two neighboring serines, which may form a noncanonical β-trcp phosphodegron (dsmcpas) (14). because none of these two serines precede a proline, they do not conform to usual cdk phosphorylation sites. however, multiple so-called transcriptional cdks (cdk7, -8, -9, and -11) (to which cdk10 may belong; fig. s11) have been shown to phosphorylate a variety of motifs in a non–proline-directed fashion, especially in the context of molecular docking with the substrate (20). here, it can be hypothesized that the high-affinity interaction between cdk10 and the pointed domain of ets2 (6, 9) (fig. 1c) would allow docking-mediated phosphorylation of atypical sites. the control of ets2 degradation involves a number of players, including apc-cdh1 (13) and the cullin-ring ligase crl4 (21). the formal identification of the ubiquitin ligase involved in the cdk10/cyclin m pathway and the elucidation of its concerted action with the other ubiquitin ligases to regulate ets2 degradation will require further studies.our results present a number of significant biological and medical implications. first, they shed light on the regulation of ets2, which plays an important role in development (22) and is frequently deregulated in many cancers (23). second, our results contribute to the understanding of the molecular mechanisms causing tamoxifen resistance associated with reduced cdk10 expression levels, and they suggest that, like cdk10 (6), cyclin m could also be a predictive clinical marker of hormone therapy response of erα-positive breast cancer patients. third, our findings offer an interesting hypothesis on the molecular mechanisms underlying star syndrome. ets2 transgenic mice showing a less than twofold overexpression of ets2 present severe cranial abnormalities (24), and those observed in star patients could thus be caused at least in part by increased ets2 protein levels. another expected consequence of enhanced ets2 expression levels would be a decreased risk to develop certain types of cancers and an increased risk to develop others. studies on various mouse models (including models of down syndrome, in which three copies of ets2 exist) have revealed that ets2 dosage can repress or promote tumor growth and, hence, that ets2 exerts noncell autonomous functions in cancer (25). intringuingly, one of the very few star patients identified so far has been diagnosed with a nephroblastoma (26). finally, our findings will facilitate the general exploration of the biological functions of cdk10 and, in particular, its role in the control of cell division. previous studies have suggested either a positive role in cell cycle control (5, 6) or a tumor-suppressive activity in some cancers (7, 8). the severe growth retardation exhibited by star patients strongly suggests that cdk10/cyclin m plays an important role in the control of cell proliferation.previous sectionnext sectionmaterials and methodscloning of cdk10 and cyclin m cdnas, plasmid constructions, tamoxifen response analysis, quantitative rt-pcr, mass spectrometry experiments, and antibody production are detailed in si materials and methods.yeast two-hybrid interaction assays. we performed yeast interaction mating assays as previously described (27).mammalian cell cultures and transfections. we grew human hek293 and mcf7 cells in dmem supplemented with 10% (vol/vol) fbs (invitrogen), and we grew lymphoblastoid cells in rpmi 1640 glutamax supplemented with 15% (vol/vol) fbs. we transfected hek293 and mcf7 cells using lipofectamine 2000 (invitrogen) for plasmids, lipofectamine rnaimax (invitrogen) for sirnas, and jetprime (polyplus) for plasmids/sirnas combinations according to the manufacturers’ instructions. we transfected lymphoblastoid cells by electroporation (neon, invitrogen). for ets2 stability studies we treated mcf7 cells 32 h after transfection with 10 μm mg132 (fisher scientific) for 16 h.coimmunoprecipitation and western blot experiments. we collected cells by scraping in pbs (or centrifugation for lymphoblastoid cells) and lysed them by sonication in a lysis buffer containing 60 mm β-glycerophosphate, 15 mm p-nitrophenylphosphate, 25 mm 3-(n-morpholino)propanesulfonic acid (mops) (ph 7.2), 15 mm egta, 15 mm mgcl2, 1 mm na vanadate, 1 mm naf, 1mm phenylphosphate, 0.1% nonidet p-40, and a protease inhibitor mixture (roche). we spun the lysates 15 min at 20,000 × g at 4 °c, collected the supernatants, and determined the protein content using a bradford assay. we performed the immunoprecipitation experiments on 500 μg of total proteins, in lysis buffer. we precleared the lysates with 20 μl of protein a or g-agarose beads, incubated 1 h 4 °c on a rotating wheel. we added 5 μg of antibody to the supernatants, incubated 1 h 4 °c on a rotating wheel, added 20 μl of protein a or g-agarose beads, and incubated 1 h 4 °c on a rotating wheel. we collected the beads by centrifugation 30 s at 18,000 × g at 4 °c and washed three times in a bead buffer containing 50 mm tris (ph 7.4), 5 mm naf, 250 mm nacl, 5 mm edta, 5 mm egta, 0.1% nonidet p-40, and a protease inhibitor coktail (roche). we directly added sample buffer to the washed pellets, heat-denatured the proteins, and ran the samples on 10% bis-tris sds/page. we transferred the proteins onto hybond nitrocellulose membranes and processed the blots according to standard procedures. for western blot experiments, we used the following primary antibodies: anti-myc (abcam ab9106, 1:2,000), anti-v5 (invitrogen r960, 1:5,000), anti-tubulin (santa cruz biotechnology b-7, 1:500), anti-cdk10 (covalab pab0847p, 1:500 or santa cruz biotechnology c-19, 1:500), anti-cycm (home-made, dilution 1:500 or covalab pab0882-p, dilution 1:500), anti-raf1 (santa cruz biotechnology c-20, 1:1,000), anti-ets2 (santa cruz biotechnology c-20, 1:1,000), anti-flag (sigma f7425, 1:1,000), and anti-actin (sigma a5060, 1:5,000). we used hrp-coupled anti-goat (santa cruz biotechnology sc-2033, dilution 1:2,000), anti-mouse (bio-rad 170–6516, dilution 1:3,000) or anti-rabbit (bio-rad 172–1019, 1:5,000) as secondary antibodies. we revealed the blots by enhanced chemiluminescence (supersignal west femto, thermo scientific).production and purification of recombinant proteins.gst-cdk10(kd)/strepii-cycm. we generated recombinant bacmids in dh10bac escherichia coli and baculoviruses in sf9 cells using the bac-to-bac system, as described by the provider (invitrogen). we infected sf9 cells with gst-cdk10- (or gst-cdk10kd)-producing viruses, or coinfected the cells with strepii-cycm–producing viruses, and we collected the cells 72 h after infection. to purify gst-fusion proteins, we spun 250 ml cells and resuspended the pellet in 40 ml lysis buffer (pbs, 250 mm nacl, 0.5% nonidet p-40, 50 mm naf, 10 mm β-glycerophosphate, and 0.3 mm na-vanadate) containing a protease inhibitor mixture (roche). we lysed the cells by sonication, spun the lysate 30 min at 15,000 × g, collected the soluble fraction, and added it to a 1-ml glutathione-sepharose matrix. we incubated 1 h at 4 °c, washed four times with lysis buffer, one time with kinase buffer a (see below), and finally resuspended the beads in 100 μl kinase buffer a containing 10% (vol/vol) glycerol for storage.6his-ets2. we transformed origami2 de3 (novagen) with the 6his-ets2 expression vector. we induced expression with 0.2 mm isopropyl-β-d-1-thiogalactopyranoside for 3 h at 22 °c. to purify 6his-ets2, we spun 50 ml cells and resuspended the pellet in 2 ml lysis buffer (pbs, 300 mm nacl, 10 mm imidazole, 1 mm dtt, and 0.1% nonidet p-40) containing a protease inhibitor mixture without edta (roche). we lysed the cells at 1.6 bar using a cell disruptor and spun the lysate 10 min at 20,000 × g. we collected the soluble fraction and added it to 200 μl cobalt beads (thermo scientific). after 1 h incubation at 4 °c on a rotating wheel, we washed four times with lysis buffer. to elute, we incubated beads 30 min with elution buffer (pbs, 250 mm imidazole, ph 7.6) containing the protease inhibitor mixture, spun 30 s at 10,000 × g, and collected the eluted protein.protein kinase assays. we mixed glutathione-sepharose beads (harboring gst-cdk10 wt or kd, either monomeric or complexed with strepii-cycm), 22.7 μm bsa, 15 mm dtt, 100 μm atp, 5 μci atp[γ-32p], 7.75 μm histone h1, or 1 μm 6his-ets2 and added kinase buffer a (25 mm tris·hcl, 10 mm mgcl2, 1 mm egta, 1 mm dtt, and 3.7 μm heparin, ph 7.5) up to a total volume of 30 μl. we incubated the reactions 30 min at 30 °c, added laemli sample buffer, heat-denatured the samples, and ran 10% bis-tris sds/page. we cut gel slices to detect gst-cdk10 and strepii-cycm by western blotting. we stained the gel slices containing the substrate with coomassie (r-250, bio-rad), dried them, and detected the incorporated radioactivity by autoradiography. we identified four unrelated girls with anogenital and renal malformations, dysmorphic facial features, normal intellect and syndactyly of toes. a similar combination of features had been reported previously in a mother–daughter pair1 (table 1 and supplementary note online). these authors noted clinical overlap with townes-brocks syndrome but suggested that the phenotype represented a separate autosomal dominant entity (mim601446). here we define the cardinal features of this syndrome as a characteristic facial appearance with apparent telecanthus and broad tripartite nasal tip, variable syndactyly of toes 2–5, hypoplastic labia, anal atresia and urogenital malformations (fig. 1a–h). we also observed a variety of other features (table 1).  figure 1: clinical and molecular characterization of star syndrome.  figure 1 : clinical and molecular characterization of star syndrome. (a–f) facial appearances of cases 1–3 (apparent telecanthus, dysplastic ears and thin upper lips; a,c,e), and toe syndactyly 2–5, 3–5 or 4–5 (b,d,f) in these cases illustrate recognizable features of star syndrome (specific parental consent has been obtained for publication of these photographs). anal atresia and hypoplastic labia are not shown. (g,h) x-ray films of the feet of case 2 showing only four rays on the left and delta-shaped 4th and 5th metatarsals on the right (h; compare to clinical picture in d). (i) array-cgh data. log2 ratio represents copy number loss of six probes spanning between 37.9 and 50.7 kb, with one probe positioned within fam58a. the deletion does not remove parts of other functional genes. (j) schematic structure of fam58a and position of the mutations. fam58a has five coding exons (boxes). the cyclin domain (green) is encoded by exons 2–4. the horizontal arrow indicates the deletion extending 5' in case 1, which includes exons 1 and 2, whereas the horizontal line below exon 5 indicates the deletion found in case 3, which removes exon 5 and some 3' sequence. the pink horizontal bars above the boxes indicate the amplicons used for qpcr and sequencing (one alternative exon 5 amplicon is not indicated because of space constraints). the mutation 201dupt (case 4) results in an immediate stop codon, and the 555+1g>a and 555-1g>a splice mutations in cases 2, 5 and 6 are predicted to be deleterious because they alter the conserved splice donor and acceptor site of intron 4, respectively.  full size image (97 kb)  table 1: clinical features in star syndrome cases  table 1 - clinical features in star syndrome cases  full table  on the basis of the phenotypic overlap with townes-brocks, okihiro and feingold syndromes, we analyzed sall1 (ref. 2), sall4 (ref. 3) and mycn4 but found no mutations in any of these genes (supplementary methods online). next, we carried out genome-wide high-resolution oligonucleotide array comparative genomic hybridization (cgh)5 analysis (supplementary methods) of genomic dna from the most severely affected individual (case 1, with lower lid coloboma, epilepsy and syringomyelia) and identified a heterozygous deletion of 37.9–50.7 kb on xq28, which removed exons 1 and 2 of fam58a (fig. 1i,j). using real-time pcr, we confirmed the deletion in the child and excluded it in her unaffected parents (supplementary fig. 1a online, supplementary methods and supplementary table 1 online). through cgh with a customized oligonucleotide array enriched in probes for xq28, followed by breakpoint cloning, we defined the exact deletion size as 40,068 bp (g.152,514,164_152,554,231del(chromosome x, ncbi build 36.2); fig. 1j and supplementary figs. 2,3 online). the deletion removes the coding regions of exons 1 and 2 as well as intron 1 (2,774 bp), 492 bp of intron 2, and 36,608 bp of 5' sequence, including the 5' utr and the entire krt18p48 pseudogene (ncbi gene id 340598). paternity was proven using routine methods. we did not find deletions overlapping fam58a in the available copy number variation (cnv) databases.  subsequently, we carried out qpcr analysis of the three other affected individuals (cases 2, 3 and 4) and the mother-daughter pair from the literature (cases 5 and 6). in case 3, we detected a de novo heterozygous deletion of 1.1–10.3 kb overlapping exon 5 (supplementary fig. 1b online). using xq28-targeted array cgh and breakpoint cloning, we identified a deletion of 4,249 bp (g.152,504,123_152,508,371del(chromosome x, ncbi build 36.2); fig. 1j and supplementary figs. 2,3), which removed 1,265 bp of intron 4, all of exon 5, including the 3' utr, and 2,454 bp of 3' sequence.  we found heterozygous fam58a point mutations in the remaining cases (fig. 1j, supplementary fig. 2, supplementary methods and supplementary table 1). in case 2, we identified the mutation 555+1g>a, affecting the splice donor site of intron 4. in case 4, we identified the frameshift mutation 201dupt, which immediately results in a premature stop codon n68xfsx1. in cases 5 and 6, we detected the mutation 556-1g>a, which alters the splice acceptor site of intron 4. we validated the point mutations and deletions by independent rounds of pcr and sequencing or by qpcr. we confirmed paternity and de novo status of the point mutations and deletions in all sporadic cases. none of the mutations were seen in the dna of 60 unaffected female controls, and no larger deletions involving fam58a were found in 93 unrelated array-cgh investigations.  by analyzing x-chromosome inactivation (supplementary methods and supplementary fig. 4 online), we found complete skewing of x inactivation in cases 1 and 3–6 and almost complete skewing in case 2, suggesting that cells carrying the mutation on the active x chromosome have a growth disadvantage during fetal development. using rt-pcr on rna from lymphoblastoid cells of case 2 (supplementary fig. 2), we did not find any aberrant splice products as additional evidence that the mutated allele is inactivated. furthermore, fam58a is subjected to x inactivation6. in cases 1 and 3, the parental origin of the deletions could not be determined, as a result of lack of informative snps. case 5, the mother of case 6, gave birth to two boys, both clinically unaffected (samples not available). we cannot exclude that the condition is lethal in males. no fetal losses were reported from any of the families.  the function of fam58a is unknown. the gene consists of five coding exons, and the 642-bp coding region encodes a protein of 214 amino acids. genbank lists a mrna length of 1,257 bp for the reference sequence (nm_152274.2). expression of the gene (by est data) was found in 27 of 48 adult tissues including kidney, colon, cervix and uterus, but not heart (ncbi expression viewer, unigene hs.496943). expression was also noted in 24 of 26 listed tumor tissues as well as in embryo and fetus. genes homologous to fam58a (ncbi homologene: 13362) are found on the x chromosome in the chimpanzee and the dog. the zebrafish has a similar gene on chromosome 23. however, in the mouse and rat, there are no true homologs. these species have similar but intronless genes on chromosomes 11 (mouse) and 10 (rat), most likely arising from a retrotransposon insertion event. on the murine x chromosome, the flanking genes atp2b3 and dusp9 are conserved, but only remnants of the fam58a sequence can be detected.  fam58a contains a cyclin-box-fold domain, a protein-binding domain found in cyclins with a role in cell cycle and transcription control. no human phenotype resulting from a cyclin gene mutation has yet been reported. homozygous knockout mice for ccnd1 (encoding cyclin d1) are viable but small and have reduced lifespan. they also have dystrophic changes of the retina, likely as a result of decreased cell proliferation and degeneration of photoreceptor cells during embryogenesis7, 8.  cyclin d1 colocalizes with sall4 in the nucleus, and both proteins cooperatively mediate transcriptional repression9. as the phenotype of our cases overlaps considerably with that of townes-brocks syndrome caused by sall1 mutations1, we carried out co-immunoprecipitation to find out if sall1 or sall4 would interact with fam58a in a manner similar to that observed for sall4 and cyclin d1. we found that fam58a interacts with sall1 but not with sall4 (supplementary fig. 5 online), supporting the hypothesis that fam58a and sall1 participate in the same developmental pathway.  how do fam58a mutations lead to star syndrome? growth retardation (all cases; table 1) and retinal abnormalities (three cases) are reminiscent of the reduced body size and retinal anomalies in cyclin d1 knockout mice7, 8. therefore, a proliferation defect might be partly responsible for star syndrome. to address this question, we carried out a knockdown of fam58a mrna followed by a proliferation assay. transfection of hek293 cells with three different fam58a-specific rnai oligonucleotides resulted in a significant reduction of both fam58a mrna expression and proliferation of transfected cells (supplementary methods and supplementary fig. 6 online), supporting the link between fam58a and cell proliferation.  we found that loss-of-function mutations of fam58a result in a rather homogeneous clinical phenotype. the additional anomalies in case 1 are likely to result from an effect of the 40-kb deletion on expression of a neighboring gene, possibly atp2b3 or dusp9. however, we cannot exclude that the homogeneous phenotype results from an ascertainment bias and that fam58a mutations, including missense changes, could result in a broader spectrum of malformations. the genes causing the overlapping phenotypes of star syndrome and townes-brocks syndrome seem to act in the same pathway. of note, mycn, a gene mutated in feingold syndrome, is a direct regulator of cyclin d2 (refs. 10,11); thus, it is worth exploring whether the phenotypic similarities between feingold and star syndrome might be explained by direct regulation of fam58a by mycn.  fam58a is located approximately 0.56 mb centromeric to mecp2 on xq28. duplications overlapping both mecp2 and fam58a have been described and are not associated with a clinical phenotype in females12, but no deletions overlapping both mecp2 and fam58a have been observed to date13. although other genes between fam58a and mecp2 have been implicated in brain development, fam58a and mecp2 are the only genes in this region known to result in x-linked dominant phenotypes; thus, deletion of both genes on the same allele might be lethal in both males and females.\""
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Text'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "53833d81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Text</th>\n",
       "      <th>Cleaned_Text</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>fam58a</td>\n",
       "      <td>truncating_mutations</td>\n",
       "      <td>cyclin-dependent kinases (cdks) regulate a var...</td>\n",
       "      <td>cyclindependent kinase cdks regulate variety f...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>cbl</td>\n",
       "      <td>w802*</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "      <td>abstract background nonsmall cell lung cancer ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>cbl</td>\n",
       "      <td>q249e</td>\n",
       "      <td>abstract background  non-small cell lung canc...</td>\n",
       "      <td>abstract background nonsmall cell lung cancer ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  \\\n",
       "0   0  fam58a  truncating_mutations   \n",
       "1   1     cbl                 w802*   \n",
       "2   2     cbl                 q249e   \n",
       "\n",
       "                                                Text  \\\n",
       "0  cyclin-dependent kinases (cdks) regulate a var...   \n",
       "1   abstract background  non-small cell lung canc...   \n",
       "2   abstract background  non-small cell lung canc...   \n",
       "\n",
       "                                        Cleaned_Text  Class  \n",
       "0  cyclindependent kinase cdks regulate variety f...      1  \n",
       "1  abstract background nonsmall cell lung cancer ...      2  \n",
       "2  abstract background nonsmall cell lung cancer ...      2  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[['ID', 'Gene', 'Variation', 'Text','Cleaned_Text', 'Class']]\n",
    "df.head(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ed5ac5bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Class\n",
       "7    952\n",
       "4    686\n",
       "1    566\n",
       "2    452\n",
       "6    273\n",
       "5    242\n",
       "3     89\n",
       "9     37\n",
       "8     19\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0a5f7881",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "x= df.drop(['Class','ID','Text'],axis=1)\n",
    "y = df['Class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f3d38b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42,stratify=y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c7d20a40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# label encoding\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "le_gene = LabelEncoder()\n",
    "le_var = LabelEncoder()\n",
    "x['Gene'] = le_gene.fit_transform(x['Gene'])\n",
    "x['Variation'] = le_var.fit_transform(x['Variation'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "128aa976",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "# Save the label encoders to a file\n",
    "with open('le_gene.pkl', 'wb') as f:\n",
    "    pickle.dump(le_gene, f)\n",
    "\n",
    "with open('le_var.pkl', 'wb') as f:\n",
    "    pickle.dump(le_var, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c09bff8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Cleaned_Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rheb</td>\n",
       "      <td>y35n</td>\n",
       "      <td>gene encoding component signaling axis frequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>pten</td>\n",
       "      <td>k128n</td>\n",
       "      <td>pten phosphatase tensin homolog phosphatase un...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Variation                                       Cleaned_Text\n",
       "68    rheb      y35n  gene encoding component signaling axis frequen...\n",
       "2222  pten     k128n  pten phosphatase tensin homolog phosphatase un..."
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9d02b5ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tf-idf vectorization\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "tfidf = TfidfVectorizer(max_features=5000)\n",
    "tfidf.fit(df['Cleaned_Text'])\n",
    "x_train_tfidf = tfidf.transform(x_train['Cleaned_Text']).toarray()\n",
    "x_test_tfidf = tfidf.transform(x_test['Cleaned_Text']).toarray()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9c702e36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the tfidf vectorizer to a file\n",
    "with open('tfidf.pkl', 'wb') as f:\n",
    "    pickle.dump(tfidf, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d8b5e540",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Cleaned_Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rheb</td>\n",
       "      <td>y35n</td>\n",
       "      <td>gene encoding component signaling axis frequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>pten</td>\n",
       "      <td>k128n</td>\n",
       "      <td>pten phosphatase tensin homolog phosphatase un...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Variation                                       Cleaned_Text\n",
       "68    rheb      y35n  gene encoding component signaling axis frequen...\n",
       "2222  pten     k128n  pten phosphatase tensin homolog phosphatase un..."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2513d5d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2652, 5000)\n"
     ]
    }
   ],
   "source": [
    "x_train_tfidf\n",
    "print(x_train_tfidf.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4bda2621",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique words in the corpus: 5000\n",
      "['aa' 'aacr' 'ab' ... 'μl' 'μm' 'μmoll']\n"
     ]
    }
   ],
   "source": [
    "unique_words = tfidf.get_feature_names_out()\n",
    "print(\"Number of unique words in the corpus:\", len(unique_words))\n",
    "print(unique_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0f0d3058",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>aacr</th>\n",
       "      <th>ab</th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>abc</th>\n",
       "      <th>abcam</th>\n",
       "      <th>abd</th>\n",
       "      <th>abdominal</th>\n",
       "      <th>aberrant</th>\n",
       "      <th>aberrantly</th>\n",
       "      <th>...</th>\n",
       "      <th>βgalactosidase</th>\n",
       "      <th>βsheet</th>\n",
       "      <th>γphosphate</th>\n",
       "      <th>γsecretase</th>\n",
       "      <th>δbraf</th>\n",
       "      <th>μg</th>\n",
       "      <th>μgml</th>\n",
       "      <th>μl</th>\n",
       "      <th>μm</th>\n",
       "      <th>μmoll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001251</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.067933</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007272</td>\n",
       "      <td>0.003815</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aa  aacr   ab  abbreviation  abc  abcam  abd  abdominal  aberrant  \\\n",
       "0  0.0   0.0  0.0           0.0  0.0    0.0  0.0        0.0  0.001251   \n",
       "1  0.0   0.0  0.0           0.0  0.0    0.0  0.0        0.0  0.000000   \n",
       "\n",
       "   aberrantly  ...  βgalactosidase  βsheet  γphosphate  γsecretase  δbraf  \\\n",
       "0         0.0  ...             0.0     0.0    0.067933         0.0    0.0   \n",
       "1         0.0  ...             0.0     0.0    0.000000         0.0    0.0   \n",
       "\n",
       "         μg  μgml        μl        μm  μmoll  \n",
       "0  0.005056   0.0  0.007272  0.003815    0.0  \n",
       "1  0.000000   0.0  0.000000  0.000000    0.0  \n",
       "\n",
       "[2 rows x 5000 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train = pd.DataFrame(x_train_tfidf, columns=unique_words)\n",
    "df_train.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "84792855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2652, 5000)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "41c1202b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>aacr</th>\n",
       "      <th>ab</th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>abc</th>\n",
       "      <th>abcam</th>\n",
       "      <th>abd</th>\n",
       "      <th>abdominal</th>\n",
       "      <th>aberrant</th>\n",
       "      <th>aberrantly</th>\n",
       "      <th>...</th>\n",
       "      <th>βgalactosidase</th>\n",
       "      <th>βsheet</th>\n",
       "      <th>γphosphate</th>\n",
       "      <th>γsecretase</th>\n",
       "      <th>δbraf</th>\n",
       "      <th>μg</th>\n",
       "      <th>μgml</th>\n",
       "      <th>μl</th>\n",
       "      <th>μm</th>\n",
       "      <th>μmoll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006431</td>\n",
       "      <td>0.01424</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.009706</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.019918</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aa  aacr   ab  abbreviation  abc  abcam  abd  abdominal  aberrant  \\\n",
       "0  0.0   0.0  0.0      0.004855  0.0    0.0  0.0        0.0       0.0   \n",
       "1  0.0   0.0  0.0      0.000000  0.0    0.0  0.0        0.0       0.0   \n",
       "\n",
       "   aberrantly  ...  βgalactosidase  βsheet  γphosphate  γsecretase  δbraf  \\\n",
       "0         0.0  ...             0.0     0.0         0.0         0.0    0.0   \n",
       "1         0.0  ...             0.0     0.0         0.0         0.0    0.0   \n",
       "\n",
       "         μg     μgml      μl        μm  μmoll  \n",
       "0  0.006431  0.01424  0.0074  0.009706    0.0  \n",
       "1  0.000000  0.00000  0.0000  0.019918    0.0  \n",
       "\n",
       "[2 rows x 5000 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test = pd.DataFrame(x_test_tfidf, columns=unique_words)\n",
    "df_test.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b57cde85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>aacr</th>\n",
       "      <th>ab</th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>abc</th>\n",
       "      <th>abcam</th>\n",
       "      <th>abd</th>\n",
       "      <th>abdominal</th>\n",
       "      <th>aberrant</th>\n",
       "      <th>aberrantly</th>\n",
       "      <th>...</th>\n",
       "      <th>βgalactosidase</th>\n",
       "      <th>βsheet</th>\n",
       "      <th>γphosphate</th>\n",
       "      <th>γsecretase</th>\n",
       "      <th>δbraf</th>\n",
       "      <th>μg</th>\n",
       "      <th>μgml</th>\n",
       "      <th>μl</th>\n",
       "      <th>μm</th>\n",
       "      <th>μmoll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001251</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.067933</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005056</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.007272</td>\n",
       "      <td>0.003815</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aa  aacr   ab  abbreviation  abc  abcam  abd  abdominal  aberrant  \\\n",
       "0  0.0   0.0  0.0           0.0  0.0    0.0  0.0        0.0  0.001251   \n",
       "1  0.0   0.0  0.0           0.0  0.0    0.0  0.0        0.0  0.000000   \n",
       "\n",
       "   aberrantly  ...  βgalactosidase  βsheet  γphosphate  γsecretase  δbraf  \\\n",
       "0         0.0  ...             0.0     0.0    0.067933         0.0    0.0   \n",
       "1         0.0  ...             0.0     0.0    0.000000         0.0    0.0   \n",
       "\n",
       "         μg  μgml        μl        μm  μmoll  \n",
       "0  0.005056   0.0  0.007272  0.003815    0.0  \n",
       "1  0.000000   0.0  0.000000  0.000000    0.0  \n",
       "\n",
       "[2 rows x 5000 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e7e2d445",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa</th>\n",
       "      <th>aacr</th>\n",
       "      <th>ab</th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>abc</th>\n",
       "      <th>abcam</th>\n",
       "      <th>abd</th>\n",
       "      <th>abdominal</th>\n",
       "      <th>aberrant</th>\n",
       "      <th>aberrantly</th>\n",
       "      <th>...</th>\n",
       "      <th>βgalactosidase</th>\n",
       "      <th>βsheet</th>\n",
       "      <th>γphosphate</th>\n",
       "      <th>γsecretase</th>\n",
       "      <th>δbraf</th>\n",
       "      <th>μg</th>\n",
       "      <th>μgml</th>\n",
       "      <th>μl</th>\n",
       "      <th>μm</th>\n",
       "      <th>μmoll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006431</td>\n",
       "      <td>0.01424</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0.009706</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.019918</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    aa  aacr   ab  abbreviation  abc  abcam  abd  abdominal  aberrant  \\\n",
       "0  0.0   0.0  0.0      0.004855  0.0    0.0  0.0        0.0       0.0   \n",
       "1  0.0   0.0  0.0      0.000000  0.0    0.0  0.0        0.0       0.0   \n",
       "\n",
       "   aberrantly  ...  βgalactosidase  βsheet  γphosphate  γsecretase  δbraf  \\\n",
       "0         0.0  ...             0.0     0.0         0.0         0.0    0.0   \n",
       "1         0.0  ...             0.0     0.0         0.0         0.0    0.0   \n",
       "\n",
       "         μg     μgml      μl        μm  μmoll  \n",
       "0  0.006431  0.01424  0.0074  0.009706    0.0  \n",
       "1  0.000000  0.00000  0.0000  0.019918    0.0  \n",
       "\n",
       "[2 rows x 5000 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d70081b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2652, 5000)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3a5dbb87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Cleaned_Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rheb</td>\n",
       "      <td>y35n</td>\n",
       "      <td>gene encoding component signaling axis frequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>pten</td>\n",
       "      <td>k128n</td>\n",
       "      <td>pten phosphatase tensin homolog phosphatase un...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Variation                                       Cleaned_Text\n",
       "68    rheb      y35n  gene encoding component signaling axis frequen...\n",
       "2222  pten     k128n  pten phosphatase tensin homolog phosphatase un..."
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "482460bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train.drop(['Cleaned_Text'],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5f92af3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_test.drop(['Cleaned_Text'],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "72c49eeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rheb</td>\n",
       "      <td>y35n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>pten</td>\n",
       "      <td>k128n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>ep300</td>\n",
       "      <td>c1385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>cdkn2a</td>\n",
       "      <td>i49s</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fam58a</td>\n",
       "      <td>truncating_mutations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3185</th>\n",
       "      <td>rara</td>\n",
       "      <td>r276w</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3245</th>\n",
       "      <td>ddr2</td>\n",
       "      <td>g774v</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2635</th>\n",
       "      <td>brca1</td>\n",
       "      <td>p1856t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2060</th>\n",
       "      <td>mycn</td>\n",
       "      <td>p44l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1424</th>\n",
       "      <td>fgfr3</td>\n",
       "      <td>a391e</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2652 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene             Variation\n",
       "68      rheb                  y35n\n",
       "2222    pten                 k128n\n",
       "355    ep300                 c1385\n",
       "685   cdkn2a                  i49s\n",
       "0     fam58a  truncating_mutations\n",
       "...      ...                   ...\n",
       "3185    rara                 r276w\n",
       "3245    ddr2                 g774v\n",
       "2635   brca1                p1856t\n",
       "2060    mycn                  p44l\n",
       "1424   fgfr3                 a391e\n",
       "\n",
       "[2652 rows x 2 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "371bfcce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3234</th>\n",
       "      <td>ntrk3</td>\n",
       "      <td>etv6-ntrk3_fusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2165</th>\n",
       "      <td>pten</td>\n",
       "      <td>p38l</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>tp53</td>\n",
       "      <td>r282w</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1809</th>\n",
       "      <td>rhoa</td>\n",
       "      <td>t37a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cbl</td>\n",
       "      <td>n454d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2596</th>\n",
       "      <td>brca1</td>\n",
       "      <td>g1788d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>map2k1</td>\n",
       "      <td>q56_v60del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2615</th>\n",
       "      <td>brca1</td>\n",
       "      <td>t47d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2486</th>\n",
       "      <td>brca1</td>\n",
       "      <td>s1841r</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>pdgfra</td>\n",
       "      <td>y849s</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>664 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene          Variation\n",
       "3234   ntrk3  etv6-ntrk3_fusion\n",
       "2165    pten               p38l\n",
       "504     tp53              r282w\n",
       "1809    rhoa               t37a\n",
       "3        cbl              n454d\n",
       "...      ...                ...\n",
       "2596   brca1             g1788d\n",
       "1998  map2k1         q56_v60del\n",
       "2615   brca1               t47d\n",
       "2486   brca1             s1841r\n",
       "904   pdgfra              y849s\n",
       "\n",
       "[664 rows x 2 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d09101d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df =pd.merge(x_train,df_train,left_index=True,right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "0c26efa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df =pd.merge(x_test,df_test,left_index=True,right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b4ef69dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2652, 2)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "cde39d33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>aa</th>\n",
       "      <th>aacr</th>\n",
       "      <th>ab</th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>abc</th>\n",
       "      <th>abcam</th>\n",
       "      <th>abd</th>\n",
       "      <th>abdominal</th>\n",
       "      <th>...</th>\n",
       "      <th>βgalactosidase</th>\n",
       "      <th>βsheet</th>\n",
       "      <th>γphosphate</th>\n",
       "      <th>γsecretase</th>\n",
       "      <th>δbraf</th>\n",
       "      <th>μg</th>\n",
       "      <th>μgml</th>\n",
       "      <th>μl</th>\n",
       "      <th>μm</th>\n",
       "      <th>μmoll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rheb</td>\n",
       "      <td>y35n</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003909</td>\n",
       "      <td>0.00344</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002622</td>\n",
       "      <td>0.029802</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>pten</td>\n",
       "      <td>k128n</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene Variation   aa  aacr        ab  abbreviation  abc  abcam  abd  \\\n",
       "68    rheb      y35n  0.0   0.0  0.003909       0.00344  0.0    0.0  0.0   \n",
       "2222  pten     k128n  0.0   0.0  0.000000       0.00000  0.0    0.0  0.0   \n",
       "\n",
       "      abdominal  ...  βgalactosidase  βsheet  γphosphate  γsecretase  δbraf  \\\n",
       "68          0.0  ...             0.0     0.0         0.0         0.0    0.0   \n",
       "2222        0.0  ...             0.0     0.0         0.0         0.0    0.0   \n",
       "\n",
       "       μg  μgml        μl        μm  μmoll  \n",
       "68    0.0   0.0  0.002622  0.029802    0.0  \n",
       "2222  0.0   0.0  0.000000  0.000000    0.0  \n",
       "\n",
       "[2 rows x 5002 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "92eb723e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>aa</th>\n",
       "      <th>aacr</th>\n",
       "      <th>ab</th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>abc</th>\n",
       "      <th>abcam</th>\n",
       "      <th>abd</th>\n",
       "      <th>abdominal</th>\n",
       "      <th>...</th>\n",
       "      <th>βgalactosidase</th>\n",
       "      <th>βsheet</th>\n",
       "      <th>γphosphate</th>\n",
       "      <th>γsecretase</th>\n",
       "      <th>δbraf</th>\n",
       "      <th>μg</th>\n",
       "      <th>μgml</th>\n",
       "      <th>μl</th>\n",
       "      <th>μm</th>\n",
       "      <th>μmoll</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>tp53</td>\n",
       "      <td>r282w</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007551</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cbl</td>\n",
       "      <td>n454d</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004827</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.014327</td>\n",
       "      <td>0.015860</td>\n",
       "      <td>0.003297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene Variation   aa  aacr   ab  abbreviation  abc     abcam  abd  \\\n",
       "504  tp53     r282w  0.0   0.0  0.0           0.0  0.0  0.000000  0.0   \n",
       "3     cbl     n454d  0.0   0.0  0.0           0.0  0.0  0.004827  0.0   \n",
       "\n",
       "     abdominal  ...  βgalactosidase  βsheet  γphosphate  γsecretase  δbraf  \\\n",
       "504        0.0  ...        0.000000     0.0         0.0         0.0    0.0   \n",
       "3          0.0  ...        0.004901     0.0         0.0         0.0    0.0   \n",
       "\n",
       "           μg      μgml        μl   μm  μmoll  \n",
       "504  0.000000  0.007551  0.000000  0.0    0.0  \n",
       "3    0.014327  0.015860  0.003297  0.0    0.0  \n",
       "\n",
       "[2 rows x 5002 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ba9420c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2652, 2)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4c541bad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2116, 5002)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6e068c1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2652, 5000)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train_tfidf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "8da2f9b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "<class 'numpy.ndarray'>\n"
     ]
    }
   ],
   "source": [
    "print(type(x_train))\n",
    "print(type(x_train_tfidf))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "50a9d650",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.sparse import issparse\n",
    "import numpy as np\n",
    "def safe_to_dense(X):\n",
    "    if issparse(X):\n",
    "        return X.toarray()\n",
    "    elif hasattr(X, \"to_numpy\"):  # DataFrame\n",
    "        return X.to_numpy()\n",
    "    else:  # NumPy array বা list\n",
    "        return np.array(X)\n",
    "\n",
    "merge_df_training = np.hstack((\n",
    "    safe_to_dense(x_train),\n",
    "    safe_to_dense(x_train_tfidf)\n",
    "))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "64d90199",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_df_testing = np.hstack((\n",
    "    safe_to_dense(x_test),\n",
    "    safe_to_dense(x_test_tfidf)\n",
    "))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65d3f1b1",
   "metadata": {},
   "source": [
    "# smote\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7e7be3c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3316, 6)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e180c745",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2652,)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b5c988ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train_numeric = x_train.select_dtypes(include=[np.number])\n",
    "merge_df_training = np.hstack((\n",
    "    x_train_numeric.to_numpy(),\n",
    "    x_train_tfidf\n",
    "))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b3831105",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_test_numeric = x_test.select_dtypes(include=[np.number])\n",
    "merge_df_testing = np.hstack((\n",
    "    x_test_numeric.to_numpy(),\n",
    "    x_test_tfidf\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2b1b9f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_df_training = merge_df_training.astype(float)\n",
    "merge_df_testing = merge_df_testing.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "ba349ded",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting imblearn\n",
      "  Downloading imblearn-0.0-py2.py3-none-any.whl.metadata (355 bytes)\n",
      "Collecting imbalanced-learn (from imblearn)\n",
      "  Downloading imbalanced_learn-0.14.0-py3-none-any.whl.metadata (8.8 kB)\n",
      "Requirement already satisfied: numpy<3,>=1.25.2 in ./venv/lib/python3.12/site-packages (from imbalanced-learn->imblearn) (2.3.3)\n",
      "Requirement already satisfied: scipy<2,>=1.11.4 in ./venv/lib/python3.12/site-packages (from imbalanced-learn->imblearn) (1.16.2)\n",
      "Requirement already satisfied: scikit-learn<2,>=1.4.2 in ./venv/lib/python3.12/site-packages (from imbalanced-learn->imblearn) (1.7.2)\n",
      "Requirement already satisfied: joblib<2,>=1.2.0 in ./venv/lib/python3.12/site-packages (from imbalanced-learn->imblearn) (1.5.2)\n",
      "Requirement already satisfied: threadpoolctl<4,>=2.0.0 in ./venv/lib/python3.12/site-packages (from imbalanced-learn->imblearn) (3.6.0)\n",
      "Downloading imblearn-0.0-py2.py3-none-any.whl (1.9 kB)\n",
      "Downloading imbalanced_learn-0.14.0-py3-none-any.whl (239 kB)\n",
      "Installing collected packages: imbalanced-learn, imblearn\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2/2\u001b[0m [imblearn]\n",
      "\u001b[1A\u001b[2KSuccessfully installed imbalanced-learn-0.14.0 imblearn-0.0\n"
     ]
    }
   ],
   "source": [
    "# imblearn for handling imbalanced data\n",
    "!pip install imblearn\n",
    "from imblearn.over_sampling import SMOTE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e4d9922d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#smote \n",
    "\n",
    "from imblearn.over_sampling import SMOTE\n",
    "smote = SMOTE()\n",
    "x_train_res, y_train_res = smote.fit_resample(merge_df_training, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "9d44b16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((6849, 5000), (6849,))"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_train_res.shape, y_train_res.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "8adb1356",
   "metadata": {},
   "outputs": [],
   "source": [
    "# random forest classifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "rf = RandomForestClassifier()\n",
    "rf.fit(x_train_res, y_train_res)\n",
    "y_pred = rf.predict(merge_df_testing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "aca19312",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy is :  0.625\n",
      "classification report : \n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "           1       0.56      0.51      0.54       113\n",
      "           2       0.59      0.56      0.57        91\n",
      "           3       0.30      0.39      0.34        18\n",
      "           4       0.69      0.64      0.66       137\n",
      "           5       0.34      0.40      0.37        48\n",
      "           6       0.82      0.60      0.69        55\n",
      "           7       0.70      0.81      0.75       191\n",
      "           8       0.00      0.00      0.00         4\n",
      "           9       1.00      0.71      0.83         7\n",
      "\n",
      "    accuracy                           0.62       664\n",
      "   macro avg       0.56      0.51      0.53       664\n",
      "weighted avg       0.63      0.62      0.62       664\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", result.shape[0])\n",
      "/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", result.shape[0])\n",
      "/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", result.shape[0])\n"
     ]
    }
   ],
   "source": [
    "# accuracy\n",
    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
    "acc = accuracy_score(y_test, y_pred)\n",
    "print(\"accuracy is : \", acc)\n",
    "print(\"classification report : \\n\", classification_report(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e88ec4ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy is :  0.5993975903614458\n",
      "classification report : \n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "           1       0.54      0.50      0.52       113\n",
      "           2       0.51      0.53      0.52        91\n",
      "           3       0.30      0.56      0.39        18\n",
      "           4       0.70      0.58      0.64       137\n",
      "           5       0.33      0.48      0.39        48\n",
      "           6       0.69      0.62      0.65        55\n",
      "           7       0.74      0.72      0.73       191\n",
      "           8       0.67      0.50      0.57         4\n",
      "           9       0.67      0.86      0.75         7\n",
      "\n",
      "    accuracy                           0.60       664\n",
      "   macro avg       0.57      0.59      0.57       664\n",
      "weighted avg       0.62      0.60      0.61       664\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):\n",
      "STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT\n",
      "\n",
      "Increase the number of iterations to improve the convergence (max_iter=100).\n",
      "You might also want to scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n"
     ]
    }
   ],
   "source": [
    "# logistic regression\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix\n",
    "log_reg = LogisticRegression(penalty='l2',C=10,class_weight='balanced',random_state=1,solver='lbfgs',n_jobs=-1)\n",
    "log_reg.fit(x_train_res, y_train_res)\n",
    "y_pred = log_reg.predict(merge_df_testing)  \n",
    "acc = accuracy_score(y_test, y_pred)\n",
    "print(\"accuracy is : \", acc)\n",
    "print(\"classification report : \\n\", classification_report(y_test, y_pred))  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e83fb506",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best K: {'n_neighbors': 30}\n",
      "Best Score: 0.6261436363368851\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "\n",
    "parameters = {'n_neighbors': list(range(30, 100, 5))}\n",
    "neigh = KNeighborsClassifier()\n",
    "\n",
    "# GridSearchCV object\n",
    "Gsv = GridSearchCV(neigh, parameters,cv=5, scoring='f1_macro')\n",
    "\n",
    "# IMPORTANT: fit on training data\n",
    "Gsv.fit(x_train_res, y_train_res)\n",
    "\n",
    "# Now you can get the best parameters\n",
    "print(\"Best K:\", Gsv.best_params_)\n",
    "print(\"Best Score:\", Gsv.best_score_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "df5b4472",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy is :  0.36596385542168675\n",
      "classification report : \n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "           1       0.44      0.37      0.40       113\n",
      "           2       0.41      0.55      0.47        91\n",
      "           3       0.10      0.50      0.16        18\n",
      "           4       0.66      0.32      0.43       137\n",
      "           5       0.19      0.42      0.26        48\n",
      "           6       0.57      0.55      0.56        55\n",
      "           7       0.72      0.21      0.33       191\n",
      "           8       0.05      0.50      0.08         4\n",
      "           9       0.17      0.71      0.28         7\n",
      "\n",
      "    accuracy                           0.37       664\n",
      "   macro avg       0.37      0.46      0.33       664\n",
      "weighted avg       0.54      0.37      0.39       664\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# knn classifier\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "knn = KNeighborsClassifier(n_neighbors=30)\n",
    "knn.fit(x_train_res, y_train_res)\n",
    "y_pred = knn.predict(merge_df_testing)  \n",
    "acc = accuracy_score(y_test, y_pred)\n",
    "print(\"accuracy is : \", acc)\n",
    "print(\"classification report : \\n\", classification_report(y_test, y_pred))  \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b56887c",
   "metadata": {},
   "source": [
    "# highly Imbalance data "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb9ac3f9",
   "metadata": {},
   "source": [
    "## this dataset is highly imbalance , specilly in class coloumn so w have to apply smote "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4f6631a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):\n",
      "STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT\n",
      "\n",
      "Increase the number of iterations to improve the convergence (max_iter=100).\n",
      "You might also want to scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABKUAAAKyCAYAAAAEvm1SAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAAzshJREFUeJzs3Qd4VGX2x/FfOiEhoffeexdBxAqIZe2KomJl7X3X1V17WXd1VSxY/pa1K6hYwYogiAgConSk9w7phZT/c95hIiC6kMzcKfl+nuc+kym5eQM3M/ee97znxJSWlpYKAAAAAAAA8FCslz8MAAAAAAAAMASlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAPZx1FFHua08mjdvrosuukjhysZmYwyVcP/3AQAA3iEoBQAAwtLKlSsVExPjtvvvv3+/rznvvPPc86mpqYo0/t9tf9sVV1yhSPbdd9/p7rvv1s6dO0M9FAAAEMbiQz0AAACAP1KlShW99dZbuv322/d6PCcnRx9++KF7PlINGjRIw4cP/83jbdu2VaQHpe655x6XEVW9evW9nlu8eLFiY5kXBQAABKUAAECYO+GEEzR27Fj99NNP6tatW9njFpAqLCzUkCFD9PXXXysSWfDp/PPPV2WSlJQU6iEAAIAwwTQVAAAIa/369VOLFi305ptv7vX4G2+84QJSNWvW3O/3Pf300+rUqZMLgjRs2FBXX331fpeT/d///Z9atWql5ORk9enTR1OmTNnv/goKCnTXXXepdevWbp9NmjTRLbfc4h4PlmuuucYtTczNzf3Nc+eee67q16+v4uLisiDdiSee6H5XG5/9Tvfdd1/Z879n0qRJbsmg3e5v+eTLL79c9tjPP//ssp9atmzpMtTs519yySXatm1b2Wts2d5f//pX97X9v/mXJNr+fq+m1PLly3XWWWe5/8uqVauqb9++Gjdu3H7HOWbMGD3wwANq3LixG8Oxxx6rpUuXHsS/KgAACBdkSgEAgLBnAZjXX39d//rXv1xgYuvWrfriiy/02muv6bPPPvvN6y0wYsvHBg4cqCuvvNItGXvmmWf0ww8/aOrUqUpISHCve/HFF3X55ZfrsMMO0w033OCCIyeffLILjljQya+kpMQ9/u233+rPf/6zOnTooLlz5+qxxx7TkiVL9MEHH5Tr98rPz3e/y77S0tKUmJiooUOHatSoUS5AY0EbPwtSffzxxy64ExcX5x6z4JEFsG666SZ3a9ljd955pzIzM/Xwww8rEL788kv3b3TxxRe7gNT8+fNdUM9uv//+e/d/c/rpp7t/E1tyaf8+tWvXdt9bp06d/e5z06ZN7t/ffqfrrrtOtWrV0iuvvOL+vd99912ddtppe73ejgFb/veXv/xFGRkZeuihh1xtsenTpwfkdwQAAB4qBQAACEMrVqwotVOVhx9+uHTevHnu6ylTprjnRo0aVZqamlqak5NTeuGFF5ampKSUfd/mzZtLExMTSwcPHlxaXFxc9vhTTz3l9vHSSy+5+4WFhaV169Yt7d69e2lBQUHZ6/7v//7Pve7II48se+y1114rjY2NLfv5fs8++6x77dSpU8sea9asmRvT/2Lf93vbW2+95V5TUlJS2qhRo9Izzjhjr+8dM2aMe93kyZPLHsvNzf3Nz7j88stLq1atWpqfn1/2mI3Nxug3ceJEty+73d+//3//+98//Bk21n3HYv9n9pjtY1/7/vvccMMNe/3fmqysrNIWLVqUNm/evOz/0D/ODh067PX/9fjjj7vH586d+5ufBQAAwhvL9wAAQNizZXhdu3Z12TfGlvKdcsopbqnXvr766itXa8oyn/YsqD1ixAiXgeRfFjZz5kxt3rzZdbqzrCQ/yz5KT0/fa5/vvPOOy45q3769y2zyb8ccc4x7fuLEieX6vex3sOyjfbejjz7aPW+ZR5YhNX78eGVnZ5d93+jRo9WoUSMdfvjhZY/Z8kO/rKwsN74BAwa4DKRFixYpEPb8Gf4sL1tqZ2bPnl2ufdrvZssm9/xdLNPLMtJsyd+CBQv2er1lae35/2W/o7EMLgAAEFkISgEAgIgwbNgwFxyy+kHW3c3u78+qVavcbbt27fZ63AIZVgvJ/7z/tk2bNnu9zpb22ev29Msvv7glarYEbc/N3yXPglvlYXWRbInhvlu9evXKXmNL+PLy8vTRRx+5+xacskCOBassaOVn47OlbhZQs+Cbjc9fRN2WuQXC9u3bdf3117vxWYDKfobVjarIz7D/h33/r4wFAf3P76lp06Z73a9Ro4a73bFjR7l+PgAACB1qSgEAgIipK3Xbbbe5jCerOzR48GDPfrbVlOrSpYseffTR/T6/Z/2pQLNMJCsObgW+LRBntaQsSGXBKj8r4H7kkUe6YNS9997ripxbEXDLXvrb3/7mxv979gxs7Wl/BdLPPvtsFxC0Qubdu3d3GU22bys4/0c/I5D8NbT25VsRCQAAIglBKQAAEBEsQ6Z///6uC5sVL4+P3/9pTLNmzdytFTffM+PJlvStWLHCZSLt+TrLgvIvwzO7du1yr+vWrVvZYxbk+emnn1ynt98L4gSTBYMef/xxV7Tclu5ZkMq/bM7Yv4l1wBs7dqyOOOKIssft9/hf/JlG+3Ym3DdDyTKRJkyY4ArIWwF1P/v329fB/BvZ/4P9X+3Lv+TQ//8EAACiD8v3AABAxLj//vt111136dprr/3d11jQyZbqPfHEE3tlz1inPVtiduKJJ7r7vXv3dsvPnn32WRew8rMudvsGaCwotG7dOj3//PO/+XmWtZSTk6NgsqyogoIC15XOug3aePaXPbTn72u/09NPP/0/921BH/v+yZMn7/X4vt+7v59hRo4c+Zt9pqSkuNt9/x3354QTTtCMGTM0bdq0ssfs39O6+lnwrWPHjv9zHwAAIDKRKQUAACKGLVGz7Y9YoMmW+VlGjy0rO/nkk10mjgVZDjnkkLI6S1Y7yoJcl19+ucuUssCPZRb997///U1NqQsuuMAtn7Oi6FbU3DK2bHmbZfPY459//rkLch2sJUuW6PXXX//N41azadCgQWX3e/bsqdatW+sf//iHC07tuXTPHHbYYS7j6cILL9R1113nMpVee+21A1rSZjWorD7Vk08+6b7PssI++eST39TJsqWBloX10EMPuWwyK7T+xRdf7Dcbq1evXu7WxnvOOee4f+s//elPZcGqPd16662ugP3xxx/vxl6zZk0XfLP9vvfee3sVqwcAANGFoBQAAIg6d999twtOPfXUU7rxxhtdoMO6uf3zn/90ARI/e8yCSw8//LCrk2R1o6yg+B133LHX/iww8sEHH+ixxx7Tq6++qvfff991/rPglRX+9hc8P1j+bnv7ssDbnkEpY4GoBx54wAWnLEi1J6uxZYGkm2++WbfffrsLUFnwzZYbHnfccf9zHBaQskCTZY0lJSW5TCz7N+ncufNer7Ouh5alNmrUKBfwsrpen376qRo2bLjX6yz4d99997n9WWaX1ZuyINP+glIWgLM6VVb7ysZhXf2s06LVzvJntQEAgOgUU0pVSAAAAAAAAHiMfGgAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJUrKGWth60TixXHtG4vVkD0f5k0aZIr7mlFOK3Qp7VtBgAAAAAAQGQJaVAqJydH3bp1cx1cDoR1bbEuLEcffbTmzJmjG264QZdddplrwwwAAAAAAIDIETbd9yxTytorn3rqqb/7GmsVPG7cOM2bN6/ssXPOOUc7d+507YYBAAAAAAAQGeIVQaZNm6aBAwfu9dhxxx3nMqZ+T0FBgdv8SkpKtH37dtWqVcsFwgAAAAAAABA4lv+UlZXlyjXFxsZGR1Bq48aNqlev3l6P2f3MzEzl5eUpOTn5N9/z4IMP6p577vFwlAAAAAAAAFizZo0aN24cHUGp8rjtttt00003ld3PyMhQ06ZN3T9MWlpaSMcGAAAAAAAQbSx5qEmTJqpWrdofvi6iglL169fXpk2b9nrM7ltwaX9ZUsa69Nm2L/seglIAAAAAAADB8b/KJoW0+97B6tevnyZMmLDXY19++aV7HAAAAAAAAJEjpEGp7OxszZkzx21mxYoV7uvVq1eXLb0bPnx42euvuOIKLV++XLfccosWLVqkp59+WmPGjNGNN94Yst8BAAAAAAAAERaUmjlzpnr06OE2Y7Wf7Os777zT3d+wYUNZgMq0aNFC48aNc9lR3bp10yOPPKIXXnjBdeADAAAAAABA5IgptT59lazYVnp6uit4Tk0pAAAAAACA0MReIqqmFAAAAAAAAKIDQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAQOULSo0aNUrNmzdXlSpVdOihh2rGjBl/+PqRI0eqXbt2Sk5OVpMmTXTjjTcqPz/fs/ECAAAAAAAgwoNSo0eP1k033aS77rpLs2fPVrdu3XTcccdp8+bN+339m2++qVtvvdW9fuHChXrxxRfdPv7+9797PnYAAAAAAABEaFDq0Ucf1YgRI3TxxRerY8eOevbZZ1W1alW99NJL+339d999p/79+2vYsGEuu2rw4ME699xz/2d2FQAAAAAAAMJLyIJShYWFmjVrlgYOHPjrYGJj3f1p06bt93sOO+ww9z3+INTy5cs1fvx4nXDCCb/7cwoKCpSZmbnXBgAAAAAAgNCKD9UP3rp1q4qLi1WvXr29Hrf7ixYt2u/3WIaUfd/hhx+u0tJSFRUV6YorrvjD5XsPPvig7rnnnoCPHwAAAAAAABFc6PxgTJo0Sf/85z/19NNPuxpUY8eO1bhx43Tffff97vfcdtttysjIKNvWrFnj6ZgBAAAAAAAQRplStWvXVlxcnDZt2rTX43a/fv36+/2eO+64QxdccIEuu+wyd79Lly7KycnRn//8Z/3jH/9wy//2lZSU5DYAAAAAAACEj5BlSiUmJqpXr16aMGFC2WMlJSXufr9+/fb7Pbm5ub8JPFlgy9hyPgAAAAAAAESGkGVKmZtuukkXXnihevfurT59+mjkyJEu88m68Znhw4erUaNGri6U+dOf/uQ69vXo0UOHHnqoli5d6rKn7HF/cAoAAAAAAADhL6RBqaFDh2rLli268847tXHjRnXv3l2fffZZWfHz1atX75UZdfvttysmJsbdrlu3TnXq1HEBqQceeCCEvwUAAAAAAAAOVkxpJVv3lpmZqfT0dFf0PC0tLdTDAQAAAAAAqJSxl4jqvgcAAAAAAIDoQFAKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAUCmUlpaGeggAAGAPBKUAAAAQ1fLz891tTEwMgSkAAMIIQSkAAABEpV27dunDDz/Ueeedp0svvVTLli1zgamffvrJbYWFhe51BKoAAAiN+BD9XAAAACCo3n77bd1yyy3atGmTu1+rVi3Vq1dPjzzyiDZu3KiOHTvq5ZdfVu/evV1gygJWAADAOwSlAAAAEJXefPNNF5CqXbu2tm7d6gJQdhsXF6fk5GQtWLBA//3vf11wqmrVqqEeLgAAlQ7L9wAAABCVpk+frho1amjt2rVq2bKlC0idffbZLlPq6KOPdplRkyZNUlZWVqiHCgBApUSmFAAAAKLOunXrtHPnTp1xxhkuM6qoqMgFqJ577jmlp6frz3/+s6pXr+4yqWxJHwAA8B6ZUgAAAIg6q1evdrcWeLJMKMuSatasmQtImc2bNyspKUl16tQJ8UgBAKi8yJQCAABA1LGsqEGDBqlp06YuMGXd9goKCsqenz17tgtWHXrooSEdJwAAlRlBKQAAKrmSkhJXW4fOY4imY7F9+/Z6//33y+63atWq7GsLRo0ePdp9PXjwYIX73xF/owCAaMXyPQAAKqn8/Hx3Gxsb6y52S0tLy56zr+1CGIjkY9E66vm76u25T1vKZ4XP//73v+uoo44Ky7EHe98AEOnsPXDP90VEJoJSAABUQp999pnOPPNM9ezZUw888IAyMzP3ysKwr+1CmJM9ROKxaK997bXX9Oqrr7oC5/4aUmPGjNH48eNdoMp+1t13362GDRuG1di92DcARDIC9tGFoBQAAJXM2LFjdcUVV7iL83nz5unBBx/UI488UnaS99ZbbyktLU3PP/88y4UQkcfi8uXLdeGFF+quu+5SfHy820+vXr10/fXX6/zzz9dFF13kakzZc+E29mDvGwAiGQH76ENQCgCASsYuZK0zmRWCbtCggXJzc91F7xdffOGeX7x4sbKzs5WcnBzqoSLKBetY/OGHH9ztDTfcoCVLlui6667T+vXrtWnTJncB8/nnn7ug1a5du8Ju7MHeNwBEKgL20YmgFAAAlYxdsNvyJbudO3eunnnmGbfEyWYcrTuZZZmYPn36hHqoiHLBOhZtX6Zbt26aMmWKtm3bpssuu0zvvvuu7r33XjVv3lwrV67cqxB6uIw92PsGgEhFwD46EZQCAKAS2bBhg3bu3KkjjjjCFXq2GcXhw4frrLPOchfATz31lDupM23btg31cBHFgnks5uXluVvLjPLv4+qrr9bpp5/uipufd955LmOqvMs7gjl2/kYBYP8I2EcnglIAAFQiGzdudLf+ZUt2azOKf/vb31S9enXdcsstLiXeLoaBSD0WTzrpJLd04+abb9aaNWtUpUoVjRo1SpMnT9b8+fM1e/Zs97rWrVuH3dj5GwWA3yJgH70ISgEAUImkpqZq4MCBrsCz1dhJSEhwj1vB0BtvvNFljliWSY8ePUI9VES5YB6Lhx12mM444wx3ETN69GgVFha6WiO23xNPPFGffvqpOnfurI4dO4bd2PkbBYDfImAfvQhKAQBQiVgtnccee0x/+ctfXPbInid4w4YN04ABA8pq8YQrWj1Hh2Aei7Y/mzW3GXTrwmTHTE5Ojn788UdXj8Rm0S1zqrx1R4I59mj4GwWAQCNgH71iSitZr0SrH5Cenq6MjAyX8gcAAH5lF+3WmczaLZd3aZOXiouL3a2dztiJqqX2WwHrJk2aqFatWqEeHsLgWFy7dq2+/fZbt4zPCp7XrFlT55xzjpo2bapI/DuKtL9RAAgEC84vXbrUBaQs8GTv5faYBaeWLVumSy65xDW2sGYWt99+e6iHCx147IWgFAAAlYR95L/++usua8QyLuxEzgpBf/PNN24Gsn379i7t3Z637JJwNHXqVCUmJrplVykpKb95/sUXX9SIESNcpsn1118fkjEiPI5F+xm2BfpYDubYo+FvFABCgYB9+CEo9TsISgEAKiubSWzTpo2aNWumFStWuBo7f/3rX122kXWt6du3r/71r3+pa9euCle27MpmSq1+ROPGjdWhQwc3XqsPdOSRR+ruu+/WE088oTfeeEPnnntuqIeLMDsWLVsqLi7OzbD7l8aF09ij4W8UAAKNgH10x17iPR0VAAA4KNnZ2e5D3TrKWIZQgwYNXDDGvj5Y1p3G3HDDDVqyZImuu+46t5zJ2EncZ5995tLi7XX+Wg3hMvY9gwq2TM/2bQVNbXvnnXfKnrdW0aZXr14VHj+Cx+tj0S5orBvfyJEjXbHcatWq6cEHH1SNGjXCauxe/7sAQCRYvny5LrzwQhewt1sC9tGFoBQAAGHIZvusQ9idd97pAi8WzLHZJjshO+aYY3TBBRe4bKGDmRWcO3duWYFkq7tgF7uXXXaZhgwZokWLFumFF15w9Zjef/99nX322WE1dmOd1Ozk02ZEbT+rVq3Sli1btH37dhf8skBVbm6uey3toMObV8einwWk7Nh48sknVVRU5B575plnwm7sXv+7AEAkIGAf3QhKAQAQhsaMGeNOvjZv3uzu21IeM2fOHH344Yd6/vnn3Uxhp06dyrJA/hfrSmMs5d2yl8zVV19d1sXLnn/ggQfc/sJt7Gbr1q1KSkpyy5uss5qdkNpjFpyy7Bc7GbWOatSSCH/BPhYt8GTHXZ06dVwA1J9FZ63DP/nkE3f/QI87L8fu1d8oAEQSAvbRjQWXAACEoWeffdYFdYYPH64333xTr776qv7+97/r+OOPV8OGDV0Gkr3Gsj8O9OL6pJNOcq+9+eab3TI4q6ljQZzJkydr/vz5mj17tntdRYM6wRi7sQCDFTH3p+dbd7127dpp8ODB7medccYZ7nH/BTzCVzCPxfHjx+vSSy/VCSecoPr166tu3bru2LNg6H333eeK4doWjmP36m8UACLJ7wXsTz/9dHd+cd5557mMaQL2Eaq0ksnIyLAj1d0CABCu0tLSShs1alS6a9euvR7fsWNH6WuvvVb2/IYNGw54n3l5eaVnnXVWaWxsbGlMTExpXFxcaWpqamnPnj1LmzVr5h7r0qVLaW5ubtiN/ffYz8jPz3dfv/nmm+53eOihhyq8XwRXsI7FMWPGlLZo0cJ9v22JiYllX9txN2LEiAofd8H8O/LqbxQAIsmECRPc+6KdO5xzzjmlycnJ7v38m2++KZ03b17pCSec4N4fZ86cGeqhohyxF7rvAQAQZqxOki1tM1OnTnXL1fZlj61bt66shtKBsgwmq8Xw7rvvuppOe7I6TJZNMmDAgLAc+4GwfdqpTUpKSsD3jcAKxrHYv39/TZs2TaeeeqorrG/L+Gxm3braWaaR+c9//qObbrop7Mbuxb4BIBLl5+e7jOj33nvPfcZbHank5GT3nmhL+VavXu268E6fPt09Hs6/h43fMmEtK9YyedPT01XZYy8EpQAACDMWWLn++uv14osvuuU7AwcOVM+ePV0wx5a/LVy4UNdcc417bObMmeX6GWvXrtW3337rTozshK5mzZo655xz1LRp07Acu52uvPHGG+723HPPdR349mwHbYEwK6SOyBOoY9ECUHYx0qRJE9epyc8K4FutEVvWZ8v3rG24LYcLxIVAsP6Ogr1vAIg0kRywt3OXiRMnupqAM2bMcOdD9erVc+UI/vSnP7kyBOWtcxjOCEr9DoJSAIBIMGnSJF1yySWuiLd9VFuBbzthsVk2fz2lW2+91dWeORi2L/8sYySN3TJdLLBlgScrYE076MgX6GPRjgvrzGjHic2mW9fHfTPmqlWr5gqeW0CzIoL5d+TF3ygARKpIDNh/8MEHrgaWdRHenxtuuEEPPfSQm3CLJgSlfgdBKQBAuLNAS1xcnCvG/PDDD+vrr792M4QW2LGTL7tYtYCOLUEKxAmMndjZz7N9W3ZTOI797bff1rBhw/TYY4+5otW2TGvPdtA2a2oBKdpBR7aKHIt2nNkxYsfc0Ucf7QrgWiF8K5BvRXJtWZ8de3bsWPemcBp7KPcNAJEgkgP29plkmd1XXHGFy5Cy8yPL6rXlez/++KPL6H355ZfdEsVoQlDqdxCUAgBEEvuYtloJNjO4fv16d2F61FFHuYvTQOzb3+lr48aNLovkwQcfVI0aNcJu7P/4xz/c2CzgYFlT1oVv33bQ27dv1//93//RDjoCVfRY9H//Rx995E7q7XzPz5ZJ2LmfLR21gKVdDNgSkHAZe6j2DQCRLlIC9pYhbu/bdr5iE3K9e/d2wTUrdfDQQw+58gQnnniiu42mGAVBqd9BUAoAEM72rJ1k6eh2EW2fXVa80z63rB6OZX4EitWAql69uqvH4890Km9dg2CO3TKrRo4c6ZbtzZo1yxWrttnFbt26uefvuOMOV6vBnh86dGi5fgZCKxDHYmFhoas3Yif6CxYs2GuZnl202HF01VVXuZ8TbmMPxb4BIJj8gfVg7TeYAftAjd0awHTs2NHVv7Sg1K5du9xSQwtOWYbvzp07XfZU7dq1XROYyhh7ia5FiwAARDgr0GyZHlY76YILLnBBFjvpsgtSq53Ur18/N6tWntpJtg+ru2OBIf9FudXX+dvf/qZPPvnE3a/ICVgwx37SSSfp8ccfd/uzQqY2Izpq1Cidf/75bgZy9uzZ7nWtW7cu9/jhnWAdi4mJiW4J32mnnaZ58+Zp8eLFbume1TVr1aqVW7oXrmMP9r4BwAsWaLEMJluaZkvU/BMGOTk57n0tEO9htg8L2D/55JNlAftnnnkmLMdumVyHHXaYPv74Y7cPe3+3QJQVNzc7d+50n13R3IXvfyqtZDIyMiwzzN0CABBu3nrrrdKYmJjSkSNHli5evLi0du3a7r5tcXFx7rZbt26lhYWFB7XfcePGlQ4fPry0devWpUlJSaV16tQpHTJkSOn//d//hf3YTV5eXulZZ51VGhsbW7a/1NTU0p49e5Y2a9bMPdalS5fS3NzcgP0+CA4vjsVIHHsk/7sAgFm6dGnp6aef7j6T+/btWzpjxozSDz/8sPTSSy9125tvvlmu/e7atat0yZIlpTt27Njr8dtvv720e/fubgvXsZvXXnutNCUlpbRKlSpu//Y+P378ePfcq6++6h6zn1FZYy8EpQAACCN///vf3cnJxIkTS1944QX39YgRI0rfe++90gceeKC0RYsWpenp6aWjR48+4H2OGTPGfZ8/QJSYmFj2dVpamtv/hg0bwnLse9q0aVPp0KFDywJce27t2rUrnTx5coV/BwSXV8dipI09kv9dAMDvzjvvdJNHFli3z+ojjjiitFGjRnt9Xt91111hGbAPxthNSUmJu33//fdL//znP5fee++9pdOmTXOP7dy50/1u1apVK/3ggw9KK2vsheV7AACEEVtqZKwWji09MtZG2F87yZ632kkHUxLSajGtXLlSp556qho0aOBS3W3/VjB8/vz5rkh4+/btXb2dcBv7niyd3rrwRWI7aHh7LEba2CP53wUA/L788kv3Gf/UU0+5zTqd2jlAr169tGTJEndu8Pnnn+uiiy5S8+bN/+f+3nnnHbd82d4fjdWq3Lp1q9vHd9995zru3nvvvapfv37Yjd3Pv+TP3t9ts5pS/i7BtmTvlVde0XPPPeeW+VVapZUMmVIAgHA2YcIEN1Nns3PnnHNOaXJyssuS+Oabb0rnzZtXesIJJ7jZupkzZx5wynt8fLzLwthTVlZW6dy5c0v//e9/u2VwXbt2dTN24TT2/c02FhcXV2iMCB0vj8VIGnsk/7sAwJ5q1qxZ2qZNG/f14MGD3Wf+1KlT3f3Nmze75fa2hG3VqlUHtL/DDjvM7eO0004rveqqq1ym0SmnnFLauXPnsuylRx55JCzHbuycZdSoUaVz5szZ63HL7P7+++9dBng0Y/ne7yAoBQAIZ4GunbR8+XK3FKhTp06lixYtKs3Ozv7Na2z/devWDbuxH4jVq1eXrlu3zv1shDcvj8VIGnsk/7sAgJ8FWCywU79+/dKXX37Zva/Ze9fKlSvd80VFRe59ziaswi1gH+ix+/3000/u3MeW6BlbtmcBtY4dO5b26NHDBdnWrl1bGq1YvgcAQASyrnKWNh4bG+ta25eUlLiuLz/++KN7vm3btq7rXHJy8gHtLyUlxXWr+/rrr3XVVVfp9NNPV7t27Vz3F1tON23aNLf/7t27h93YD6RVsy17ClY7aARW0I7FzNVS3tbgDDq5tpTWNKh/R17+jQJAsNhnv3XZHT9+vC6++OKyJWrXXnutLrzwQrcUecGCBe41B8KW6Ns+rQOpLZ1r3Lixe79MTU1V586d3Xbfffe5c4CKdq4L9Nj9fv75Z3d7yCGHuN/nyiuvdPvymzNnjhYuXKjJkyeXnddURgSlAAAIM4GqnWQnOLav6667TjNnztTEiRPdZho2bOhO4uxkyE6+zjrrrLAae6jaQSM4gnYsWkDqpXZScX5wBh5XRaUXL1Ldus2C8ncUir9RAAiG2rVr6/zzz3cTURs2bNC5557rguuPPPKIPvnkE/caCygdf/zxYRewD/TY/azmlenTp4+++eYb/fTTT+rUqZNOPvlkLV261O17+fLlrp7VoEGDVFnFWLqUKpHMzEz3AZ+RkaG0tLRQDwcAgN/YvbzezdwFQmFhoctcevHFF91MnxVQ9rNgkRVPthO+6tWrh93Y/SzwtGLFCneSuOc477jjjrITRn9GFsJXwI/FTbOl13spqM6fJdXrGdS/Iy//RgEgGCw72j77161b5yaNLLC+Y8cO3XLLLS6AZJNUl1xyictqrlWr1h/uy5819NFHH2n48OHuGt5v34D9ww8/7AL74TL2PR155JGuYLplellB9jfeeEPvvfeeTjvtNPf84MGD9dVXX+n999/XKaecosoaeyEoBQBABLATIuvMYheotkyuPGxmcd68ee7kyL5OSkpSq1at1L9/f4Xz2C2dfvTo0a7Tju3LPr+tG47Nmo4YMSIoY0ZwBexY9DAo5cXfUaj+RgEgWLKysjR16lQ1atRIbdq0OajzgFAH7CsydmPnK9nZ2S5T3M5hLOBlv4dlZplhw4a57HLLlO3Z89fPmWhBUOp3EJQCAEQS/2zhzTffHHG1kwIx9v21g7Z2ysb2N3To0IC1g0YECkFQCgDw20yjZ5991gXQu3XrVnYOYFlC9lltAR1b/haOAftgjd2CUVb3ysZrZQxsnxbUsjhEfLyvipI9v2bNGq1fv94tV4w2BKV+B0EpAECksZk1mwX0104qLi6OmGKYFR27nSRa6vypp56qBg0auP3YTOmyZcvKioX+5z//cbOlqIQISgFAyFlBb6vtdMEFF+iVV17R9OnTddlll7mAj00mWaHvu+66yxUrryxjtywvq61p5yw2sfbLL7+4c5jHH3/cxSFsKd9FF13ksr4tG6wyx14odA4AQBjZX+0k6zxj2UL+2knhGpAK9NhtfzNmzFDz5s01duzYvWYf7QTPlvVZ5x07ibz00ksr3H0HAAAcvH27zF1xxRV7dZmz5y3LKRy7zAVr7ImJiTrmmGN+sxzQzovs1jrvWWbWRRddpMqOoBQAAGHij2onWfDFtso0di/bQQMAgPKJ5C5zwRy7ZVsZC2TZZssBjU3cPfroo0H4bSITQSkAAMLA/monbd26VZ9//rkL9NhJU7jWTgrW2L1sBw0AAMrHsn6M1Yy0z35jn/v7dpmzJf2Vaez7diL2Z1pZlz/72oq1V61aVZUdQSkAAMLAyJEjXVDn92onvfDCC2rfvv2B107KXC3lbQ3OYJNrS2lNgzf23SdudevWdW2erSvNxIkT3ba/dtBnnXVWUH5NIJh2rd2kkm07g7Lv2FrVldC4XlD2DQD7+vHHH92t1V6yySjrLmeTSn7+bnNNmjRRuAnF2B977DFt2LAhoprXBBOFzgEACDEL4iQnJ7sTHksR/73aSS1btnQ1Df7nUjULSL3UTirOD86A46pIlyx2gamAjz3M2kGj4jZn5CkjtzAo+07aPkdNPztKwbK6RNp6/OtSrQ4B3W/Mhu1KO/mfUoGvk2SgxSQlqsn3bxKYAhB0kdxlLhRjj+TmNQeLQucAAESIgNdOsgypYAWkjO3bfkZa06DXfbJCocOGDXNp9MFqB43gBqQuGTVJu4p9dTUCrXXMUo1KDl5Aql2ulP/e+QHfd6fNKfqwoIeCpbSg0JeFRVAKQJDZ5/RLL730my5zVszb32XOJpVsCf7/DOoEM8t7P5neAR17lDWv8RJBKQAAQiySayd5NXbLxrLOOLYhcliGVLACUsG2tVQKYmgXAKJCwLrMBTvLe59M74COPcqa13iNoBQAACEUybWTInnsAAAgMALSZS7YWd77ZHoHs0NeJDevCQWCUgAAhJA/bXvIkCF6+umn96qdZPULbPPXTho+fLjCSSSPHQAABEYkd5kLxtiD0QAmmhGUAgAgDERy7aRIHjsAAAiOSO4yV96xWwBqxowZat68ucaOHfu7DWBeeeUVXXrppeWqtxltCEoBABBGIrl2UiSPHQAABIZlGlmXuSeeeKKsy9wzzzyjyjD2YDeAiUYEpQBgD7aufNeuXS7zY89uGPahtHr1atf23taFAwAAAIjsLnOBHnskN6+ptEGpUaNG6eGHH3aRQqts/+STT6pPnz6/+/qdO3fqH//4h0uF2759u5o1a+bWbJ5wwgmejhtAdLIPENt69OjhCjXbOnJbfjRhwgRdeOGFuvrqq3XXXXeFephAVNu1dpOvnX0QxNaqroTG9YKybwAAKptI7jIX6LHTACYCg1J2AFhxr2effVaHHnqoCy4dd9xxLs3N/jP3VVhYqEGDBrnn3n33XTVq1EirVq0qi2gCQEU999xzeu+999zXtobcMqM6duyotWvXuq4ZS5cudZt9qNiMB4DAB6TW9B2m0oLCoOw/JilRTb5/k8AUAAAVFMld5oIxdhrARGBQylosjhgxQhdffLG7b8GpcePG6aWXXtKtt976m9fb45YdZQeJf/mMFRADgECxIs2mSpUqysrKch8kthl737Fg+htvvKEvvvhCAwcODPFogehjGVLBCkgZ27fLwiIoBSBKUHoAoRLJXeaCOXYawERIUMqynmbNmqXbbrut7DErCGYXebbOcn8++ugj9evXzy2f+fDDD12Wgv1nW4QzLi5uv99TUFDgNr/MzMwg/DYAosVRRx3lMqEsK/Oyyy5zASn7wHrttddc0UP/SZ19eAEVtTkjTxm5gQ/AJG3PUlMFz+oSaeuWhVJJ4Pcdu3WNqgV+twAQtSg9gFCI5C5zXo2dBjBhHpSy1Lji4mLVq7f3TKXdX7Ro0X6/Z/ny5e4N97zzznMHil04WvEwmxn4vTdaa994zz33BOV3qMxsvWy4FqsDKsLq2lntujFjxriihA899JDL0Hz++efdrMeXX37pPqw6deoU6qEiCgJSl4yapF3FgY/stI5ZqlHJClpAql2ulP/e+UHZf6fNKfpQPYKybwCIRpQeQChEcpe5SB57NAp5ofODTU21elL/93//5zKjrAjZunXrXKH03wtKWSbWnil3lillb9QoXyDKOgXYH6sFpAhMIdrYMR0fH++W59mH1H/+8x9Xu+6YY45x7z8dOnRw3TT69u0b6qEiCliGVDACUsG2tVTKD/UgAABlKD2AUIjkLnORPPZoFLKgVO3atV1gydZt7snu/14xMVsuY2+sey7Vs4tEi2DackDLYtiXrdu0DRVjf5xvv/22nnjiCVdY3i7WLSj4zTffuAt5yxqxP2K7cLeoMxCJLMjqP4atxp3d//vf/1627rtLly7ulmMcAACEC0oPwGuR3GXOjb1eZI49WoUsKGUBJAtq2FpnKy5m7GLQ7l9zzTX7/R67MHzzzTf3CnwsWbLEvcHuLyCFwHn11VfdxfmOHTvcfesmYG9EtrTJ/rBt7bp98B1//PFkUCFi2XuLn30QWRDWAt7+lHh7zzJ2jAMAAIQDSg/Aa5HcZS6Sxx6tQrp8z/6jrfhe79691adPH1cB39Lk/N347CCwpTNWF8pceeWVeuqpp3T99dfr2muv1S+//KJ//vOfLsqJ4LKLcgtIWWaa/bvbh579X9WqVcsVirMPvrfeekuHHXYY624RsfbNgLIlfI8//rjrnGFZgccdd5x7/PcaKwAAAHiJ0gMIpUjuMhfJY482IQ1KDR06VFu2bNGdd97pluDZms3PPvusrPi5tTDd8yLRakF9/vnnuvHGG9W1a1f3hmsBKuu+h+Cy1EZbnmftMW2WxVIaR4wYoWOPPdbNvlhngm+//dalCBOUQiSyEzdbsnf44Ye79xf/YytWrHDFDk8++eSyYBSZgAAAIBxQegDhIJK7zAVq7LvWblLJtp0Khtha1ZXQeO8GcdEk5IXObane7y3XmzRp0m8e69evn77//nsPRgY/S2G0lGArAGcfehZ4ss4etrTJIslnn322qzdlmVSsVUekshkSey+64IILXJB1+vTpri6DZQLaUj77oLr33ntdMBwAACAcRFvpAft9LLDGBCAiiQWk1vQdptKCwqDsPyYpUU2+fzNqA1MhD0oh/FnLTH+Xgvz8fNfZ49BDDy0rIG8ZbZZB8nsF6oFI8PPPP7tbCz7ZMX/FFVe4zMA9g1a2dHXy5MnUTQMAAGEhWkoPZGdnuw7f/t+Hcy1EEsuQClZAyti+XRZWlAalyOPE/2QFzc844wy1bt3afdBZO1lLD/abMWOGMjMzXT0pIFL98MMP7tbq29lJ3E8//eSWqt52222u84al9i5fvtwtV+UkCQAAhENWkRVq9k+s7Vt64JFHHlH79u3d4+F67mIZ6VaX1s6/rDv75ZdfrrVr15aNt6CgwF1nAIheZErhf2rRooXeeeedvWp7+VmBcyt6bqwNLRCp5syZ425r1KjhatcZW65nM41m8ODB+uqrr9zyVQAAgFCLhtIDFpCy+sBWX9i88MILLqvrX//6l9LS0vT+++/rkksu0cMPP6yrr7461MMFEAQEpXDQ/MUUjXUoOProo91mRc+BSPXjjz+627vuukvfffedm62zbjV+dn/foCwAAECoaieFqvRAIOs+vfrqqy4g1bRpUzVr1kxTpkxxKzKsO7sFoxYtWuTKh9BIKXxszshTRm7gl6olbc9SUwXX6hJp65aF0q+l2AIidusaVQvsLisVglI4oI5k1sGjW7du7jELSFmnPZuBsZTgK6+8cq9AFRBprI5BzZo13RI9y/yzEzernVa9evW9TvyswH+7du1COlYAABC5Alk76fdKD1jH4KVLl+qTTz4pKz0QiBUNwaj7NHv2bJcRZZ2+7TzMstXPPPNMPfDAA658iC1F9P+OCI+A1CWjJmlXcYCjOpJaxyzVqGQFNSDVLlfKf+/8gO+70+YUfageAd9vZUEUAQeUFvzoo4+6+5YWbK1lbb23BaP++te/upkZAlKIZImJiXrppZc0cuRI3XfffbrwwgvdiVBWVpaKi4vdLN6CBQvcCZ0V/A9XkdJZBwCAyiYYtZN+r/SABXRGjx7tJpWti3ZFSw8Eq+6TZUjl5OS47uq2XzvHsm7ftn8LRtnvYplSpm3bthX6HRAYliEVjICUF7aWSvmhHgT2i0wpVCgt2D4M7cOCjmSI9KDUMcccs9djFpCywv52a8e5ZQpedNFFCmd7/v1Z9qL9TdoWHx+vTZs2adq0aerQoQPZXgAAREHtJK9KDwSr7tPmzZvdeYplYJldu3a5lRg26W1ZXo899pj7OS1btqzQ+AGEN9Jb8IfoSIbKwoI4/kCOsaV6diJkS/gsU9BO/E466SSFK+uKaYXYt2zZ4u5b9qKN3wJSZty4cW728aOPPgrxSAEAqHz2rJ1kgSM737ASGe+++657/mBrJ/lLD9SpU8eVHrDJY8s6CkbpgUCP3c9ebw1l7HrClhtaQMr2bUGof/zjH+41lrHeowfLooBoRqYU/hAdyVBZ7LsE1Z/5t27dOve1nfhZ5lS4siWHlg1l7ATVgsfdu3d3mzUi8M+mUqgdAADvBbp2kr/0gH32r1y50hU0t2V2luFtP+eNN95wpQdsQqqipQeCVfepcePGeuqpp9yEWr169dxjNkFok2o27s8++8xNpnXt2rVC4wcQ3ghK4Q/RkQyVnaWOb9iwwc00Pvjggy5AG45sxtSKs9vJnJ3cTZo0yW1+/oypnj17hnCUAABUPv7aSTZJ5D939tdOeu6558pVO8mr0gPBGLufBZ9sn/79+h8zdevW1ahRo3T++eerb9++FfodAIQ3glL4XXQkQzgLdg0z27dlAD7xxBNu5tE888wzYTluO2G0sVp2lJ182pJae8w2C1Bt3bpVO3bscK+lUCgAAN4KVu0km4gydl5hm52TG3/pgXAfu51XWTF2y+r283f4bt68uRo1auQysgBEN4JSCIu0YCDURb3t2Lb0c1v65g+82myjFfa0k659f2a4jNvs3LnTjdtqPdx0003KyMhwf6cWlLLnrDac1WZo06ZNhccPAABUrtpJ27dvd7WTWrduvVftpKuuuqpctZO8KD0QrLFbh+9rr71WF1xwgV555RXX4fuyyy5z50ZJSUmuydIdd9zhlvgBiG4EpRD1HckQfayot52I2fFnwZh9T8qsqLed2Pz73/92M3n/y/jx413rZFuiaoVCLejaq1cvF3C977773BaO4/azfV1//fUuk9F/Ammb/0TOv3Svc+fOAfk9AABA+NZOCmTpgWCN/UA6fC9cuJAO30AlQFAKIU8LBkJZ1Pudd95x2VCWDWgsZdyWu1kRTwtSWZaR1UuoX79+WI17T7Vq1XIdMff927XsLwsuW8q9pcFbYwIAAOAtr2onBaP0QLDG/nsdvk8++WSXkWVZ6v4O34MGDarQ7wAgvBGUQlR3JEN0CmRR75EjR7qA1KmnnqoGDRq4kzgLHC1btszN2L3wwgtq3769WxYXTuP+I/a3aX+7FpAyJ510kgYOHFjB0QMAgIMVzNpJwS49EMyx0+EbgB9BKUR1RzJEn0AW9baTuRkzZriTqrFjx+4VPLJAlS3rs6V7Vuvg0ksvdUviwmHc+54wPvvsszr88MPLUuftBNR/wmg1H+xk1b+0DwAAeCdYtZO8KD0QzLpPdPgG4EdQCmHRkQwIRVFvO4mzjCKbVVy8eLE7qbKi/ampqa7+km12Umf7rkhAKtDj3veE8ZprrvnDE8Y777zTzWQCAABvBaN2klelB4JV94kO3wD2RFAKYdGRDAhFUW8LQNms3Ndff+26x9jsop382M/Iy8tznfBycnL2SlkPh3HviUKhAACEr2DUTvKq9ECw6j7R4RvAnghKIeQdyYBQFPW2AI0V6Lzuuus0c+ZMTZw40W2mYcOGLmBkwRybfTzrrLPCZtz7olAoAADhK9C1k7wsPRCsuk90+AawJ4JSCHlHMiAURb39GUNDhgzR008/rRdffNHNytms3fr1691mqeU2yzh8+PCwGfe+KBQKAED4CnTtJC9LDwSz7hMdvgH4EZRCyDuSAaEs6m1BoWHDhrkgjtVnshM8W7pn9ZhatWrlOs6E47j9KBQKAEB4CkbtJK9KDwS77hMdvgH4EZRCSNOCgXAp6m0nXfb9tkXKuCkUCgBA+Ap07SQvSw+Equ4THb6ByoegFEKaFgxUhqLewRo3hUIBAAhfga6d5GXpAa/rPtHhG6i8CEohpB3JgMpQ1DuY3WsoFAoAQPgKRu0kL0oPBGvsfnT4BuBHUAoh7UgGVIai3sEcN4VCAQAIX8GsnRTM0gPBHDsdvgHsiaAUwqojGRCNRb2DOW4KhQIAEHkiuXZSRcZOh28A+yIohZCmBQPRXtQ7VOOO5JNdAACiVSTXTgrE2OnwDWBfBKUQ0rRg4EAFtKh35mopb2twBppcW0prGpxxR/nJLgAA0SaSaycFeux0+AawPwSlAESEgBX1toDUS+2k4vzgDDSuinTJ4rLAVLCLkUfyyS4AANEskmsnBWPsdPgGsD8EpQBEjIAU9bYMqWAFpIzt237GHtlSwSpGHsknuwAARLNIrp0UrLHT4RvA/hCUAhAxIrWodzDGHcknuwAARLuA107ysPRAMOo+0eEbwO8hKAUg4kVqUe+KjJtCoQAAhKeA107ysPRAsOo+0eEbwO8hKAUgYkVqUe+KjptCoQAAhK+A107ysPRAsOs+0eEbwL4ISsGbtOD9pAYDlaWod6DHTaFQAADCVyTXTvJq7HT4BuBHUArepAXvpysZUBmKegdj3JF8sgsAQDSL5NpJkTx2AJGLoBS8SQv+na5kQDQX9Q7GuDlhBAAgfEVy7aRIHjuAShqUKiwsdEtSbP1vfDzxLQDBEalFvYMxbk4YAQAIf5FcOymSxw4g8pQrkmQFeq+99lpXRNcsWbJELVu2dI81atRIt956a6DHCaCSitSi3sEeNyeMAACEv0iunRTJYwcQOWLL80233XabfvrpJ02aNElVqlQpe3zgwIGudgoABMq+Rb2tVpLxF/W+5ZZb3P1wK+rt1bj9J4znn3++RowY4bKjCEgBAAAAiNpMqQ8++MAFn/r27btXx6hOnTq5ZSkAECiRWtQ7UscNAAAQSLvWblLJtp1B2XdsrepKaFwvKPsGEMZBqS1btrhCu/uyC6xwbMcOIDJFalFvN+56kTduAACAQAek1vQdptKCwqDsPyYpUU2+f5PAFFDZlu/17t1b48aNK7vvD0RZ0d5+/foFbnQAKrV9i3offfTRqlfPd9JhBb0tsGNFve+6666wKuodqeMGAAAIJMuQClZAyti+g5WFBSCMM6X++c9/6vjjj3cdn6yY7+OPP+6+thbn33zzTeBHCaBSi9Si3pE6bgAAAAAI26DU4Ycf7gqdP/jgg+rSpYu++OIL9ezZ09VIsfsAEAyR2gUmUscNAAAO3OaMPGXkBj4rKGl7lpoqeFaXSFu3LJRKAr/v2K1rVC3wuwVQmYNSu3bt0uWXX6477rhDzz//fHBGBQDwtFCooVgoAADlD0hdMmqSdhUHPrLTOmapRiUraAGpdrlS/nvnB2X/nTan6EP1CMq+AVTSoJQV5n3vvfdcUAoAEB2FQg3FQgEAKB/LkApGQCrYtpZK+aEeBIBKrVyFzk899VR98MEHgR8NACAkhUINxUIBAAAAhH1NqTZt2ujee+/V1KlT1atXL6WkpOz1vLVBBwAAAAAAAAIalHrxxRdVvXp1zZo1y237tkInKAUAAAAAAICAB6VWrFhRnm8DgJB2r4nkDjZ0rwEAAAAQbcoVlNpTaWlpWYYUAIRz95pI7mBD9xoAAAAA0aZchc7Nq6++qi5duig5OdltXbt21WuvvRbY0QGodCK1e42hgw0AAAAABDlT6tFHH9Udd9yha665Rv3793ePffvtt7riiiu0detW3XjjjeXZLQAAAAAAACqJcgWlnnzyST3zzDMaPnx42WMnn3yyOnXqpLvvvpugFAAAAAAAAAK/fG/Dhg067LDDfvO4PWbPAQAAAAAAAAEPSrVu3Vpjxoz5zeOjR49WmzZtyrNLAAAAAAAAVCLlWr53zz33aOjQoZo8eXJZTampU6dqwoQJ+w1WAQAAAAAAABXOlDrjjDM0ffp01a5dWx988IHb7OsZM2botNNOK88uAQAAAAAAUImUK1PK9OrVS6+//npgRwMAAAAAAIBKoVyZUuPHj9fnn3/+m8ftsU8//TQQ4wIAAAAAAEAUK1dQ6tZbb1VxcfFvHi8tLXXPAQAAAAAAAAEPSv3yyy/q2LHjbx5v3769li5dWp5dAgAAAAAAoBIpV1AqPT1dy5cv/83jFpBKSUkJxLgAAAAAAAAQxcoVlDrllFN0ww03aNmyZXsFpG6++WadfPLJgRwfAAAAAAAAolC5glIPPfSQy4iy5XotWrRwm31dq1Yt/ec//wn8KAEAAAAAABBV4su7fO+7777Tl19+qZ9++knJycnq1q2bBgwYEPgRAgAAAAAAoHJnSk2bNk2ffPKJ+zomJkaDBw9W3bp1XXbUGWecoT//+c8qKCgI1lgBAAAAAABQGYNS9957r+bPn192f+7cuRoxYoQGDRqkW2+9VR9//LEefPDBYIwTAAAAAAAAlTUoNWfOHB177LFl999++2316dNHzz//vG666SY98cQTGjNmTDDGCQAAAAAAgMoalNqxY4fq1atXdv+bb77R8ccfX3b/kEMO0Zo1awI7QgAAAAAAAFTuoJQFpFasWOG+Liws1OzZs9W3b9+y57OyspSQkBD4UQIAAAAAAKDyBqVOOOEEVztqypQpuu2221S1atW9Ou79/PPPatWqVTDGCQAAAAAAgCgSfzAvvu+++3T66afryCOPVGpqql555RUlJiaWPf/SSy+5jnwAAAAAAABAwIJStWvX1uTJk5WRkeGCUnFxcXs9/84777jHAQAAAAAAgIAFpfzS09P3+3jNmjXLszsAAAAAAABUMgdVUwoAAAAAAAAIBIJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCgAAAAAAAJ4jKAUAAAAAAADPEZQCAAAAAACA5whKAQAAAAAAwHMEpQAAAAAAAOA5glIAAAAAAADwHEEpAAAAAAAAeI6gFAAAAAAAADxHUAoAAAAAAACeIygFAAAAAAAAzxGUAgAAAAAAgOcISgEAAAAAAMBzBKUAAAAAAADgOYJSAAAAAAAA8BxBKQAAAAAAAHiOoBQAAAAAAAA8R1AKAAAAAAAAniMoBQAAAAAAAM8RlAIAAAAAAIDnCEoBAAAAAADAcwSlAAAAAAAA4DmCUgAAAAAAAKicQalRo0apefPmqlKlig499FDNmDHjgL7v7bffVkxMjE499dSgjxEAAAAAAABRFJQaPXq0brrpJt11112aPXu2unXrpuOOO06bN2/+w+9buXKl/vKXv2jAgAGejRUAAAAAAABREpR69NFHNWLECF188cXq2LGjnn32WVWtWlUvvfTS735PcXGxzjvvPN1zzz1q2bKlp+MFAAAAAABAhAelCgsLNWvWLA0cOPDXAcXGuvvTpk373e+79957VbduXV166aUejRQAAAAAAACBFK8Q2rp1q8t6qlev3l6P2/1Fixbt93u+/fZbvfjii5ozZ84B/YyCggK3+WVmZlZw1AAAAAAAAIj45XsHIysrSxdccIGef/551a5d+4C+58EHH1R6enrZ1qRJk6CPEwAAAAAAAGGcKWWBpbi4OG3atGmvx+1+/fr1f/P6ZcuWuQLnf/rTn8oeKykpcbfx8fFavHixWrVqtdf33Hbbba6Q+p6ZUgSmAAAAAAAAKnFQKjExUb169dKECRN06qmnlgWZ7P4111zzm9e3b99ec+fO3eux22+/3WVQPf744/sNNiUlJbkNAAAAAAAA4SOkQSljWUwXXnihevfurT59+mjkyJHKyclx3fjM8OHD1ahRI7cMr0qVKurcufNe31+9enV3u+/jAAAAAAAACF8hD0oNHTpUW7Zs0Z133qmNGzeqe/fu+uyzz8qKn69evdp15AMAAAAAAED0CHlQythSvf0t1zOTJk36w+99+eWXgzQqAAAAAAAABAspSAAAAAAAAPAcQSkAAAAAAAB4jqAUAAAAAAAAPEdQCpVOSUmJSktLQz0MAAAAAAAqNYJSqDTy8/PdrXVzjImJ2SswZV9bsAoAAAAAAHiDoBQqhc8++0xnnnmmevbsqQceeECZmZkuMOVnX1uwigwqAAAAAAC8QVAKUW/s2LG64oorNH78eM2bN08PPvigHnnkkbLMqbfeektpaWl6/vnn9wpUAQAAAACA4CEohahnwabVq1erRo0aatCggXJzc11g6osvvnDPL168WNnZ2UpOTg71UAEAAAAAqDQISiHq/fDDD6pataq7nTt3rp555hkVFRW5ZXwFBQVavny5e12fPn1CPVQAAAAAACoNglKIahs2bNDOnTt1xBFHqGXLlm6Z3vDhw3XWWWe5INVTTz3lMqVM27ZtQz1cAAAAAAAqDYJSiGobN250t7t27Sq7tWV6f/vb31S9enXdcsstrs6UBawAAAAAAIB3CEohqqWmpmrQoEGKj4/X+vXrlZCQ4B63Lnw33nij67aXl5enHj16hHqoAAAAAABUKvGhHgAQTC1atNCjjz6qdevWqUqVKu6xkpISxcbG6rzzztOECRM0efJkdevWLdRDBQAAAACgUiEohahmGVIdOnRw254sQ8qW7FnAyrrwnXnmmSEbIwAAAAAAlRFBKUQ1qyE1Y8YMVz+qU6dO7jHLkrIle5YxZcv4unfv7h4DAACVUczu29IQjwMAgMqHK3FErezsbD322GMaMGCAunTpojPOOEMFBQW6//77df311+vJJ5/Uli1bCEgBAFCpWTCKgBQAAKFAphSi1rRp0/TAAw8oLi5OMTExmjp1qs466yx98sknZa/5/PPPNXHixJCOEwAAhEjdHlLbM6Wt86UtP0kZy6Si/FCPCgCASoOgFKKWBZuysrI0fPhwV+jciprbZplTVlPKgla//PKLqyk1ePDgUA8XAAD8zuK6gOYxxcRKpSVS02OlIa9IqQ1+fW7rPGn9NGnFOGnJx4H8qQAAYD8ISkXpsrXMzEwtXrxYiYmJatCggRo3buy+rkxjnzlzpru1YuaPP/64C0hde+21uvzyy11NqQsuuEA//vijcnJygvCbAEA0oNYOQi9oR1+7s30BqRWfSUnpUoNDpdqdfVt6C4JSAAB4gKBUFLHC3Z9++qnuvPNOzZs3zwVz0tPT1axZMx1zzDEuCGMFv+114VZHKRhjX79+verXr6+qVatqxYoV7rGLL75YLVq0cF/bvi0oZbcAgP0hGIXQiI2JUUlpqc4+rJU2ZeRq6sKNKiopVWyMVFLRw9KypExiNWntZOnT86X8HVJ6S6leL6nN6dKC1xQRYmKspXCoRwEAYYQJtUhDUCqKjBkzRjfccIM2b97s7vsDMXPmzNGHH36o559/Xm+99ZbrQmfL16zOUrSOPT8/X7Vq1XJZV1WqVFH//v2VnJysdu3a7RW0Mm3btg3q7wYAEYlaOwghC0iZi49pp9yCIqUmJWjc7NUVD0j5WUCqIFNq0E9Ka+E7tjOW+7Yl7yhi7BuQsom7kt1BNwColAhGRRqCUlHk2WefdUEdq6E0ZMgQFRUVadGiRS4b6KeffnIZSPaaf//73y57KJrHHh8frzvuuMPtw/iX7PlNnz5dM2bMUO/evZWamhrU3w0AIsYB1NpJXzZOmsuyJgRffGyMRk9dpsHdGuuaEzrriI4N9MaUX/Tzqu0VP8Y7nC91HeF77JjHpZWfS1vnSpmrfJtlToW55CN6KalbO5XmFyjny2kqWrmegBSAyo0JtYhEUCqKWACnYcOGeuGFF1xQxm/nzp2u49zVV1+t999/X//4xz/CLigV6LHbPgYOHOg2P8uU8tetsuLmthzw9NNPD9JvBAAR7A9q7dSxrBKCUvCALdcbM3WZ1m7L1hl9W6pr81pKqZKg0VOX6tuFG91ceEx5l+7Fxks7l0sJVaWGh/m2/J1SUY405e/SwtcVzqpfM0zpV5ytuDo13P2C+ctU7fRBKvhpkXK/nsGSPgCVB80rIl54FRZCuW3ZskVJSUkqLCwsW/rmZ8GX888/X3Xr1tX27dtdnaXKMPbi4uL9Pp6SkqIrr7xSs2bN0hVXXFHh8QNAqAR8Efb+au28fbj0Ujvpk3OkxWO0edbjighhtEQd5ZdbWKSvfl6na1/4Vm99u1St6qfp72f01M0nd1PN1KTyL9L48Unp9V7S+ydKX17hjm3lbJBSG0m5vlIC4SquYV2lXzlUJZnZKpy/VKX5hcr/bo6qnXO8qg07UUpgzhlAJZ9Qs0CUndPYhJplxXa/JtSjwx/gUytKWKDllFNO0Ysvvqhu3bq5DKGePXuqTZs2LgNp4cKFWrZsmXussow9Li5uv4/bMkF7zvZtwTAAiFRByYX4H7V2Mvcf7w8/1NqJeNce39nFFtOrJqpF3TQlxP86l3p054ZqXreaRn+0Ucoux87jkqTEVCmphm8Gfe7zUkp9X6HzVV8qLO0+hqsee6hiqiQq+8UvVaVfN5Vk+UoVlOTlKzYtVSrcFeqRAoB3oql5RSVFUCpK2JK28847TxMmTNCqVavckrcvv/zSFQS3ot/GCn+fe+65qqxj9xdIf+ihh7RhwwZVq1ZN//rXv1Sjhi/1HQAqdVeyA621k0OtHQRfu4bVdUKvpu7rvMIiVUmI17asfM1etkUJ8XHKzCtU//b1NaBnF2nK7mP3f7LsuVJfNlS/O6UO50nFu6SCHdLPz0szHpSWj1PY2n0Mxzeq64KusbWrK6lrW+145GXFN2+o+Hq1lDf1x1CPEgC8Fy3NKyopglJRwgIuRx11lN5991395z//0ddff+0ygqy2Uu3atV0w5uabb9b111+vyjp2209ubq6efPJJV0jdWPF0AIgkQetKdqC1duZRawfBlxAXoxcnLHKBqB3ZBVq9NVtVEuOUkVOonIIit3SvfvWq6tS6uTT5IAOO/e+XOg2XCrOl3I2+INXh90u7sn3L+sJcwayFir0+WSlDDldstRSVFhap9n3XKrZGmnLGTQ718ADAO1HUvKIyIygVJSzgYmyJ2xtvvOEyjhYsWKD169e77nKDBg1y2UaVaewWeLIaVXXq1HG1qfxZWX/7299cNtaePxsAVNm7kvnZRfn8V6TqlvZ+iNT0GKl2F6lWh8iptZORpeLN25TQqqmrtVN35K1K7NRKuVNms7QpQhQUlWjphgzNWbltv88XFhVryfqdSizYokEHvNdSKSZOan2K74Jl8i2+5XvNBkl9bpVa/Uma/7JUmKVwlj/9Z+VNma3kI3u7+7X/faO7zX7nC+VP+ynEowMAD0VJ84rKjqBUlLAAzA8//OCWpHXu3FnNmzd3W2Ud+/jx4zV69Gh99913WrNmjdLS0tSrVy/Xbe++++5zGwBEa1eyvZYqHSxq7SDEkuJj9ddTumnDjlwtWLtDhUW+i46OjWtozdZsZeXvUk5+kd77foXqFCw5iKCUpPqHSIlp0qZZvq5MZt0UqenRviUf4RqQ8h/jg/q5jKht9z+nqlNmqergwxRXp6ay35+gjOffZZkqgLBXzrOTqJ1Q24slTFTCjG6CUlEgOztbTz/9tG699VZ3/7TTTtPrr7/uaietXr1arVq10ogRI1zGUGUY+zvvvOOyoVauXOnuJyQkaOvWrfr8889dkMoCYPfee2/YdSEEgPJ0JZs0b72GHdFG5x7e2nUlm/DzOn09cb1UfDAnNdTaQfhcqHRvUVtNaqfq8zlrygJSackJ7hifvXyL3p+x0r1u3fYcJcfkSckHOatekCG1PVvasUTa/KNUramvIO6G7ysw9hiXfV1yQLWtymH3RUqNv1yk2KrJWn/69dr55Jtuc+LjpKJI6UIAoDILSsglEifU9qe0cjZpISgVBaZNm6YHHnjAdZSzE6KpU6dq6NChZUvUjBUOnzhxYlmx72ge+8iRI11A6tRTT1WDBg1cJtamTZtcB7/58+frhRdeUPv27XXTTTcF+bcDgNB0JWtX9xTp88elbfMPbsfU2kEYTBC3aZDu7i/flFn2XN30ZLVukOaypEy5C/tvmSNtniU1PVYa8G+pYKdUo41UlCctervcYx/aeaiapjfVt6u/1aqdq7Q9b7vybJ+BYv8wMTHatXSN4urXVvGWfeqjEJACUOmatET4hNo+kitxkxaCUlHAAjZZWVkaPny41q1b57rY2TZgwAAXyLHAzy+//KIvvvhCgwcPVjSP3QJQM2bMcMv/xo4du1dGlgWqbFmfLd175ZVXdOmllyo93XfiCwDR1pVM3a6QJl5/gF3JqLWD0PNfmNjFS3FJqQ5v30CbduZp/Y5cHdq2nqqnJLlMqYOSkCLFJ0t5W333iwul8edLxzzlW95RtZ6vS9OkG6VlH5V77I8d95jqptRV3q48Ldy6UNPXTdeMdTM0b/M8rclYo535O7WrZFeFonWJ7Zorvml9JXVpq9QzBinvuzkq3raTZakAKm+TliiYUDOVvUkLQakoMHPmTHf76KOP6vHHH3dBnWuvvVaXX3658vLydMEFF+jHH39UTk6Oon3sVj8qNjbWFTRfvHixGjdurJSUFFcw3epV2WZBqY0bNxKQAhC1XcmaVY9V0xYnSF9fe+A7ptYOwsS3izbo3AGtXfDV6qUlxMWqekqiNu7I1YxffLVBDvhCpsc1UrtzpQ9O9t1PayZtnCGNO1eq1VFKruXr0JRnTQJKy710b9QPo9S/SX/1bthbPRv0dNuVva90wag5G+fouzXfaf6W+fpw0YcHn0G1OyhVbfjJqtKni8uKqnXXlS4IWzD3F+1asU4FPy5U0dpN5Ro/AERmk5bomFCjSQtBqahgXeqsPpIFYqzbnLn44ovVokUL93WzZs1cYMduo33sFoCyLKuvv/5aV111lSts3q5dO1eTyoJclnllAa7u3bsH9fcCgFB2JVu9YbOaFn96cDsOUq2doKPWTtRZvilLz32xQCf1aqa6aVWUmBCnzNxC/d9XC5WZd5An5nW6SZahlLNBGviM1Pliaf00X2BqzURpy89S3v7/lg5UqUp1/+T73dfVq1RX38Z9dUyLY3RE0yPUrX43HdX8KLeZhPsSDv4H7A6oFsyYp5y6NZXQvJESWjdVyslHu81suvROglIAKl+TlkidUDM0aSlDUCrC5efnq1atWsrMzFSVKlXUv39/JScnu0DMnoEf07ZtW0Xz2G25X926dXXddde5DCxbGmibadiwocuMWrhwoSt8ftZZZwXxNwOA0HYlm/D9bB2u/4RFrZ2go9ZOVPpgxkrNW71d7RvVUHb+Lq3Zlq1lG3+tMXXAfnjIzvx9gam4RClnk+8ipmE/qef1UtZaKX+7b0nfwdZg2+3aPteqWlI1fb70c/206Sd9tvQzt5kmaU3Uv2l//antn1wR9IoUQs/+YILbYpISldCysar076Hkw7qryqFdVfDT4nLvFwAis0lLBE+oGZq0lCEoFeHi4+N1xx13aNEiX0TVv+zNb/r06a7GUu/evd0Stmgeu78I+pAhQ1xHvxdffFELFixwRc4tuGVbzZo1XYFzq2EFANHalWzz9p1S8paQ19oJekcyau1EHSvef0znhkpNTtDidRmaNH+dsvOLyr/DzXN+/XrSTVK1JlKdrlLdnlK9nlL1Nr772evK/SNGDhnpbu87+j4t277MFTuftGqSvl/7vZZsW6K3573ttoqyQFRS13ZKaNNUBT8vUearHynzhfcUV6+WijdVLNsLACKySUukTqjtoYAmLQSlIp0FdgYOHOg2P8s28hf3tgLh1atXd8vYKsvYExMTNWzYMJ122mmaN2+eqy1lwa6kpCS1atXKZWQBQDR3JQuXWjtB70hGrZ2oUietim49rYc6NvEVejUlJaVasTlLP6/cpue/WnjwR2JMrG8mPbm27xjftsA3kz7/FSkh1ReYSm3su5Aph9iYWN3w2Q06tsWxGthyoFrVbOW2C7tf6J7/ccOPmrhyoit8Pnbh2IMP0O5e3mFdmWrff50S2jV390vzCpT9wdfadvfTBKQAVJ4mLR41r/BSPk1aCEpFg+LiYsXFxe23vtKVV16p888/XzVq/HqCV1nGbgGuQw45xG0AEMmC0pXMg1o7Qe1IZqi1E1Wtwq3orQWkbClqapUEJSXEucK4reqnuZl1qyl1kNVGftX9al+78PHn+Zbw2dc5630z6cXln4m2INOTM550m+nVoJdOanuSjm99vA5pdIh6NOjhNhN3b1y5j/Gat1+u+Mb1lD9jrkqycpXUpY2qnXeiCn9ZpYxnRpd7/AAQUU1aPJhQ8wRNWvZCUCoK7C+oYzZv3uyes3pKliUUjiJ57AAQtl3JYkJfayfoHcn2QK2dSOe7cLBg66K1OzVj6WYN7d9KY6evUIu61dSkVqpe/Nq31N+Wg1oNyQPf9e4T+i4jpKw1vuV8Z30lNTrc93hKfWnGvys0+jjr/mQTbaXFmrVhltvu+eYeVU2oqiOaHaGhnYZW6PiOb9bA1RnJeOE9bbv9ScU1qKPk/j1U+/5r3XIPe1y7KrDEEQAipUmLBxNqnqBJy14ISkUhO1mzk7aHHnpIGzZsULVq1fSvf/0rbLOlomXsABBRXcmCXGsn6B3J9kCtnchmgVSLozasWdUFpIqKS1yW1CczV2lg10ZKio/TwrW+AvaWUXXgdudVWbA1tYEvENuovy8gNftxqf25UuvTKxyUsmCUfynfuZ3PdbeWEbh42+K9ip6Xd3mq1UozJRnZvp+3YYuyx36l6tcMU4xduBCQAlBZmrR40LzCEzRp2QtBqShkQZ3c3Fw9+eSTKirynag8++yzigSRPHYAiKiuZEGutRP0jmTU2okqddKT3ZLUgl0lZRcsQ3o0cUv60pITtTOnsPzF2BocKpUU+wJTjQ6TVn0lzXpUaj4kYO3CqyVW07MnPatT25+qKvFV3GMbsjbom1Xf6KvlX+nlOS+7QO1B2R2A27V8rUpz85V+xdmKSUxQ4aIVSuzQUgmtGivj2TEBGT8ARESTFg+aV9CkxXsEpaKABW9WrFihOnXquMLgpmrVqvrb3/6mTz75ZK/OdOEmkscOABHdlSzItXaC3pGMWjtRxbKjPp65Sis2Z7oAlbngSF+G0PdLNpUV8ffXVzsg/guK7Yul2DjpkFt899873tcy3C5mFrxWoXFbVpRduJzQ5gSd0/kcVystMS5RNZNrqkG1Bu4x2/4757/l/hm7lq1R9rjJqnbasS4wZZlRMSnJ2rVklVu2CgCVpklLkCfUDE1avEdQKsKNHz9eo0eP1nfffac1a9YoLS1NvXr1ch3r7rvvPrdVhrHvWrtJJRZZDoLYWtWV0LheUPaNSOb/9AzjIoqIGkHpShbEWjtB70i2G7V2omcm/cy+LbV0U6bWbMvRlsx8vTdtuY7q1FCzlm/Vu9OWHeRO43wdmgp3XwCt/Ub65q9Sm9OlxWOkVV9I5/3g+8lLxwbk97BlexZsfWfBOxrRc4Re/flVVU+qrgHNBmjk974AbXmVFhRqy7X/dMs7kg/3FU232fTt9z+nwvkH+W8DAJHcpCXIE2qGJi3eIygVwd555x2XUbRy5Up3PyEhQVu3btXnn3/uAj0//PCD7r33XtWvX1/RPHYLSK3pO8ydtAWDFc5t8v2bBKawD4JRiOSuZMGttRP0jmTU2okadswmJ8bptL4t9NXPa/X13HUq2FWsFyYscpv/b8AccJZUh2FSy5OkiTdIu7KlxDTfcj3bTGyCL1C19ANfFlUF+AOqXet11bS109zFil3MPDn9SReQSohLKF+WlH956pG9VXVgP2W8NFbb731GcbWrW5cYlqainJhQQ4Q3aQly8wqatIQGQakINnLkSBfUOfXUU9WgQQO3FG7Tpk1atmyZ5s+frxdeeEHt27fXTTfdpGgeu2VIBSsgZWzfLguLoBT86vaQ2p4pbZ0vbflJylgmFeWHelSISkHqSuZBrZ2gdiSj1k5UsZpRc1dtV17Bbwu7Hlxh8906XexbwmHHcN9/SB0vlJZ/4uvOZBcwO3+RvvlLYAYvqV5KPTdrbsv2cnfluseObXmsTmxzovo06qPsQl/QtFydma4/X0nd2yvns29VtGKdirfudHXUShrUUcEc398+cOAIRiGSm7QEv3kFTVpCg6BUhLIgzowZM9S8eXONHftr6nl2drYL9tjSOFv+9sorr+jSSy9VerpvLW84zMCE79iBA1jD3vRYacgrvg9Dv63zXDva9GXjpLkfh3KUiDJB60rmQa2doHUk2wO1diKbPwvq2K6NXM2RxrVS9MuGDC1at0PbswuUU1DOTLdvbpJSGviypBofKaXUk7pc6tssUGWBqYzl0rR7pUxfxnZFZBRk6O5Jd2vbHm3IHzjmAVfw/Id1P5QFqg6KdSNOSlRSr07KGv2Z8qf+WPZU+p/PUlKPDlp77KUq3ri1wuNHJcGEGiK9SYsHE2o0aQkNglIRymowxcbGuqLgixcvVuPGjZWSkqLU1FR17tzZbRbY2bhxo8dBndIIHjtwANqd7fsQXPGZlJTu+2Cs3dltddJaEJRCeHcl87jWTlA6ku2BWjuRzR9IHdytiaokxikpIVaXD+6oVVuytHRjhjbuzHPL+TJyDzIb2nVn2t2h6d1BUt3uUrPjpCZH+S7MGw/wbd/dGbDf5a15b7nbZunN9NHij3Ryu5PdMf7QVGtfXj6J7Vsopkqi4mpU+/XBuDiV5Oa7CzMCUvifmFBDNDVp8WBCjSYtoUFQKkJZEGfAgAH6+uuvddVVV7ni4O3atXNd7PLy8jRt2jTl5OSoe/fuYTcDE5ZjBw70gzCxmrR2svTp+VL+Dim9pVSvl7uo3zyvYh+EQNC7knlUayeoHcmotRN1nv5svto1Sle7htXVrE41Vz/NX9T/q5/WHvwOY+OlkiIpuY7vInzTbGndVN9zVetJDfv53rezyrHv/Xho0EMuAHvN+Gu0KmOVTht9mgvCFpUUua28SvILVLRmo5IH9FK1c09Qwc+LlXLc4Uo9YYCyx08JyNhRSTChhkhu0uLRhBpNWkKHoFQEspohdevW1XXXXaeZM2dq4sSJbjMNGzZ02UULFy50xcPPOuuskM/AaMU4adnHv469XpiMHThYFpAqyJQa9JPsJM6CrrYExLYl7yjzt+VQgPDqSuZxrZ2gdCSj1k5USYyPdUtTbTP1qierTf10dWteS9WSE8rahR8U/4VC39ulNmdIn1/iu2hxzxX5zkks8FoBVgzXsvwOa3KYrj7kaj32/WPK2ZVTVofkziPv1PsL39eU1eUMHsXEaNfilcr9eobSLjxZdUb+TSU7sxRbvZoLvGa9ThAh2hxcs4oDxIQaoqFJi0cTajRpCR2CUhHIitmaIUOG6Omnn9aLL76oBQsWuELh69evd1vNmjVdkfDhw4eHfAZG6S3KglJhOXbgQIOvHc6Xuo7wPXbM49LKz6Wtc6XMVb4tx1fbBwjbrmQe1doJWkcyQ62dqOA/fq8Y3FE1q1XR7GVbtHRjplu69+2ijW47kEZMf3gh3vYs33v05l+PEfW62XdeMulGKWdjucfvbyxgs+nWNnzPGmmNqjXSyW1PdoGrcgeldv9tZzw3RjEJ8ap69CFuNr1w3lJtf/B55U3dvTwRUSNoJcijZULNX08Ila9Ji4cTajRpCQ2CUhEsMTFRw4YN02mnnaZ58+a5+ky2/C0pKUmtWrVS//79gz+IA5iB2d/a3rAYO3Cwx7ktCdm5XEqoKjU8zLfl75SKcqQpf5fmvR7qkSLKBLwrmYe1doLSkWw3au1EPv/xe1z3JoqNjdGhbeq6mmkrN2e6uiOrt2bpy599gdhySWsuVanpCzzlbfn18dSGvnOTLy6r4Ph9nwsNUhsoOSHZHed+Hep0UKO0RlqXte6g92uBJ+vIZO3AS7NzXTH/LTf+2xXwj62W4o77opXrKzR2hF9w9uzDWmlTRq6mLtyoopLSA1+SXZkm1Pb9zNu9lBuVoEmLh80raNISGgSlokBycrIOOeQQt4XE/5iBCeuxIyoFJQXe/PikNP8VqboFXQ+Rmh4j1e4i1eog5fqWnoQ9Zhord1cyD2vtBKUj2W7U2oke/3hzhlve0bVZLbVvVF1tG/o289mPa8q/46p1pPztvuO58RG+bCkrgmsXNHYBs3upXUXNXD9Tf+71Zz123GPq2aCnFm5ZqH8e8093IfPhog8Pen+2VK/65Wdry1/+o+z3vlRilzau9khMXJxyPp9KQCrK+C/ILz6mnXILipSalKBxs1dXPCAVZRNqtjQ7qVs7leYXKOfLab6/AwJSladJi8fNK2jS4j2CUgj+DIxlTgEeClrIJS5JSkyVkmr4aqXNfV5Kqe+76Fn1pSICM42VuyuZB7V2gt2RjFo70WXOym1ue3PKUlVNitfFR7fTMV0auSL+FcoY2TTL13DFLsAHPS/lbpKqt5aSa0uzHgvY+CetnKT1WevVumZr3XPUPe6x/KJ8PT/7ef2y/ZeD3l+1s4eocMEy5Xw6RTHJSarzyF/dxbixeiObRtzlsqgQPazGzuipyzS4W2Ndc0Jnd9H+xpRf9POq7YH5ARE+oWbHvWWTxNXxFcsumL9M1U4fpIKfFrnPASbaKkGTFo8m1GjSEjoEpRD8GZiF4T8Dg8gXvBT43XlXqY2kfndKHc6TindJBTukn5+XZjwoLR+nSMBMY+QJeFcyD2rtBLsjGbV2oieb9U+9m7nleRZo3bAj12WKTFuyyWUGWu0RU+7rTTvWrcbI0Y/7LlbSmkq2xG7mI9KiNwNWe2TZjmXq/1J/DesyTH0a9lHLGi017pdxemDKAwe9vyqHdlF8o7ra+dSbrt5I9RsvcO/ZVsTf6o5Yx6aUEwYQlIoydq4yZuoyrd2WrTP6tlTX5rWUUiVBo6cu1bcLN7q/lXLXV4vwCbW4hnWVfuVQlWRkqXjzNiW0aqr87+ao7shbldiplXKnzJYKy9EMAZHVpMXjCTWatHiPoBRU2WdgEB2CmgJv+t8vdRouWR2c3I2+INXh9/vWt9vfQJhjpjHyBKUrWZBr7QSzIxm1dqKHvdvExcboyuM6Kie/SJsz87RuW47W78hVgxpV1b1F7bLCtxV6Z9owXfrwNKnVn3yz6Ru+93VoKj7IDMM/qD3SqkYr1Uutp9UZq/Xugne1dPvScu+vSp8u7rZo9QZXN63aOcerJDNHW297TAWzF6rJlFeV2LFVQMaO8JJbWKSvfl6nSfPWa9gRbXTu4a319zN6asLP6/TS14tUmlPJJtR2Z5VUPfZQ996e/eKXqtKvm0qyfBftJXn5ik1LJSBVWZq0eDShRpOW0CEohYqL4BkYRJfgpMBbtcY4qfUpvqWpk2/xHevNBkl9bvVd7Mx/2VdoMUwx0xhZgtqVLMi1doLZkYxaO9GlfvWqWrUlW1US4tS4Zopa1E1zJe/Muu05Wr6pAu+plsFt79Hth/nuz/yPtOUnBZJl/V3c/WI9cfwTKiwu1Kqdq9T3xb7q27ivq5nmL5Z7MEoLfRmEqWcOVlzNdCU0b6TMVz5S7uffKbFrW3ehXrhweUB/D4Tetcd3dsd+etVE93eQEB9b9tzRnRuqed1qGv3RRim7Ek2o7c7itsxBu5iPrV3dvd/veORlxTdvqPh6tZS3x0U9orxJiwfNK/xo0hIaBKUizOaMvIOvIXIAkrZnqWmQZ2BWl0hbtyyUArxaKHbrGu3xp41KLGgp8PUPkRLTfDVKts7zPbZuitT0aF9x/3ANSDHTGJGC3pUsiLV2gtWRzFBrJ7pY4OnmV6apae1UdWhcQ83rVFN6ihW/LdB3izZVrGlF779KvW+WqvgyQ13rcJtYsGWjdoyXt533HjVH/tT2T7rv6Pu0I2+HalWtpQ3ZG9QkrYmmXjJVF31wkV77+bedh/+X7I8nKfW0Y91mSgt2KfOlsa4obsqQ/opJSlDuhO/LPXaEH1uafUIv3xl4XmGRqiTEa1tWvpuMSIiPU2Zeofq3r68BPbtIU3bXcq0kE2qmYNZCxV6frJQhh7uMWAvc1r7vWsXWSFPOuMmhHh68atLiYfMKmrSEBkGpCAtIXTJqknYVB74GTOuYpRrlq0EXlBkYC0i1y5Xy3ztfgdZpc4o+lK9zARD4FPjdacMFGVLbs6UdS3Z/EDaV0lv6loOEK2YaI1rQupJ5UGsn0B3JqLUTfSzYWj0lSV/MWaNF63a6AJS1DM/fI9haroBUQorU+yYpa43v/bnF8b5OwKd/6svqnvN0hYJSflbs1gKsVsz/L/3+og8W+eqZ2Cx6nZQ65dpn8frNrvtSyqnHuEBU7qffqnDRCnd8p196ugrmLFb+9LkVHjvCR0JcjF6csMgFonZkF2j11mzX3CIjp9BdwNdMTXJZhZ1aN5cml1SOCbU95E//WXlTZrvi0Kb2v290t9nvfKH8aYHNfkQYN2nxsHmFoUmL9whKRRD7Aw5GQKp8Dm4GZmuplB/qISPqBS0FfsscafMsqemx0oB/SwU7pRptfBc2i95WuGOmMTIFrSuZB7V2At2RjFo70aValQRdf2IXLdmQ4WqOFJf4mmv3aFlbKzdnuaLnB293XpXNmsen+IreWtDVJhIsEBubICVU871/ByAb0GpJWQAqNTFVNZJraOzCsTq86eHu/vzN88u9/7xvZ7utrDNqTIyS+/dQ/qyFyvzv+xUaO8JPQVGJlm7IcO/1+1NYVKwl63cqsWCLBlWWCTV/lvegfu48Zdv9z6nqlFmqOvgwxdWpqez3Jyjj+Xdp1FKZmrR4NKFmaNISGgSlUH5RMAOD6BHQFHibaY9PlvJ2r+22C/Xx50vHPOUr5G8X8UX5vqKKyz5SuGOmMXIEvSuZB7V2At2RzFBrJzr4A6k9W9ZWjdQkTVu8yQWkjGWEnN2vlX5etU3/nViejLfdfxDWLty6MVVvJdVs78uSsjqXNpEQoAtxW5Y6a8MsXdT9Ive11U9rnNZYVx1ylTLyM/TNqm/Kv/O4OKm42GUCJnVvp6wxnyvrzfHKGT9FpXkFARk/wkNSfKz+eko39x6/YO0Ot0zbdGxcQ2u2ZrtmFtYI4L3vV6hOwZKDD0pF6oSav0vZXy5SbNVkrT/9eu188k23OfFxUlE5l68jcpu0BHFCjSYtoUdQCuUXqTMwiEoBTYHvcY3U7lzpg5N999OaSRtnSOPOlWp1lJJr+bp/5FkB9TDtWsdMY0QKeleyINbaCVZHMkOtnejgD6Q2q+urBLk9+9cc6trVqqh+jWSt3Zbk7pc7E3D9NN/Ft6t1WSglVZcGPecrhhugC3ErbP7xko91XpfzXJcmuz926FjVT62vZ2c+67ICy63Yd7GdctIRriNT9ntfub91W+aB6Jp8sPfzJrVT9fmcNWUBqbTkBFd2YPbyLXp/xkr3Oqu/lhyTJ/2vMhseTKjZxbs1tPBnDAbtjcKWOi1do7j6tVW8ZcfezxOQqnxNWoI8oUaTltAjKIXyi9QZGESlgKbA1+kmleyScjZIA5+ROl/su9CxwNSaidKWn6U9CiCGJWYaI1bQupJ5UGsnGB3JDLV2ooM/xpSbX6Si4hId36OpC74u25Sp43s2dTWmLCOwQrYtkJZ9InW73Le0o9sVvsfnviit/PVCo6KshtRfv/yrrjv0OjVNb+oCUfdPvl9PzXiqwvuOq1NDSd3aq2jtJpVkZrsL9PKlRiIc+f87rRC0Wb4ps+y5uunJat0grSyb5KCCsx5MqA3tPNQd79+u/ta9v2/P2668AExm7PuPk9iuueKb1ldSl7ZKPWOQ8r6bo+JtO2nMUlmbtAR5Qo0mLaFHUAoHLsqWNCF6BDwF/gcrZGiZRrt8FzU5m3zLVRv2k3peL2Wt9XUAseN/W/lrhwQVM40RK/BdyYJfayeYHcn8qLUTPb5bslGXDmyvfu3qqU+bumWz59alafZy3znGQWVJxexekt36VN+x/f290qI3pOZDpPQW0vbF0ox/+d7TK7As1S7IHx70sMv8e/WnV/XkjCf1zMxn1K5WO7Wo0UKbsjdpS+4ercrLKaFFYyUP6Kmstz/1PWDH/O4MKkQ+/7FtmSW2fPXw9g20aWeey4g9tG09F5y1TKmD5sGEmjWvqJtS12WTLNy6UNPXTdeMdTM0b/M8rclYo535O7WrAn9n/qBUteEn+2oJFhWr1l1XujIEBXN/0a4V61Tw40IXsEUladLiUfMKQ5OW0CEohQMXDUuaEFWClgJvrWX9Jt3kazlbp6tUt6dUr6dUvY3vfva6Cow9iCnwzDRGtMB3JfOu1k4wOpKVodZO1Fi/PVd3vv2DTj2kuSuCWyutiitwbsVx9zzOD5g/i6jThVKzwdIv70vrpkrrv5fKmZ23ryObH6mnjn9K6VXS3TK93g17a+XOlW753iEND3Gz6fZ13L1xFf5ZVh9ty80Pq2Dm7kkPsqSi0reLNujcAa1dPcyuzWspIS5W1VMStXFHrmb8svngg7NBnlCz85ZRP4xS/yb93fFvF+22Xdn7SheMmrNxjr5b853mb5nvLuDLlUG1u6RAwYx5yqlb09UPTGjdVCknH+02s+nSOwlKVYomLd40r9gTTVpCh6AUDlw0LGlCVAlaCrx/1t3azFrw1ZaDWM20+a9ICam+wFRq4wp9EAY1BZ6ZxogVnK5kwa+1E+yOZA61dqJiEqFlvTT3fjx31XZtycxXvfRkVwvQvi5Xm3Bn95v7+u987802o+4ernhAyp8FaAFXC0jdPelu14HJLsQ/PvdjJcX7amAZax0eCCVZOcp6/ZNf38+p/ReVbCn2c18s0Em9mqluWhUlJsQpM7dQ//fVQmXmlWPyKMgTavZpZEtU/cWfbVn2MS2O0RFNj1C3+t10VPOj3GYS7ktQRWR/MMFtMUmJrkh0lf49lHxYd1U5tCtZJZWmSYt3E2qGJi2hRVAKBy4aljQhqgQtBd6v+9W+C/jx5/mOd/s6Z73vAr54cvimwDPTGHGC25XMm1o7Qe1Ithu1diKX/3/pjjN7uizRK56b7AKttll3pqzyXITvvhhyaraTOl8i1Wgrdb9SWvSWlL1Bch2USisccLWlqeOWjNMT059QRkGGC1JZnZEvl33pjm1bqmpfH6yEds0Vm1JVRavXqzgzZ69MVuvOZM0qijdvV/53lbtdeLT6YMZKzVu9Xe0b1VB2/i6t2ZatZRt/nWA7KEGeULu2z7WqllRNny/9XD9t+skVg/YXhLZl2v2b9nd/J/Y3U9EscAtEJXVtp4Q2TVXw8xJlvvqRMl94T3H1aql4ExPglapJiwfNKwxNWkKLoBQOnAdLmoCwSIH3n0x1GeGbcbdj/6yvpEaH+x63GZoZ/y73eD1JgWemMaIErSuZB7V2POlIthu1diJ7Fr1tg3TVr1FVr0xcXLZMz5am3nFmL81fs13PfL7goPdd9qfQ/HhfQKqkWOp3l68z8MbpvvfvNZN85QXKybJabRIhIS7BZYfYBUvHOh01Zv4Yjfh4hMsOLK+6T9zmOjHlfTNT+T8uchfgRSvWqnDJKpfpWveJvyvzlQ8JSkWZ9KqJOqZzQ6UmJ7gi0JPmr1N2vi+rosKCNKE2cshId2t1A5dtX+YyvSetmqTv136vJduW6O15b7ut3HbXC0w+opdq33+dC9jafVuenf3B19p299MEpCpjkxYPmlfQpCX0CErhwHmwpAkIfQr87ssnywK0lGHLEGzU3xeQmv241P5cqfXpFQpKeZUCz0xj5AhaVzIPau141ZHMUGsnMvkT2to39nVOsqV6ftWrJioxPtbVlnKvLc9Meu3OUnwVaemHvvOUBodKtTr4tk6Spt5ZoaDU8a2Pd7fWLrxL3S5uubUt6ysuKXbBKQtKWaDqoINTMTG+eiKxsUo+uo/bTNGajcqfOd/NotuWN/XHco8d4adOWhXdeloPVwzar6SkVCs2Z+nnldv0/FcLy5fbF8QJNTveb/jsBh3b4lj3d9CqZiu3Xdj9Qvf8jxt+1MSVE13Wty3bLlem1O4s75q3X674xvWUP2OuSrJyldSljaqdd6IKf1mljGdGl2v8iLAmLR5NqNGkJXwQlEJYLWkCQp4C7796sosam3G3wFSjw6RVX0mzHvV9IBZWrIZNUFPgmWmMaAHvShbEWjtedyQz1NqJTP5jtnB3dpRliFi9kY07czWgYwPXqOKLOWvdc7a0z7/M43+Jj43R4H69pSOvkjJXSZtnS1PvkDJXSvX7SI2PkNqcIa3YnVlXTp8u/VT3fHOPTmt/mssCbFCtgXvclvAd0ugQbcvdpidmPHHQWSKuLXiplPHCe9r1y2rXhalK326Kb1JfqU3ql70ub9KMCo0f4cFKDZSUluqIjg1cQMq6A6dWSXDZgnYst6qf5jKobELt4IOzwZ1Qs3MRe2+3zfRq0EsntT3JBWztb6BHgx5uMxUp9m8ZJJY5aH8T225/UnEN6rgL+Nr3X6uUIYe7x7UrQBllCN8mLR5PqNGkJfQISiFsljQBYZEC7z/ObRYmNk465Bbf/feOl6o19S1bXVD+GZOgp8Az0xjRAtuVbPdFSpBq7QS7Ixm1dqLPjyu3ugzAHi3ruNpppkW9NN9M+pKN7r5dtB8oq1NywhE97bt8s+fVW0rVW0uv95LWfevbpv+zwuO2LKh7v7nXbSkJKepev7tO73C6hnYa6ro02Rb3QzkuxONiVZKVreL1W5T58gdus+XWiR1aKvWswUq/7Ay31Lo0v7wF4BFefMe21b9ctHanZizdrKH9W7kC0C3qVlOTWql68etFBx2c9WpCzSYhjC1lsvqBtlmwtmpCVR3R7Aj391Duhi27x2/dgk1Jhi/rsHjDFmWP/UrVrxmmmPg4AlKVpklLcCfU/GjSEj4ISiFsljQBIU2Bt5OthBSpcHd21dpvpG/+KrU5XVo8Rlr1hXTeD76/g6Vjyz1mL1LgmWmMLMHrSlYatFo7XnQko9ZO9LFGFKM+m6cRAzu4YJR1ZsrIKdQjH/3sArIHyl9X7chODdx7vj4/T8peL53whlScL6U28tW3tPf1AF/E5OzK0dQ1U9128xc3q3bV2u7ipTxFzq0b6upDhynGMlxNXJyrO1IwZ5Fi01JcUCpvkn3uIBrYoWrv9w1rVnUBKVuqbZkkn8xcpYFdGykpPk4L1+7Y/drSsJtQ89fVsff/czuf626tQcvibYv3yvgul92/767la1Wam6/0K85WTGKCq79jQdqEVo2V8eyYCo0fkdKkJfgTanuiSUt4ICiFsJmBAUKaAt9hmNTyJGniDdKubCkxzXds2+Z+aIIvULX0A99JXzkFNQWemcaIFKyuZMGstRPsjmTU2olOyYlxmjhvvZZuzNQJPZq4Y9+Oc8uUKs9ywENa1dXiFavVdfl4qSjX1wG4yPa1+wUV7AB2ILbmbi1b6nEwqialqs6DNyjzzXEqmLWgbImHn2VMFW/dQWemKFMnPdkFZwt2laiwyHd8DunRxJ3PpCVb/Z2DnIDwaELNr1piNT170rM6tf2prji02ZC1wb3ff7X8K70852VXN7O8di1bo+xxk1XttGNdYMrOVywzdteSVa5xCypDk5bgTajtD01awgNBKYTNDAwQ0hT4Thf7ivVbcLXvP6SOF0rLP5E2zvB9AO78RfrmLwEZfdBS4JlpjDjB6koWZyc8fW6VDrs3aLV2gtmRjFo70SMtOUHXHN/ZvQ9XTYpXmwbprhFFUXGp2jZKV3JivB54b7YLVB2MuunJrojuBgu02/FstS2tA/DcF3xZU074zkQf0/VEVTv/JOV+PV0FCfFKu/AUFa3brNxPp7jnc7+cplWdTg31MBFglh318cxVWrE50wWozAVH+iaSvt/dzOKgOqx6NKHmz449oc0JbvJhTcYal1VSM7mmq7Fmj9n23zn/VUVYpuCWa/+p4i07lHy4b4KueNtO18GscP6yCu0bEdSkJcjNK/ZFk5bQIyiFsJqBAUKWAv/NTVJKA99JXeMjpZR6UpdLfZsFqiwwlbFcmmYX+SvDNwWemcaIEqyuZO2aN5H6/dn3HUGqtRO0jmSGWjtRw4JHJ/Zq6r62wJRNFPgV7A7ALl538F17D2nlKz7brkVTqe4Lviwp69iUWE1qeJiUt0XKWufLoApDAzod57JY7WIlsV1z1X7gOlcA1wWl/Mv5KOQfNfzv32f2bamlmzK1ZluOe79/b9pyHdWpoWYt36p3p5Uj6OLhhJqxcxarffnOgnc0oucIvfrzq6qeVF0Dmg3QyO999TIPmr9By5G9VXVgP2W8NFbb731GcbWruwxCmrNUniYtXkyouZ9Dk5awQ1AKYTEDA4Q8Bd5O3rS7Ns27g6S63aVmx0lNjpLq9pAaD/Bt390ZkPEHMwWemcbK25XMP8Peq1NbqbRI+vSioNXaCVZHMkOtnehimVGrt2ZrwdodrmaaZUsNP6qtvp67Ts99ubAsOHUwfli2Ra9P/kXHtKuhxvUa+MoKmHbnSPV6+4JUPz7h6wwchvq2O0oF3y91ma0ppx7jHrNl1mV/yEUs7YgmpbuXr57Wt4UrBG3Hvh33L0xY5DZ/WYKD7rDq0YSaf8m2vddPWztNO/N3ukzZJ6c/6QJSljFb7iyp3b93jevPV1L39sr57FsVrVin4q07XSfhkgZ13Hs/or9JixcTaoYmLeGHoBTCagYGCFkKfGy8VFIkJdfxXdxsmu1rP2uq1pMa9pPq9ZKyfAGCsEuBZ6YxogWqK5n/eO/Yqrm06kspiLV2gtWRjFo70SetaqKbPFiwZodmLtui5ruz/35auc1dmO95QX6gNmfk6Y3Jv2j6lKUalfYP30SCNVxpN1Sq0cb3ojnlb00fTM3Sm6lGam3lxS5T1cGHuQYUJn/a7osUAlJRySbMLCibV/Db/9+DLmweggm1ein1tKtklztnyd3ly0A8tuWxOrHNierTqE/5l2xbBmVSopJ6dVLW6M+Uv0etQOtWltSjg9Yee6mKN/oybhB9TVo8m1CzbFrRpCUcEZRC2CxpAkKaAu+/SO97uy81+PNLfMtT3XNF0rKPfRmB4ZoCz0xjRAtUVzL/cqnUlGQpLsmzWjuB7EhGrZ3oYhlS63fk6tA2dV2HpskLNqhTkxrasD3XLe2o0AW5n3VjskkE2765WUqu7esObIHZMDSk9RB3W/WoQ5TUtZ0r2l+yLUMpJx7pAq5W0L9o41ZXGxCRzx90PbZrI5cl2LhWijv2F63boe3ZBcopqEDzEY8m1Iw1s7CL+G15v05yPXDMAy7b+4d1P5QFqsojsX0LxVRJVFwNX8DaiYtTif0NlJYSkIryJi2eTajRpCVsEZRCWC1pAkKWAu//oGt7lq944uY9PjB63exLI550o5Tjy1oJuxR4ZhojWqC6ku1Za0fNBkq1u4Sk1k55O5IZau1EF8uQemnCIm3NzNcxXRq6JdbWMnzR2h171VALqLytAZ1ECDRb+vrMpw/qsmZn+y5gdqv71N+1a/UGlWzd6Yr8Z79PDcBo4D8nGdytiaokxikpIVaXD+6oVVuytHRjhjbuzHPnMgeaVRLKCbW35r1Vlu1nmSQntzvZlRx4aOpDFdpvSX6Bu3BPHtBL1c49QQU/L1bKcYcr9YQByh7vm5BAdDdp8WpCjSYt4YmgFMJmBgYIaQq823FzqUpNX+DJLtz9Uhv6ivt/cZkCIVgp8Mw0RpZgdSWzWjufTp6uEztU8Z3QUWsHYZAt9dSn8/Tzqm266Kh2alCzqjo1rem6plpdqMrGLoKeGf+ghoz5zNUYSerSRinHD1DqaccqoXkjybb/hm9QDeVj9XXaNUpXu4bVXd2djk1quM189VM5z6M9mlAzDw16yNXDvGb8NVqVsUqnjT7NZUnZ8ifbyi0mRrsWr1Tu1zOUduHJqjPybyrZmaXY6tVc6YGs1z+u8NgR/k1avJpQo0lLeCIohbCagQFClgJvqtbxfQhaoNVmaezkrloT39JVyxq0JSIBEKwUeGYaI0uwupJZrZ3xU+bqxJk3+LqnUmsHYcKW7q3YnKWTezfTSb2bKT4u9uDr/0WZ0pw85X//s9u23TVKsbWqK/nQLsr9Zmaoh4YAsot06xZsm7HlTW3qp6tb81qqlpygrPwDX+rk5YRajGJc05XDmhymqw+5Wo99/5hbrm2qV6muO4+8U+8vfF9TVlfgHGP3ZGLGc2MUkxCvqkcforgGdVQ4b6m2P/i88qZSeyfam7R4OaFGk5bwRFAKYTMDA4Q0Bd5smiVlLPPNyAx6Xsrd5Ov6YfVJZj0W0N8l4CnwzDRGpGB0JdsLtXYQZtZszdYLXy3Ulsy8slbhFS0pFU1Ktu1UDpMIUTeZdsXgjqpZrYpmL9visl/tvOXbRRvd9ut0RPhNqPkDCpZVkrcrT58t/azsuUbVGunktie7wFV5glIWeEpo2diXOZKdq13L1mjLjf922YOx1VJc5nfRSv+yLURzkxYvJ9Ro0hKeCEohbJY0ASFNgfcHYa2b5NGP+07u0ppKcYnSzEekRW8GbPxBSYFnpjEiBaMr2R+i1g7CQEFRicZ8t7zsPjEpRCv/+/dx3ZsoNjbGFfy39/yVmzO1eF2GVm/N0pc/+2pjllsQJ9T8dTAbpDZQckKyKzvg16FOB1ccep0tqSoHm0CrfvnZ2vKX/yj7vS+V2KWN61oWExennM+nEpCqZE1aQjGhRpOW8EFQCmGzpAkIeQq82TBd+vA0qdWffHXTNnwvrf1GKq7Y+u9gpcAz0xi5POlKFmGotQMgGv3jzRlu8qxrs1pq36i62jb0beazH9dUbOceTKjNXD9Tf+71Zz123GPq2aCnFm5ZqH8e809Xi+fDRR+Wa5/Vzh6iwgXLlPPpFMUkJ6nOI3/1FYq286JrhmnTiLvcuQ0qR5OWcJhQo0lL6BCUQlgtaQJClgJvRf2bDZLaD/Pdn/kfactPARp98FLgmWmMXCHpShZBqLUDIFrMWbnNbW9OWeoaW1x8dDsd06WRvl+ySUUlpRWvqxakCTW/SSsnaX3Were06Z6j7nGP5Rfl6/nZz+uX7QffrKDKoV0U36iudj71pluOXf3GC1xAqnjrTsUkJii+WQPXuYygVOVp0hLpaNJSMQSlEDZLmoCQpsD3/qvU+2apim8ZoJZ/IrU+xdd90oKvRXlhmQLPTGNkoyvZgaPWDoBI4u869qfezdy5idW+tGLQtsRp2pJNrlnL2Okr3GvLnRQb5Ak1ExcTp2U7lqn/S/01rMsw9WnYRy1rtNS4X8bpgSkPlGufVfp0cbdFqze4zsHVzjleJZk52nrbYyqYvVBNpry61xJuRH+TlkhGk5aKIyiFsJiBAUKaAm8FFXvfJGWt8R3bLY6Xlrwjnf6plJgqzXm6wkGpYKTAM9MYPehKBgDRxd6642JjdOVxHV0x6M2ZeVq3Lcct225Qo6q6t6hd1pWs3G/zQZ5QM8WlxWpVo5XqpdZzS6zfXfCulm5fWqF9lhb66memnjlYcTXT3fLszFc+Uu7n3ymxa1u33MmWQaGSNmmJMDRpqTiCUgj5DAwQuhT43fOYVh8tPsW3Xt2yAXcs8WUIxiZICdWkgsDN+AQyBZ6ZxuhCVzIAiC71q1fVqi3ZqpIQp8Y1U9Sibpo1y3XWbc/R8k1Z5d+5BxNq1oTl4u4X64njn1BhcaFW7Vylvi/2Vd/GffXDuh9cwKo8sj+e5GoF2mZKC3Yp86WxKi0oVMqQ/u6ini5l4cvzJi1hjiYtFUdQCiGfgQFClwK/+8WpDaTSIql6K6lme99JXUp9XwtaO9EL0xR4ZhqjD13JACB6WODp5lemqWntVHVoXMNdvKenJLpC0N8t2rTXOc6BC/6EmmVvW9mBP7X9k+47+j7tyNuhWlVraUP2BjVJa6Kpl0zVRR9cpNd+fq1c+y9ev1nb73/O1d6xQFTup9+qcNEKl+GdfunpKpizWPnT55Z7/AgemrT8Fk1aKo6gFMJiSRMQ0hT49dN8J28dzvMtSU2qLg16TkptKC16O2C/R6BT4JlpBAAgfNmFe/WUJH0xZ40WrdvpwklJCXHK32N508Gft3g3oXZO53Ncvcu35r2lv/T7S1lnsuzCbNVJqVOhfed9O9ttrjOZdSWLiVFy/x7Kn7VQmf99PyDjR+DRpOWP0aSlfAhKISyWNAEhTYHftkBa9onU7XJfEf9uV/gen/uitPLXLnnhlgLPTCMAAOGpWpUEXX9iFy3ZkKGvfl7rLtwtnNSjZW2t3JzlMr4rJIgTav7mLDaRZgGo1MRU1UiuobELx+rwpoe7+/M3z6/Y+OPipOJiVwszqXs7ZY35XFlvjncNLUrzCiq2bwQVTVoOHE1aDgxBKYR8SRMQkhT4mFhfkLX1qb6g6/f3SovekJoPkdJbSNsXSzP+JZXsqtC4g50Cz0wjAADhw1/b0pY21UhN0rTFm1xAytRMTdLZ/Vq5i/n/TqxgE5IgT6hZl+BZG2bpou4Xua/zduWpcVpjXXXIVcrIz9A3q76p2A8o9k3GpZx0hNL/fJay3/vKncOV7KzAJCM8RZMWBApBKYR8SRMQkhR4/3r3ThdKzQZLv7wvrZsqrf9eKmfhzpCkwDPTCABA2PCfXjSr6yv+vD371yVNtatVUf0ayVq7Lal8F+8eTagZy+r+eMnHOq/Leepar6u7P3boWNVPra9nZz7rmrRUVFydGkrq1l5FazepJDPbTazR4SOy0KQFgeALZwJ/NANj/DMwLU8K6JIm4GBS4Pu1q6dYO4PbHYCyFHirKVU+uz8x13/nO9atdpp7OLABqT9Kge9ct3PFU+D3mGmsde81irEg1e6ZRlvSBwAAvOO/Hs/NL1JRcYmO79FUHRvXcBNpx/ds6ibYrElL+Xa+x4Rar5uklIa+CbXv7pbGny99f19AAlJ+NoH21y//6upgWraUBaLun3y/7p50d0D2n9CisZIH9FT+jN2lBizrGxHbpGXpxkx3n5gUDhaZUgjZDAwQuhT43Qv9araTOl8i1Wgrdb9SWvSWlL1B2pUT0I/UYKfAM9MIAEB4+W7JRl06sL2bUOvTpq478zDWocyfUXLwS5z2mFBLbRzwCTXrEjy081A9POhh14jl1Z9e1ZMzntQzM59Ru1rt1KJGC23K3qQtuVsC8vOsQ/CWmx9Wwczdk3OcuwCVEkEphHRJExCaFPjdL2x+vC8gVVIs9btLanu2tHG6tHmOtGaStDUwRcKDnQLvn2nMevvTX2cad2dQAQAA763fnqs73/5Bpx7SXM3qVFOttCquwPnTn83fq/RAOE2oHdn8SD11/FNKr5LuzlF6N+ytlTtXunOXQxoeog51Oriv4+71ZWVXVElWjrJe370qwybUrDYmgEqHoBQ8nYEBKpoCn5NfpGWbMiueAl+7sxRfRVr6oS87sMGhUq0Ovq2TpKl3BiwotWcK/HWHXqem6U3LUuCfmvFUhffNTCMAAKHnb7jSsl6amyybu2q7tmTmq156srZl5buvM3LLu7Q+eBNq/qYsVv/SAlK2PK+opEhX9r5SH5/7sZLifROA5qPFH5Vr9Antmis2paqKVq9XcWaOVPjrqouYlGRVHdRPxZu3K/+7OeXaP4DIRVAKIVvSBIQiBT7Osoj63Coddq+UuUraPFuaeoeUuVKq30dqfITU5gxpxe6sowhIgWemEQCA0POfitxxZk/FxMToiucmu+wo2+pVT1ZWXgXLXwRpQs1f/9I6BY9bMk5PTH9CGQUZLkjVKK2Rvlz2pSs1YJ2D7evyqPvEbUrq2lZ538xU/o+LVPDzEhWtWKvCJatUpU8X1X3i78p85UOCUkAlRFAKIVvSBIQiBb5d8yZSvz/7jnerk1a9pVS9tfR6L2ndt75t+j8rPOZgpcAz0wgAQPhq2yBd9WtU1SsTF5edo1iR8zvO7KX5a7brmc8XHPQ+vZhQsyzuuil1lRCXoL6N+7ri5h3rdNSY+WM04uMRrllLucXEKLFjK1deIPnoPm4zRWs2Kn/mfMUkJbgtb+qP5f8ZACIWQSmEdEkT4FUKvL/mVK9ObaXSIunTi6Ts9dIJb0jF+VJqIyl7nRQTV6HlqsFOgWemEQCA8NW+cQ13a+cpftWrJioxPtZNrO15jhNOE2rHtz7e3Q5sOVBd6nbR9rzt7pymuKTYBacsKGWBqvIEp5K6tXNDz3jhPe36ZbVSThigKn27Kb5JfaU2qV/2urxJMyr0OwCITASl4OmSJiBUKfD+5X0dWzWXVn0pLR8vFeVK+dulooJff+LuFPbyCmoKPDONAACEtcLd2VHHdG6oDTtytXFnrgZ0bKAmtVP1xZy17jk7ryk9gPqPXk2omU+Xfqp7vrlHp7U/zWVyN6jWwD1u5y+HNDpE23K36YkZT+jteW8f/M7jYlWSla3i9VuU+fIHbotJSlRih5ZKPWuw0i87QwU/LVZpfnnrbQGIZASl4OmSJuBA+GcQA50CXzc9WakpyVJcki/QmrNeqtNVmvuC7yTPqXjdtGClwDPTCABAePtx5VbXlKVHyzqqkerLjm5RL007cwpcjUxTcoANSbyaUDN2rnLvN/e6LSUhRd3rd9fpHU7X0E5D1bpma7fF/VC+rnsFc3/R6kOHKcYmwU1cnEoLClUwZ5Fi01JcUCpv0g8V/h0ARCaCUpWclzMwwIGyWt12vhboFPhDWtXxfdFsoFS7i++kLiZWSqwmNTxMytsiZa3znfCFYwo8M40AAIS1TTvzNOqzeRoxsIMLRuUWFCkjp1CPfPSzq5F5sLyaUNtTzq4cTV0z1W03f3GzaletrcObHl6uIudVk1JV58EblPnmOBXMWuACUir+9ZrCzmOKt+5Q7oTvA/o7AIgcBKUqOS9nYICDPS4DmQJvfli2RZ9Onq4TO1TxndCl+lLT1e4cqV5v33H/4xPSonKkpnuQAs9MIwAA4S05MU4T563X0o2ZOqFHE3cmbaUHLFOqPLyaUPsjW3O36oNFH5Tre4/peqKqnX+Scr+eroKEeKVdeIqK1m1W7qdT3PO5X07Tqk6nBnjEACIJQSmEZAYG8DoF3mzOyNP4KXN14swbpIQUqW53qfXpUruhUo02vhfNKV9qerBT4JlpBAAgfNVJq6Krh3RS1aR4tWmQrszcQhUVl6pto3QlJ8brgfdmu0DVwfJqQi1YBnQ6TsUbtqhw4XIltmuu2g9cp6wxn/uCUv5JthImv4HKjKAUwmIGBvAiBX4vu3KkdVN92zc3S8m1pUaH+zIGwzAFnplGAADCV3rVRJ3Uu5nL3rYsbr+C3Vnfi9ftLNd+vZpQC5a+7Y5SwfdLtWv5WqWceox7LHvsV7/WESmiPAhQ2RGUQsTPwCB6BToF/g/lbZWWli813YsUeGYaAQAIX8UlpVqwdofb5q7a7rKlhh/VVl/PXafnvlxYFpyKhAm1QGmW3kw1UmsrL3aZqg4+TClDDneP50+b43sBASkABKUQDTMwiC5pyQm65vjObqYx0CnwkYyZRgAAwldcbIwKi0q0YM0OzVy2Rc13N2T5aeU2F5CKjYk5qJID4TShVl5DWg9xt1WPOkRJXdspJilBJdsylHLika7kQNGajSrauFWlub82tAFQ+RCUQkTPwCA6a5yd2Kup+zrQKfCRiplGAADCW05BkXq2rO22yQs2qFOTGtqwPVe/bMhwzwc8IBUBrPHLM58+qMuana3Ejq3KHq/71N+1a/UGlWzdqYwX3lP2+xNCOk4AoUVQChE9A4PoZJlRq7dmBzcFPoIw0wgAQHjblpWvifO265guDXVExwZuOd+itTu0JbPyfjZb45dnxj+oIWM+U0xKspK6tFHK8QOUetqxSmjeSLLtv1xnAJUdQSkAYSetaqL3KfBhjJlGAADCm523PPXpPP28apsuOqqdGtSsqk5Na2po/1Z6ffIvquxKc/KU//3Pbtt21yjF1qqu5EO7KPebmaEeGoAQ210dN7RGjRql5s2bq0qVKjr00EM1Y8aM333t888/rwEDBqhGjRpuGzhw4B++HjhgeywTQ+hYhtT0Xza79Pe7h/bWbaf30KmHNq/UKfD+mca1R1+iFS2HaP0p1yrj2TEq3rLDzTQm9e4kxVH3DQCAULOle3eNmalPZq5y9+PjYsvKP+JXJdt2Kmf8FBesAlC5hTwoNXr0aN1000266667NHv2bHXr1k3HHXecNm/evN/XT5o0Seeee64mTpyoadOmqUmTJho8eLDWrVvn+dgRZfYNdPg7msHzmcaXJizSuFmrlVdY5FLgq6ckaXt2fqVOgd93ptFmGVd1PV0rO56iTRffrhzrwgcAAEJuzdZsvfDVQv3360X6duEG91glm08DgAMW8qvuRx99VCNGjNDFF1+sjh076tlnn1XVqlX10ksv7ff1b7zxhq666ip1795d7du31wsvvKCSkhJNmMCyFZRf8hG9VP3aYUofcYbimzf0PVhSEuphVepsKUuBH/nJXJchZR1t/Cnw2BszjQAAhJ+CohKN+W55WbdgYlIAEIY1pQoLCzVr1izddtttZY/Fxsa6JXmWBXUgcnNztWvXLtWsWTOII0U0q37NMKVfcbbi6tRw9wvmL1O10wep4KdFyv16BlNbIU6BX7E5Syf3bqaTejfbKwW+hP8WAAAAAIhoIQ1Kbd26VcXFxapXr95ej9v9RYsWHdA+/va3v6lhw4YukLU/BQUFbvPLzPTNVgAmrmFdpV85VCUZWSrevE0JrZoq/7s5qjvyViV2aqXcKbOlwl2hHmal5k+B35KZp9nLt7rHiBMCAAAAQOQL+fK9ivjXv/6lt99+W++//74rkr4/Dz74oNLT08s2q0FVmVFjce96UVWPPVQxVRKV/e6XKt6WoYI5vmBoSV6+YtNSCUiFCVLgAQAAACD6hDQoVbt2bcXFxWnTpk17PW7369ev/4ff+5///McFpb744gt17dr1d19nSwMzMjLKtjVr1qgy42Jee9WLim9U16XdxNaurqSubZUz7htXUyq+Xi0V/uLrmgIAAAAAAKIsKJWYmKhevXrtVaTcX7S8X79+v/t9Dz30kO677z599tln6t279x/+jKSkJKWlpe21VTaxMb78qLMPa6UjOzVQ/O6etLSmlQpmLVRsSrJShhyu2GopKi0sUu37rlVsjTTljJsc6uEBAAAAABC1QlpTytx000268MILXXCpT58+GjlypHJyclw3PjN8+HA1atTILcMz//73v3XnnXfqzTffVPPmzbVx40b3eGpqqtvwWyW7C/BcfEw75RYUKTUpQeNmr6ZQtKT86T8rb8psJR/pC27W/veN7jb7nS+UP+2nEI8OAAAAAIDoFfKg1NChQ7VlyxYXaLIAU/fu3V0GlL/4+erVq11HPr9nnnnGde0788wz99rPXXfdpbvvvtvz8UcKy44aPXWZBndrrGtO6KwjOjbQG1N+0c+rtqvSseOppERVB/VzGVHb7n9OVafMUtXBhymuTk1lvz9BGc+/W7bEDwAAAAAARGFQylxzzTVu259JkybtdX/lypUejSq6FJWUaszUZVq7LVtn9G2prs1rKaVKgkZPXapvF27cowx6JUif2p05VuMvFym2arLWn369dj75ptuc+DipqDi0YwQAAAAAIMpFdPc9HJzcwiJ99fM6XfvCt3rr26VqVT9Nfz+jp24+uZvSUqtWjoCUPygVE6NdS9eoaPN2FW/ZsffzBKQAAACA/2/vPsCrqNI/jr83IY2QhBp670iRIkXAgiggdtfF3rD3XRdd/e/ay66uq6siqFh3FzuirggqiiKIiDTpHUKkh/Se3P/zO3jjTQghCeSmfT8885A7d+5k5ubMmTnvvOcMANSOTClUvFvH9FQcxmLqhlr72GgLqfNbPPLkni2sa+zZZrP+ZbZvpdVo+hK8Xgvt2s7qtGlmYb26WL3zT7WM+Ustb1+iWXZOZW8hAAAAAAC1AkGpWqBri/p2ev827ueM7FwLD6lj+1IybfHGPRZSJ9iSM7JtaLdmZn1uMPv6djNvfo0PSkVdfpaFD+zlsqIa3X+jG/A86+f1lrM53rKWrLbc7bsqe0sBAAAAAKjRCErVAiHBHntl9hoXiNqfmmXb9qZaeGiwJaVlW1pWrjWsF2Zt6wdZm/anm311q9Vovw5enrVwhaXFNrSQdi0tpFMbizzrZDfJrvH3EZQCAAAAAKCCEZSqBbJy823DjiRbumVfse9n5+bZth27rU3eZ1ZbpE6f7SZPWKiFdGhl4UP7WsTxx1r4oN6WtWxtZW8eAAAAAAA1HkGpGi6sTpBNOLuP7difbqu277fs3AOZQj1aNbC4vamWkpljaZm5NnvBYhtm/7DaQoGosN5dLaRzG8tavs6S3/zYkqd8YMFNG1neruKDdwAAAAAA4OghKFVDeX59lt6x7Rtb68b1bNbSuIKAVHREiF00rJMt3rTHPly4xS23OyHRLGJPZW+2efTP47H8ihjXKijIdd+LOKG/NX7kNgvp2s699mZkWer0r2zfAy8QkAIAAAAAIEAIStXs8bytc/MY93rTruSC92JjIqxT82iXJSVBimBVEeN6jrM2MW3su23f2dbErZaQkWAZuRlHdTyphn+53uq0amqZC3+2/JR0C+vV2aIuGWvZ67da0qR3js7vAgAAAAAAJSIoVUPlK/3JBZw8lpfvtWHdmtuuxAz7ZX+6DerS1OpHhrlMqarm6VFPW2xkrGXkZNjqvavth/gfbGH8Qluxe4XFJcVZYmai5eQfCKaVR522zS2sdxdLmvKB7fvLcxbcvIlFDO1rjR+51SJHD3PzLSf3qO4TAAAAAAA4GEGpGu67NTvsouGd7PT+bax3u0YWEhxk9SNDbef+dFu4fvdvAawqkC2lrnsTf5xoQ1sPtQEtBli/5v3cdOOAG10waunOpTY/br6t3LPSPlrzUdkyqH5NHQvr1cW9zE9Kdf/n7dhjqdO+tPq3XGyeOsEEpAAAAAAACBCCUjXcpl0p9uLnq+yM/m0tNjrcQkOCLTk92176crUlZ5Q/46gieM1rj3z7iPu5fnh9G9xqsI1oP8JOaHOC9WnWx05qd5KbJOThkDKu/EDqWM6m7eZNz7SYG35vntAQy16z2UK7d7CQjq0safK7R3+nAAAAAABAsQhK1QLTF26xFdsSrFvLBpaamWNx+1Jt487fxpiqKm4deKtFhUXZrA2zbNmuZTZzw0w3Sevo1ja0zVA7s8uZbhD08g6EnrMxzlI//daizj3FBaaUGeWJjLCcdVstdfrso7xHAAAAAADgUAhK1WAxdUNtRM8WVi8ixNbGJ9mclfGWmll1u6c9M/oZ9//DJz9sGxM2usHO52ydYwu2L7B1+9bZ2yvedtOR8GZl255bH7O8PfstYlhfNy9vX6IlPPKiZa/ceFT2AwAAAAAAHB5BqRqqSXS4/fncvtajdYOCefn5Xtu8O8WWb9lnL3+52n4dC71KCPIE2R0z77BT2p9iIzuMtI4NO7rpimOvcO8v2bHEvt7ytRv4fNrqaWXLlAoKck/eizhxgNUdOcSSXp1mCQ9NsuDG9c2Cgy1v176K2zEAAAAAAFAsglI1jJ62l+/12gk9mruAVNzeVKsXHmJhIcFWJ8hjHZtFuwwqjSmlsc2rSmBKQabnFj7nJunfvL+d0eUMG9NpjB3X8jjr27yvmyT4oeByjSfV4PZLLezYbpY28zvL3RxveXsTLeKE/pbfvIllLV1z9HcKAAAAAAAcEkGpGudAAGZYt+a2ZnuiLdyw28YN7WjTfths7WOjrHWjevbKVwcCMB6Px7y/BmyqgmDPgWBTnjfPftrxk5se/OZBqxtS105oe4KNO2Zc2Z645+P1mics1ML6H2Mp78y0zHlLCt6Kue4CC+vb3bafMt7ydu49mrsDAAAAAABKQFCqhsn3msuAatGwrgtI5ebluyyp/y3aaiN7t7SwOsG2evv+X5etOgEpXzDK15Xvop4Xuf8Xxi+0tfvWFhr0vDxCu7U3T3ioBTeI+m1mcLDlp2e6oBUBKQAAAAAAAougVA3UJCbCdiVmWFZOvmXnHhh7aXTf1q5LX3REqCWmZVtVFRUaZZPPmGzndDvHwuuEu3k7UnbYN1u/sS83fWmvL33dvOXodJifmWW5cTstYnh/i7rodMtavtYiRw2zeqcPt9QZcytgTwAAAAAAQEkIStVAyo76ZNFW27w72QWo5LITu7j/F6zb5f4P8hzIqqoqlBWlcaVO73y6XdjzQotLirPQ4FBrGNHQmkc1d/M0vbb0tbKv3OOxnLVbLP2rhRZ9xVnW5Jm7LT8xxYLqR7lBzlP+80lF7BIAAAAAAChBUElvonpRtz353eAOluf1Wty+NFsVt98++H6T7UvOtM+XbrdXZ1ftAb3VbW/dvnX25vI33ZhX+v+DVR/Y7rTddu/se8u30l+7KSa9+K6lTJ1heTv2uIBU9ooNtuePT1jGvKVHdycAAAAAAMBhkSlVgyj0EhEabOcObm9fLt9uX/0cb1k5eTZl9ho3+Z7MJ1UpS0qUJSW9m/a277d/b4mZiRYbGWvP/fCcDW873EKCQ8qcJRXcvImFdGhlWcvWmjc13XI2xtmeP/zdPJERFhQV6caYyt3ySwXtEQAAAAAAKAmZUjWMxoz6eWuCZWQdGDTcX1Ub2LyoppFNLSc/x3XbS89Jd/NO6XCKndvtXBvRfoSlZqeWaX3qqtf8P3+zyFFD3evQXp0t6pKxFnX+qS57ioAUAAAAAACVh0ypGsKXBXVK75bWuXmMtWoUaet3JNma+P2WkJplaVm5lb2Jh5WUlWQPzHnA9mXsK5j36IhH3YDnP8b/WBCoKq2o34+27FUbLe2zueaJCLMmT02wsD5d3Xv1b7nYdl17v8uiAgAAAAAAgUdQqobwZUGd1qe1hYcGW1hIkF1/Wg/buifFNuxMsp2JGa47X1J61X3ynry14i33f9uYtvbx2o/trK5nuafuPTHviTKtJ3xQL6vTMtYSn59q3vRMq/+Hy1xAKm9vonlCQ6xO2+YWefpwglIAAAAAAFQSglI1zAszV1rXljHWtUV9a9skynq0buAm+XLZdqvKnjj1CYsKjbJbZtxiW5O22rnvnOuypHLzc91UFuEDe7n/c7ftsNBu7S3qwjGWn5xme+952rIWr7bWc9+00B4dK2hPAAAAAADA4RCUqkFC6wTZwg273SRN60dY52Yx1qddI4uKCLGUzByrajzmMa957fjWx9vNx91sTy942tJy0tx79cPr230n3mcfrv7Q5m6bW6b1erMPBLHq/e40C24YYyHtWlryGx9b+qz5Ftq7i1lQkGWv3lQh+wQAAAAAAA6PoFQNGk/qhtN6WMOocFu8cY9t2Jnsuu59t2anmzxWNXk8HvN6vTayw0jLyMmwmRtmFrzXMqqlndXlLBe4KmtQKvWTOVbv3FPcJN6sHEt+dZp5s7ItcvRQ84SFWPrsBUd9fwAAAAAAQOkQlKpB40mNOra1BQV5bFDnWMvOzbctu5NtbXySbdubYl8sj7esnIOfyFfZ8r357v/m9ZpbREiEe/KeT/cm3a1ldEuLT4kv83rzftltCY+8aJHnjHCBqPTPvrPsNZvdWFIx48+zrKVrLfOHn4/qvgAAAAAAgNIjKFWD/N/UhW78qN5tG1m3lvWtS4sDk8xcEmdV2aJfFtl1/a+zp0c9bf2a97PVe1bbYyMesyBPkH205qNyrTPju8VuUlc9y89XWpZFDO1rmT+ttuTXPjzq+wAAAAAAAEqPoFQNsnTLPjdNnbvB6obVsatO7mojerW0Bet2WW6+14I8ykyyKmnOljn2S8ov1qlhJ3vwpAfdvMzcTHt58cu2PmF9+VYaHGyWl+eeuhd2bFdLeXeWpUydYWkz5po3I+vo7gAAAAAAACgTglLVnMaKUpzpzAFtXfe8DTuTbMf+dEvPyrXv1+2yVo0ibdoPm92yv/byq3KCPcG2cf9GG/rqULu418U2sMVA69Cgg326/lN7dO6j5V9x3oHuipFnnGAx111gqR986b6r/MSUo7fxAAAAAACgXAhKVXMKsgQHeezGUT0sLTPXdidnWPy+NPtlf7o1b1DXjm3f2A0k7lu2Ksrz5lnHBh2tab2mti1pm72/6n3bkLDhqKw7uEkDC+vTzXK377L85FTXha/KRucAAAAAAKhFCErVAM3q17Wte1ItPCTYWjWMtPax0S72IvEJabZpV9XNDAqvE25XHXuVPTvmWcvOy7atiVtt8CuDbXCrwfZj/I8uYHUkQtq3sojh/Szl7c8OzND4Ur9mUAEAAAAAgMpDUKoGUODpzje+tzaN61n3Vg2sXZMoi4kMtcS0LJu/Zlehbn5VhQYw15P3zuxypj188sO2P2O/NarbyHak7rDW0a1t3tXz7MrpV9q/l//7iH5P9upNtufOJy1r0coDM8iSAgAAAACgSiAoVQMM6hxr9SPD7POlcbYmPtEFoMJCgi0z57eMoKoairmw54UWnxJvb614y/405E82fc10Nz81O9WaRDY54vXnp6RZyn/+d+CF0sf0FD4AAAAAAFDpgip7A3BkosJD7PaxvWxI16YWpMfr/RqA6tuhsRtTqqpSlpRoLCkFoOqF1rMGEQ1s2upp1jO2p3u9cvev2U2lFNK1nYX162HBjeubhYYUes8TGWGRZ59s4ccfe1T3AwAAAAAAlA+ZUtWU4k/5XrN+HRpbg3ph9v3aXZanGWbWsF6Y/X5IR1u+dZ+99vVaq6pCg0Ptpx0/2ZXHXul+zsjJsFbRreym426ypMwk+2brN2VaX+yz91hY7y6W8c0iy1yyxrKWr7Pczdste91WCx/Yy2KfvdeS3/jIMucvrbB9AgAAAAAApUNQqpryDY3UNjbK/Z+QmlnwXuOocGvWIMK27wsrFMCqajSw+SfrPrFLel1ivZv2dq+njZtmzeo1s8mLJltm7m/7dFgej4X26OgGMo84eaCbJDdup2UuWmmesBA3ZcxbUnE7BAAAAAAASo2gVDXlizGlZ+Zabl6+jenbxtIyc23jrmQb06+NG2Pq+3UHBjmvyjSG1IQvJthtg26zNjFtXCDqkW8fsecXPl+m9YT16eq+lKQpH1jO+m0WefpwCx/cx+q0bmb1WjcrWC5jzsIK2AsAAAAAAFBWBKWqufnrdtr4kd3cmFIDO8e6Qc5l/Y4kW7xpr/u5KmVJBXuCbVzPcfbkqU/ahoQN9uayN+25hc/ZpEWTrGujrta+QXvblbrL9qTvKeOKgyw/JdXyftljya9Pd5MnLNRCu3ewehecZjHXnG9Zy9aaNzO7onYNAAAAAACUAUGpau6XhHS77+0f7Zzj2lnbJlHWKDrctuxOsRdmriz09L2q4sR2J9rzY563mPAY101vQIsBtiVxi+u+d1yL46x7k+7u5+CHgsu03qyf19u2QRebJ+jXsfuDg82blW1ZS9dYUHSkC0plzPmxYnYKAAAAAACUGUGpakjZUEp+6tA02o0X9fPWBNuTnGlNYyJsX0qm+zkpvYplBHkULMq3C3te6AJSD8x5wHLzc+3GATfaJxd9YmF1Dox/JR+v/bhMq64bVs+aPH6HJU/91LJ+WuUCUpb3W0BOGVN5e/db+uwFR3WXAAAAAABA+RGUqoZ8vfH++rt+5vF47IYXv3XZUZqa1o+wlIwcq3K8+e6/M7ucaZ+u+9Se/eFZS8pKckGqltEt7YuNX7in7e1I3eF+LosRvcda1KVnWPpXP1hWSB2LvuJsy43fbemfzXXvp3/xvW095pwK2S0AAAAAAFA+BKWqaZZUl+Yx1qxBXXvj67UF3fTCQoLtr7/rbyvjEmzSrFVW1Wgg89jIWAsJDrHBrQbbtqRt1qNJD3t35bt27SfXWmp2arnWO/yYUZa3Y49lr95koV3bWeNHb7OUd2cdCEr5uvPlHwiKAQAAAACAqoGgVDXj8Zh5vWbdWjVwr9VVz6d+3VALrRPkxpZyy/plVVUFYzqNcf+P7DDSesX2soSMBAvyBFlefp4LTikopUBVWYNTg7ueZFkLNljOpu0Wec4INy912pcH3lT/xtyqN7YWAAAAAAC13a9pJKgufE/Sy/41O2pEzxbWo1UDa1gvzIb3aG6tG9ezRRsOPLlOXfuqks82fGYPfvOgrdi9wppHNbdjYo9x89WF79/n/tumnDnFzuhyRpnW2TamrTWo19gFn+qedrxFjh7m5md+v/TAAgSkAAAAAACoksiUqqaWbNlraZm51rdDE2tQ78Ag4e2bRltiWpbNX7fTvc5XSlUVoiyoh755yE2RIZF2bLNj7bzu59m4Y8ZZp4ad3BT8Y9meuje602j3f92TjrOw3l3NExZi+fuSLHLsiW5w89y4nZa7c69503/LKAMAAAAAAJWPoFQ1tSsxwybOXGHXjuzuglHpWbmWlJZtT3283H5JSLeqLi0nzebFzXPTnZ/faY3rNrZhbYaVeZBzZV9N+uxxu6bt7y20R8eC+bHP32s523ZY/t5ES5rygaV+OLsC9gIAAAAAAJQXQalqKiI02L5e8Ytt2Jlsp/dt7caO0tP3lClVHe1N32vT10wvV/bVpBmP2+h3Z5onMsLCenW2yDHDrd65p1hIu5Zmml4r+3oBAAAAAEDFIihVjURHhNgtY3qa1+u1umF1rHPzGEtOz7bcPK91aRljEaF17NEPFrtAVW3kTcuwzAXL3bTv/okW1Ki+RQzqZenfLKrsTQMAAAAAAEUQlKpGYmMibGz/Nu5nBab8BzLP+nXg87XxiZW2fVVN/r5ES5sxt7I3AwAAAAAAFIOgVDWjzKhte1Nt1fb99vPWBJctdflJXeyrn+PtxS9WFwSnAAAAAAAAqrKgyt4AlE103VDLzs23VXH7bdHGPZaTm+/mL9uyzwWkgvyypwAAAAAAAKoqMqWqEWVI/bI/3QZ1jrV+HRrbt6t22DGtG9iOhHRbvyPJLZPv1ZDnAAAAAAAAVRuZUtWIMqRenb3GPv1pm2Vk59oJPZpb/cgwS0jNtD3JmZW9eQAAAAAAAKVGUKoaZks9/9kKe+Z/P7sMqeAgjx3TpqGNG9qxsjcNAAAAAACg1Oi+V02p697m3Sl21oC2dsaAtlYn+EB8McijLnyVvXUAAAAAAAAlIyhVjcXtTbUpX662PckZtnjTXjePIaUAAAAAAEB1QFCqmsvKzbd3528qeE1MCgAAAAAAVAeMKQUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgIAjKAUAAAAAAICAIygFAAAAAACAgCMoBQAAAAAAgNoZlJo4caK1a9fOwsPDbdCgQbZw4cISl3/vvfesW7dubvlevXrZjBkzAratAAAAAAAAqAFBqXfeecf++Mc/2v3332+LFy+2Pn362KhRo2z37t3FLj9//ny76KKLbPz48bZkyRI755xz3LRixYqAbzsAAAAAAACqaVDqn//8p1177bV21VVXWY8ePWzy5MlWt25de/XVV4td/l//+peNHj3aJkyYYN27d7eHH37Y+vXrZ88//3zAtx0AAAAAAADVMCiVnZ1tP/30k40cOfK3DQoKcq+///77Yj+j+f7LizKrDrU8AAAAAAAAqp46lfnL9+7da3l5eda0adNC8/V6zZo1xX5m586dxS6v+cXJyspyk09SUpL7Pzk52aqb1JRky81Mr5B1Z3syLdljFSY1z8wyK2bdedleS8nPtYqUnJpqYdWwzFRHFVnOK7qsV/dyLpT1wKmudXpFlnOhTq9ZqNOLRzmveajTi0dZr1mqazkX6vTA88VcvF5vyQt6K1F8fLy2zjt//vxC8ydMmOAdOHBgsZ8JCQnxTp06tdC8iRMnemNjY4td/v7773e/g4mJiYmJiYmJiYmJiYmJiYnJAjbFxcWVGBeq1Eypxo0bW3BwsO3atavQfL1u1qxZsZ/R/LIsf88997iB1H3y8/MtISHBGjVqZB5PBYZicVCUtHXr1hYXF2fR0dGVvTlAhaCcozagnKO2oKyjNqCco7agrAeeMqRSUlKsRYsWJS5XqUGp0NBQ69+/v82ePds9Qc8XNNLrW265pdjPDBkyxL1/xx13FMz74osv3PzihIWFuclf/fr1j+p+oPRUAVAJoKajnKM2oJyjtqCsozagnKO2oKwHVkxMzGGXqdSglCiL6YorrrABAwbYwIED7ZlnnrG0tDT3ND65/PLLrWXLlvb444+717fffrudeOKJ9tRTT9nYsWPt7bfftkWLFtlLL71UyXsCAAAAAACA0qr0oNS4ceNsz549dt9997nByo899libOXNmwWDm27Ztc0/k8zn++ONt6tSp9pe//MXuvfde69y5s02fPt169uxZiXsBAAAAAACAahWUEnXVO1R3vTlz5hw074ILLnATqg91obz//vsP6koJ1CSUc9QGlHPUFpR11AaUc9QWlPWqy6PRzit7IwAAAAAAAFC7/NYvDgAAAAAAAAgQglIAAAAAAAAIOIJSAGosj8fjHoRQkiuvvNLOOeccqy7b6++BBx5wD4dAzdSuXTv3RNryev31161+/fpW22lsSh1biYmJlb0pqML1KVBbzhdHem4BgKONoBTKrLhG/Pvvv2/h4eH21FNPufd1Mfi3v/2t0DK6ONT8og2FY445xvLy8gotq4aUGlSoXXxl54YbbjjovZtvvtm9p2XKY8uWLe7zS5cuLTT/X//6V5Upazt27LAxY8aU+/MKUhX3/WmfNV/fgf934ZtCQ0OtU6dO9sgjjxjDDJZeRQc0f/zxR7vuuuvK3cjQ023XrVtX6t930kknFZQJ1eddunSxxx9/vNqXCT21V8dWTExMZW9Kra3TNYWEhFj79u3trrvusszMTKsp/OtS3zRs2LBK3yYCclXzGNDTxU899VR79dVXLT8/v1LOF2VZtrzHgP+kaxOgouzZs8duvPFGa9OmjRvAvFmzZjZq1Cj75ptvrHHjxge1R30efvhhdzzm5OS419nZ2fbEE09Ynz59rG7duu6zQ4cOtddee61gGVQcglI4YlOmTLFLLrnEJk2aZHfeeaebpwbN3//+d9u/f/9hP79p0yZ78803A7ClqA5at25tb7/9tmVkZBTMUwNm6tSp7oRztKmhWlWySXQiPdIngujYe+WVV2z9+vWHXfbLL790jXUt++CDD9qjjz7qLpRRNTRp0sRdGJVXRESExcbGlukz1157rSsTa9eutXvuucfuu+8+mzx5slUkXQhWJAVddWz53xRB4IwePdqVKZ3rn376aXvxxRfd049qEjVatI++6eOPPy73umj81NxjQDeEPvvsMzv55JPt9ttvtzPOOMNyc3MDfr440nNLUf5lXzdHoqOjC83705/+VLCsbnIcrX0G5Pzzz7clS5bYG2+84W7Eqf7VTbakpCS79NJLXf1clMqhbkhffvnlLlis6xAFshTAUsB2/vz5tnDhQndD/LnnnrOVK1dWyr7VJgSlcEQUUb711ltdEOGqq64qmD9y5EjXCNBd9sPR53WBmpWVVcFbi+qgX79+LjA1bdq0gnn6WQGpvn37lpgZoq5sh7ojpzv0onWocaoTVnHZLpp/2223ubv5DRs2dOW46Dq3bdtmZ599ttWrV89dfP3+97+3Xbt2HdStTgEebbeWu+mmm1xGoI4ZrVPBAgWBSrq7fffdd7tsFV08dujQwf76178etsHStWtXd8H7f//3f3Y4jRo1ctvStm1bF1jWHaHFixcf9nM4PN2hGzhwoAsyNm/e3P785z8XuhBPSUlx33lkZKR7X411lb077rij2DKuCyiVK9+dwBYtWrhyKvrc1q1b7Q9/+EPBnelDdd/75JNP7LjjjnPBS90FPPfccwu9r7LmKxOq03v37m1ffPFFwfuqp9XAaNmypdv2QYMGuaxXfy+//LI7hrUurf+f//xnoe3wHR+6oaHjUtsi6l53zTXXuAaTjqsRI0bYsmXLCj6nn1W2o6Ki3Pv9+/e3RYsWufe0/2eeeaY1aNDAbZcycGfMmHHI7nsffPCBW0bfpb5nZfn607zHHnvMrr76avf79L2/9NJLZSgB8PHduVaZUF2r6wNfmdq3b59ddNFFrjypvPTq1cveeuutQp8vTZ2swPoJJ5zgylKPHj0KlVmfn3/+2ZUpBWtV96nhkZqaWvC+71ygv7vunqvMPvTQQ+64nTBhgvvdrVq1KraBo2W1Xb5Jy4oyYbQOfU7fg8r9zJkzCz7ny1p955137MQTT3Tb/9///te9p+Oje/fubl63bt3shRdeKPicGlC33HKLqzv0vo5X3/WWyq7o2NO6fa9R+ceAyrmuce6991776KOPXIDKl6l9uPrvcPV3ac8XRZctyzXNv//9b/dZ3cy78MIL3XlM/Mu+3lO5871es2aNq0O1r6qztT3fffedOzZUZnUO0DGp7BT1uvC3YsUKlz2u7dIxedlll9nevXuP+t8H1ZeOm7lz57pECF0fqC7UtZduqp111lk2fvx4F6hSmSt6jaYbJXpfdDx8++23Nnv2bBeIUnnXdffFF19sP/zwg3Xu3LmS9rD2ICiFclODWamP//vf/w5q2AQHB7sLO0WXt2/fXuJ61AjTRZ+WBUQNQf8LfwV3/IOe5aE7Hv7ZQf5Br6J0t0UNW52IFERSo8LXyNGFlC7eEhIS3ElN83ViU1cpfxs3bnQXYWqAqJGl7KWxY8e640Gf0wn0L3/5i/sdh6ILOV2wrlq1ynUzVGNfwYvD0Z0eNbp9DfbS0LI//fSTCzLgyMTHx9vpp5/uGg9qVCiLVH9/dY/0+eMf/2jz5s1zd/RUhnRRVVJAUH9PX5aJGuAKXqoBLyrLavSqnPruTBfn008/dXW1tk13FXXxpYu34qhRo21Sg0KZRj5qCH///ffuRsTy5cvtggsucFkAvsw87ZO6jyoLQN1G1U2laPBVNmzY4PZJ2+7rUqt17d692x03KotqvJ1yyinuWBMF8bSf6nqi9xXo0x1O0UWkAma6qFTwQceXGjLF0WfV6FKjSsuqwaWAb9FuvApUDRgwwH1XCiqre4AyyFB+amTqDrSvTCkLVg1VlU29p0CRGp6++rq0dfJ5553n1qn3ldmn6xN/aWlp7i64gpYqP++99547F6g8+/vqq6/sl19+ceVIwVTdMFM2iz6ndatsX3/99Ye9rvFRva1y9I9//MMdL9oGNZSKZrKqLOuYWb16tVtGgSllKerY0TxdT6mM6nuQZ5991tUd7777riuTWt4XfNL++Wdv+V6jalHQSYEY37XI4eq/stTfJZ0viirLNY3Wo2t+TVr2UN2iiqMyruVVnnWzQwEp9ZLQ8aosFN1UUVaL1usLNug70o1EXZ/oWkqBMtXdgI/O85pUNotLblC517VY0V4Aqh/VtV8Bf1Edqhsm/je/fXSdofMPKpgXKKMrrrjCGxoaqkFGvLNnzy72/bPPPtv9PHjwYO/VV1/tfv7www/dZ3y+/vpr93r//v3eyZMnexs2bOhNTEx078XExHhfe+21gO0TqgZf2dm9e7c3LCzMu2XLFjeFh4d79+zZ497TMtK2bVvv008/Xejzffr08d5///0Fr1W+VO5k8+bN7vWSJUuK/Z0+J554onfYsGGFljnuuOO8d999t/v5888/9wYHB3u3bdtW8P7KlSvduhcuXOheaxvq1q3rTU5OLlhm1KhR3nbt2nnz8vIK5nXt2tX7+OOPF7u9xXnyySe9/fv3L3it36N9Lu71hRde6B0xYoT7Wfusdes78P8uIiIivJGRkd6QkBD3+rrrrjvk78bBipYdn3vvvdf9bfPz8wvmTZw40VuvXj3391e50Hf+3nvvFbyvuk9l5vbbby+Y51/Gn3rqKW+XLl282dnZxW5LcceD6lDVpT5DhgzxXnLJJYfcH5V9bZd/mdCxN2/ePPf+1q1bXdmPj48v9LlTTjnFe88997ifx40b5x07dmyh9/U7/bdD5VTr13HuM3fuXG90dLQ3MzOz0Gc7duzoffHFF93PUVFR3tdff73Ybe/Vq5f3gQceKPY9/3ONXHzxxd5TTz210DITJkzw9ujRo9D3eemllxa81t8yNjbWO2nSpEN8ezjUMaIyozKlOl1/h6CgIO/7779/yM+o/Nx5552lrpNnzZrlrVOnTqFy+dlnnxWqT1966SVvgwYNvKmpqQXLfPrpp25bdu7cWbCt+rsXraOHDx9e8Do3N9fty1tvvVUwz3ecaL5v8v3eFi1aeB999NGDtv2mm24qVBc/88wzB5X7qVOnFpr38MMPu2NYbr31Vle/+9cx/g53LkHlnyd89WX37t1LVf8drv4u7/mivNc0qjMHDRp00LqLnnd89e/06dML5mk/tb758+cX+uz48eO9F110UUF5P+200wq9HxcX59a1du3aQ34PqH10PlH9rnr4+OOPd9cjy5YtK3hfbUxdf6WkpLjXKscqf1OmTClYRtfDt912W6VsPw4gUwrlorscuiunu4j+6e9F6W617uzpzkhJlD6pdHotDyh9XVlFylzQ3Qz9rFT1QJZvf+oioTuYorKsbiiafNRdRN03/Mu5jg9lOvko9VzLBQUFFZrnW29x1KVDXeqUAq87QcqsUpp9aSgrR5kun3/+eYnrV5aKsnl0x13dCXQ3E0dG5WDIkCGFxjDS31F1pTIsdBda3TD973Kry4O6Xh6K7qJrnDWlk2vcpw8//LDM43Lob6077yVRNpKWU8aTuk2oG6juJoqyitQFVV1KfXcnNenOtu6ii7I2it69L+5uvlLsdZz7qAzq+9F5wH/dmzdvLli3ssvUvUV3M3XH3Tdf1DVFZV7fs85Lykop6e+j5fzptTIK/B+64V8P+LqjlHS8onjqUqEypUyjK664wmW9agwQ0fetjGvdzVaXN/3NZ82adVA9V5o6WV2UfHT8+dMyykrxv9utv7myRPyz39Sls2gd7Z9hoixwldGi5UBZKdpH36QMweTkZJd1VVxZK3pNpIw8/6wulW1dF/kfCyrfvjKvrob6PaozVPZLqudRdSl+qLqlNPVfaerv8pwvyntN438MloZ/GVembHp6ujtO/PdXmVO+/dV38vXXXxd635fV4l/3AzqfqK5V9qgyt9VlX5mGvuxndRHXuUbXub5rX9Xz/tmA1f2BLjVBncreAFRP6hevvt+62FQFoHRj/5OVj8Z4UCq6+vaW9NS0OnXquDR1LVM0nR61twufryxMnDjxoPd1Qil6EjlaA8T6ugT56KKxrE/JKW4dZVmvukgpQKAByHUMKWihLlNFx745lI4dO7qLUQWZ1HWsOLoI1VP3RGOX6EJPXUTUnck3zg+qBv2t1HhWlyN1r1B3sieffNIFhIqWq0PRuB2Ho3LmKxO6gNPPgwcPdoEgNZrUKFfXEv3v71Bd5Q6laCq81q1GTtHxqcQ3HpXKpcZ3UDcWnXMUfNIxoS4tClbpONF7aqCra4iOFY1ZWJn1AA78rX1lSl0oFBxSnaSgi8qwurhpPA8Ff7SsuvQXHfw+UH+L8tbbClj69tFHQanyHA++G33qrl20O7XvuFODSwELHQeqE9SlScdo0TF5ULUp6KMxlUpT/5Wm/j6a54uijvQYLK6Mq75We8Kf72EvWkbjBBZ3s1rfFeBP16wKcmrSdayuCXSNoHalxkn73e9+525y+4YHUZ3pf92im20argCVh0wplJvudOsEt3PnTheY8g14WJTuaGtwRjWyD3dnR3cp1QgHVKbUMFGgSY3NopRl4T92jhoAukg/FN8YJv6ZEOWh4E1cXJybfDTmk8Y/0N3Fo0XjrugYU6aK7jBqkEUN5lwWGpNEAzyq4V4aavDobmpFPw2tplMZUX3nHzRV5pEC9xoTSXevdYHvP9aLnhKjv1VJ1CjRRbrGk1HjRb9D2Uu+8n24sq1sE41DUlq6YNM4NxrYXPuisRb0O3R3XA1w/0mNclHmRtExbEozpo0a2TqX6AZF0XX7Z0nqwlFjjyjwpHGE/MeeU0NMY/5ojBY9CVaN+kP9ffT38KfXWnfRYBuOLt1M0CDPyvpUJoe+d41no7FkFKzSsXG44+BQdbL/+WDBggUHLaPMC2Uh+eh3a3tKylA8EmoIKXuruLJW0rlC2Vn6nDIqix4Lvgd2+NavO/0q57rzr3GEfOMPqX450nMdKpbGL1P9rSyP0tR/Za2/SzpfVMY1jT+tV8EnZUQW3V9fxpa+E401pQytosswvg9KU8b863vdBNFg5xoPTdfXvgHOfXTDS0FcjddWlNoh/utCxSAohSOik4dOdmqkKHBQ3J1B3f1UxodOjIejAJbupHLwQ41D3UXUxVFxDUUNgKknwaiLmi601C2kpAalnnanizTfYJkKApSH7kb7yrQGptaAvHqkrJ6c5J+efqQUhNIFmwJKymDS8aMU/LJQ40Zdng517OnJV7oQVpcy3XFXxoKyH9XYQemoHPl329GkwZp1ga8sHd15U7dI3bHT30KNYAWnVF71RC91T9CFty6Q9J5/lz9/SkNXdokGg1Zj9T//+Y8rzwpcii7cNTizBlk/1NOJtA0adF//69jyDQheEg3qrCCBGrwK2qjcq7wr8KMgsMq/spJ0x1u0z3rqnQaJVnc4DbSrsnWo/fI/rtTlSk8/U8BJTyXThaOCshrkVgEMZU7qfKPgrBr2CnapQSXKrlG3L22Tjkt9r773ilLASo07dRvTvqmL+fPPP1/oseWoOLoBpbpaGbCq55TJob+1yqTKm/9Tv0pDZUdlU8eUAk86JxR9+qjKre6kaxkdQyofKqsaVF31ZEXRMa5jTEEjZa4oc1V1hIK9JdHNOR1XqrtVRnWsKgCr40r0v45l1S96XwO3KzDsy6pRfaAyrvp9//79FbZ/KB0NwKy/hepn1U8auF7BWA2ir/r0cPVfWevvw50vKuOaxp/OgapvdYNB9a+ucfS79cAj32D+eniFgqzqeqW6Xsuojlf3XwKu8L+OVXtAZVzd9nUNoPpQD8TQMebfc0cBTZVtdQP1DUvgo2sIda1WF1mdm3Qu0bGjjHFlixd9OAWOPoJSOGK686+GghpChwpM6Uk5pUnzVcWiqaxjpaBmUnDkUAESdQnVRZMu6jTmlC7m1GXtUHQHUhf4aiTrLrT/yaos1LhWkEFPY9JJThd0uruvRsfRpCc06YJNDXE9mlYXqEpJLitd+B2qa5W2XWnwasAokKKn+hzt/ajpVPcpg8h/UrBDgRld3Cv7Q9k7CjopO8RHjUo1QlR+9XfQxZDv8e/FUWNTGRFaTnfMdUdPGagag8RXx6oho2PAf6wmfyeddJK7WNO4CypTqmuLPuWsKI3zo4s4dZ1THa6GsV4rsKMMEx13ajDo0eOi7dPTlLR/2ncFgVWOD9cdVMeVvjMdU2p0KMigp+MpAKWggYIYuvjU79Z7Sr3XmFe+zFo1UtSI0XeoLEst88ILLxT7u3QHXheaCvj27NnTZRTq+yupizmOHtXFqtfUaFA50t9D1w4qnwqsqEyVhYK5CtgrcKnxy9Rto+gTH+vWresatGrk6klM6sqhxoeCkRVJ4z0pGK39VMNfx4OOv8M9Xlz7MGXKFHe86XM61ynQ4MuUUqNe35+CBtofHfs6fnzjYanrqoJ9unFY3NOkEFj6u/vOtaqfFBTV9YiuJVS3Ha7+K2v9fbjzRWVc0xSl86SuaRR89dXburnhK+O+LEPV7aeddpo7DhQ40L75j/uG2k3Xt+rmrHH9VH51Tle50vAV/vW7yrm67ilIr/+LUuae6sy77rrLtRMUiFLdquNU9bjWi4rl0WjnFfw7AABACZQdqrE11JgsmlZe3eniUBkdymABAAAA/DHQOQAAAaZxCxSoUWaHugAqU0fKm8FXlfzjH/9wg41q3A913VN3jENlLQEAAKB2IygFAEAlBW80zowGKe/fv7/LJPIf1Lu6UpcSdS3Swy/UDUTp7+qOBAAAABRF9z0AAAAAAAAEHCPFAQAAAAAAIOAISgEAAAAAACDgCEoBAAAAAAAg4AhKAQAAAAAAIOAISgEAAAAAACDgCEoBAABUQ3PmzDGPx2OJiYml/ky7du3smWeeqdDtAgAAKC2CUgAAABXgyiuvdEGjG2644aD3br75ZveelgEAAKitCEoBAABUkNatW9vbb79tGRkZBfMyMzNt6tSp1qZNm0rdNgAAgMpGUAoAAKCC9OvXzwWmpk2bVjBPPysg1bdv34J5WVlZdtttt1lsbKyFh4fbsGHD7Mcffyy0rhkzZliXLl0sIiLCTj75ZNuyZctBv++7776z4cOHu2X0e7XOtLS0Ct5LAACA8iEoBQAAUIGuvvpqe+211wpev/rqq3bVVVcVWuauu+6yDz74wN544w1bvHixderUyUaNGmUJCQnu/bi4ODvvvPPszDPPtKVLl9o111xjf/7znwutY+PGjTZ69Gg7//zzbfny5fbOO++4INUtt9wSoD0FAAAoG4JSAAAAFejSSy91waGtW7e6ad68eW6ejzKZJk2aZE8++aSNGTPGevToYS+//LLLdnrllVfcMnq/Y8eO9tRTT1nXrl3tkksuOWg8qscff9zNv+OOO6xz5852/PHH27PPPmtvvvmm6zIIAABQ1dSp7A0AAACoyZo0aWJjx461119/3bxer/u5cePGhTKccnJybOjQoQXzQkJCbODAgbZ69Wr3Wv8PGjSo0HqHDBlS6PWyZctchtR///vfgnn6ffn5+bZ582br3r17Be4lAABA2RGUAgAACEAXPl83uokTJ1bI70hNTbXrr7/ejSNVFIOqAwCAqoigFAAAQAXTWE/Z2dnm8XjcWFH+1C0vNDTUdetr27atm6fMKQ10rq54oiynjz/+uNDnFixYcNCg6qtWrXLjUQEAAFQHjCkFAABQwYKDg10XPAWN9LO/yMhIu/HGG23ChAk2c+ZMt8y1115r6enpNn78eLfMDTfcYOvXr3fLrF271qZOneq6A/q7++67bf78+S4jS4Oha/mPPvqIgc4BAECVRVAKAAAgAKKjo91UnL/97W/uqXmXXXaZy3jasGGDzZo1yxo0aFDQ/U5P55s+fbr16dPHJk+ebI899lihdfTu3du++eYbW7dunQ0fPtz69u1r9913n7Vo0SIg+wcAAFBWHq9GwAQAAAAAAAACiEwpAAAAAAAABBxBKQAAAAAAAAQcQSkAAAAAAAAEHEEpAAAAAAAABBxBKQAAAAAAAAQcQSkAAAAAAAAEHEEpAAAAAAAABBxBKQAAAAAAAAQcQSkAAAAAAAAEHEEpAAAAAAAABBxBKQAAAAAAAAQcQSkAAAAAAABYoP0/SRvkzG7cInwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1200x700 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "import pandas as pd\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "# Define models\n",
    "models = {\n",
    "    'KNN': KNeighborsClassifier(n_neighbors=60),\n",
    "    'MultinomialNB': MultinomialNB(alpha=1),\n",
    "    'LogisticRegression': LogisticRegression(penalty='l2',C=10,class_weight='balanced',random_state=1,solver='lbfgs',n_jobs=-1),\n",
    "    'RandomForest': RandomForestClassifier(n_estimators=100, random_state=42),\n",
    "    'DecisionTree': DecisionTreeClassifier(random_state=42),\n",
    "    'SVC': SVC(kernel='linear', C=1, random_state=42)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "results = []\n",
    "for name, model in models.items():\n",
    "    model.fit(x_train_res, y_train_res)\n",
    "    y_pred = model.predict(merge_df_testing)\n",
    "    results.append({\n",
    "        'Model': name,\n",
    "        'Accuracy': accuracy_score(y_test, y_pred),\n",
    "        'Precision': precision_score(y_test, y_pred, average='weighted', zero_division=1),\n",
    "        'Recall': recall_score(y_test, y_pred, average='weighted', zero_division=1),\n",
    "        'F1-Score': f1_score(y_test, y_pred, average='weighted', zero_division=1)\n",
    "    })\n",
    "\n",
    "# Convert to DataFrame\n",
    "df_results = pd.DataFrame(results).set_index('Model')\n",
    "\n",
    "# Plotting\n",
    "ax = df_results.plot(kind='bar', figsize=(12,7), legend=False)\n",
    "plt.title(\" Model Evaluation\")\n",
    "plt.ylabel(\"Score\")\n",
    "plt.ylim(0, 1)\n",
    "plt.xticks(rotation=0)\n",
    "\n",
    "# Bar color mapping to metric names\n",
    "colors = {\n",
    "    'Accuracy': 'steelblue',\n",
    "    'Precision': 'darkorange',\n",
    "    'Recall': 'green',\n",
    "    'F1-Score': 'crimson'\n",
    "}\n",
    "\n",
    "# Manually set colors to bars + add metric label inside each bar\n",
    "for bars, metric_name in zip(ax.containers, df_results.columns):\n",
    "    for bar in bars:\n",
    "        height = bar.get_height()\n",
    "        ax.text(bar.get_x() + bar.get_width() / 2, height + 0.01, f'{height:.2f}',\n",
    "                ha='center', va='bottom', fontsize=10, fontweight='bold',rotation =80)\n",
    "        ax.text(bar.get_x() + bar.get_width() / 2, height / 2, metric_name,\n",
    "                ha='center', va='center', color='white', fontsize=9, fontweight='bold',rotation =75)\n",
    "        bar.set_color(colors[metric_name])  # Set custom color\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "1c242e9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting prettytable\n",
      "  Downloading prettytable-3.16.0-py3-none-any.whl.metadata (33 kB)\n",
      "Requirement already satisfied: wcwidth in ./venv/lib/python3.12/site-packages (from prettytable) (0.2.14)\n",
      "Downloading prettytable-3.16.0-py3-none-any.whl (33 kB)\n",
      "Installing collected packages: prettytable\n",
      "Successfully installed prettytable-3.16.0\n",
      "+--------------------+----------+-----------+--------+----------+\n",
      "|       Model        | Accuracy | Precision | Recall | F1-Score |\n",
      "+--------------------+----------+-----------+--------+----------+\n",
      "|        KNN         |   0.3    |    0.53   |  0.3   |   0.32   |\n",
      "|   MultinomialNB    |   0.51   |    0.54   |  0.51  |   0.51   |\n",
      "| LogisticRegression |   0.6    |    0.62   |  0.6   |   0.61   |\n",
      "|    RandomForest    |   0.62   |    0.63   |  0.62  |   0.62   |\n",
      "|    DecisionTree    |   0.52   |    0.54   |  0.52  |   0.52   |\n",
      "|        SVC         |   0.58   |    0.62   |  0.58  |   0.59   |\n",
      "+--------------------+----------+-----------+--------+----------+\n",
      "The Best model is :  RandomForest\n"
     ]
    }
   ],
   "source": [
    "!pip install prettytable\n",
    "from prettytable import PrettyTable\n",
    "\n",
    "# Create PrettyTable\n",
    "table = PrettyTable()\n",
    "table.field_names = [\"Model\", \"Accuracy\", \"Precision\", \"Recall\", \"F1-Score\"]\n",
    "\n",
    "# Add rows from df_results\n",
    "for model_name, row in df_results.iterrows():\n",
    "    accuracy = round(row['Accuracy'], 2)\n",
    "    precision = round(row['Precision'], 2)\n",
    "    recall = round(row['Recall'], 2)\n",
    "    f1 = round(row['F1-Score'], 2)\n",
    "\n",
    "    table.add_row([model_name, accuracy, precision, recall, f1])\n",
    "\n",
    "print(table)\n",
    "best_model = df_results['F1-Score'].idxmax()\n",
    "print(\"The Best model is : \",best_model)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b5eeb058",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create pckle file for Rndom forest model\n",
    "import pickle\n",
    "# Save the model to a file\n",
    "with open('rf_model.pkl', 'wb') as file:\n",
    "    pickle.dump(rf, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "68245f9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['abl1' 'acvr1' 'ago2' 'akt1' 'akt2' 'akt3' 'alk' 'apc' 'ar' 'araf'\n",
      " 'arid1a' 'arid1b' 'arid2' 'asxl1' 'asxl2' 'atm' 'atr' 'atrx' 'aurka'\n",
      " 'aurkb' 'axin1' 'axl' 'b2m' 'bap1' 'bard1' 'bcl10' 'bcl2' 'bcl2l11'\n",
      " 'bcor' 'braf' 'brca1' 'brca2' 'brd4' 'brip1' 'btk' 'card11' 'carm1'\n",
      " 'casp8' 'cbl' 'ccnd1' 'ccnd2' 'ccnd3' 'ccne1' 'cdh1' 'cdk12' 'cdk4'\n",
      " 'cdk6' 'cdk8' 'cdkn1a' 'cdkn1b' 'cdkn2a' 'cdkn2b' 'cdkn2c' 'cebpa'\n",
      " 'chek2' 'cic' 'crebbp' 'ctcf' 'ctla4' 'ctnnb1' 'ddr2' 'dicer1' 'dnmt3a'\n",
      " 'dnmt3b' 'dusp4' 'egfr' 'eif1ax' 'elf3' 'ep300' 'epas1' 'epcam' 'erbb2'\n",
      " 'erbb3' 'erbb4' 'ercc2' 'ercc3' 'ercc4' 'erg' 'errfi1' 'esr1' 'etv1'\n",
      " 'etv6' 'ewsr1' 'ezh2' 'fam58a' 'fanca' 'fancc' 'fat1' 'fbxw7' 'fgf19'\n",
      " 'fgf3' 'fgf4' 'fgfr1' 'fgfr2' 'fgfr3' 'fgfr4' 'flt3' 'foxa1' 'foxl2'\n",
      " 'foxo1' 'foxp1' 'fubp1' 'gata3' 'gli1' 'gna11' 'gnaq' 'gnas' 'h3f3a'\n",
      " 'hist1h1c' 'hla-a' 'hla-b' 'hnf1a' 'hras' 'idh1' 'idh2' 'igf1r' 'ikbke'\n",
      " 'ikzf1' 'il7r' 'inpp4b' 'jak1' 'jak2' 'jun' 'kdm5a' 'kdm5c' 'kdm6a' 'kdr'\n",
      " 'keap1' 'kit' 'klf4' 'kmt2a' 'kmt2b' 'kmt2c' 'kmt2d' 'knstrn' 'kras'\n",
      " 'lats1' 'lats2' 'map2k1' 'map2k2' 'map2k4' 'map3k1' 'mapk1' 'mdm2' 'mdm4'\n",
      " 'med12' 'mef2b' 'men1' 'met' 'mga' 'mlh1' 'mpl' 'msh2' 'msh6' 'mtor'\n",
      " 'myc' 'mycn' 'myd88' 'myod1' 'ncor1' 'nf1' 'nf2' 'nfe2l2' 'nfkbia'\n",
      " 'nkx2-1' 'notch1' 'notch2' 'npm1' 'nras' 'nsd1' 'ntrk1' 'ntrk2' 'ntrk3'\n",
      " 'nup93' 'pak1' 'pax8' 'pbrm1' 'pdgfra' 'pdgfrb' 'pik3ca' 'pik3cb'\n",
      " 'pik3cd' 'pik3r1' 'pik3r2' 'pik3r3' 'pim1' 'pms1' 'pms2' 'pole' 'ppm1d'\n",
      " 'ppp2r1a' 'ppp6c' 'prdm1' 'ptch1' 'pten' 'ptpn11' 'ptprd' 'ptprt' 'rab35'\n",
      " 'rac1' 'rad21' 'rad50' 'rad51b' 'rad51c' 'rad51d' 'rad54l' 'raf1' 'rara'\n",
      " 'rasa1' 'rb1' 'rbm10' 'ret' 'rheb' 'rhoa' 'rictor' 'rit1' 'rnf43' 'ros1'\n",
      " 'rras2' 'runx1' 'rxra' 'rybp' 'sdhb' 'sdhc' 'setd2' 'sf3b1' 'shoc2'\n",
      " 'shq1' 'smad2' 'smad3' 'smad4' 'smarca4' 'smarcb1' 'smo' 'sos1' 'sox9'\n",
      " 'spop' 'src' 'srsf2' 'stag2' 'stat3' 'stk11' 'tcf3' 'tcf7l2' 'tert'\n",
      " 'tet1' 'tet2' 'tgfbr1' 'tgfbr2' 'tmprss2' 'tp53' 'tp53bp1' 'tsc1' 'tsc2'\n",
      " 'u2af1' 'vegfa' 'vhl' 'whsc1' 'whsc1l1' 'xpo1' 'xrcc2' 'yap1']\n"
     ]
    }
   ],
   "source": [
    "print(le_gene.classes_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "15efe3d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1_2009trunc' '2010_2471trunc' '256_286trunc' ... 'yap1-mamld1_fusion'\n",
      " 'zc3h7b-bcor_fusion' 'znf198-fgfr1_fusion']\n"
     ]
    }
   ],
   "source": [
    "print(le_var.classes_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3ad211a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:14: SyntaxWarning: invalid escape sequence '\\|'\n",
      "<>:14: SyntaxWarning: invalid escape sequence '\\|'\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/2222957323.py:14: SyntaxWarning: invalid escape sequence '\\|'\n",
      "  sep=\"\\|\\|\",\n"
     ]
    }
   ],
   "source": [
    "# main pipeline \n",
    "\n",
    "import pandas as pd\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.compose import ColumnTransformer\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "import pickle\n",
    "\n",
    "# Load and merge training data\n",
    "train_var = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_variants\")\n",
    "train_text = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_text\",\n",
    "                          sep=\"\\|\\|\",\n",
    "                            engine=\"python\",\n",
    "                              names=[\"ID\",\"Text\"],\n",
    "                                skiprows=1)\n",
    "df = pd.merge(train_var, train_text, on=\"ID\")\n",
    "\n",
    "# Clean columns\n",
    "df['Text'] = df['Text'].str.lower()\n",
    "df['Gene'] = df['Gene'].str.lower().str.replace(' ', '_')\n",
    "df['Variation'] = df['Variation'].str.lower().str.replace(' ', '_')\n",
    "df.dropna(inplace=True)\n",
    "\n",
    "# Optional: Clean text further (lemmatization, stopwords removal, etc.)\n",
    "\n",
    "# Final feature set\n",
    "df = df[[\"Gene\", \"Variation\", \"Text\", \"Class\"]]\n",
    "\n",
    "# Define preprocessing pipeline\n",
    "preprocessor = ColumnTransformer(transformers=[\n",
    "    (\"gene_enc\", OneHotEncoder(handle_unknown=\"ignore\"), [\"Gene\"]),\n",
    "    (\"var_enc\", OneHotEncoder(handle_unknown=\"ignore\"), [\"Variation\"]),\n",
    "    (\"text_vec\", TfidfVectorizer(max_features=5000), \"Text\")\n",
    "])\n",
    "\n",
    "# Full pipeline\n",
    "pipeline = Pipeline(steps=[\n",
    "    (\"preprocess\", preprocessor),\n",
    "    (\"model\", RandomForestClassifier())\n",
    "])\n",
    "\n",
    "# Train pipeline\n",
    "pipeline.fit(df[[\"Gene\", \"Variation\", \"Text\"]], df[\"Class\"])\n",
    "\n",
    "# Save pipeline\n",
    "with open(\"final_pipeline.pkl\", \"wb\") as f:\n",
    "    pickle.dump(pipeline, f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f5513222",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:18: SyntaxWarning: invalid escape sequence '\\|'\n",
      "<>:18: SyntaxWarning: invalid escape sequence '\\|'\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/1066002218.py:18: SyntaxWarning: invalid escape sequence '\\|'\n",
      "  sep=\"\\|\\|\",\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random Forest Accuracy with SMOTE: 0.6280\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[75]\u001b[39m\u001b[32m, line 65\u001b[39m\n\u001b[32m     59\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m name, model \u001b[38;5;129;01min\u001b[39;00m models.items():\n\u001b[32m     60\u001b[39m     pipeline = ImbPipeline([\n\u001b[32m     61\u001b[39m         (\u001b[33m\"\u001b[39m\u001b[33mpreprocess\u001b[39m\u001b[33m\"\u001b[39m, preprocessor),\n\u001b[32m     62\u001b[39m         (\u001b[33m\"\u001b[39m\u001b[33msmote\u001b[39m\u001b[33m\"\u001b[39m, SMOTE(random_state=\u001b[32m42\u001b[39m)),\n\u001b[32m     63\u001b[39m         (\u001b[33m\"\u001b[39m\u001b[33mmodel\u001b[39m\u001b[33m\"\u001b[39m, model)\n\u001b[32m     64\u001b[39m     ])\n\u001b[32m---> \u001b[39m\u001b[32m65\u001b[39m     \u001b[43mpipeline\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX_train\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     66\u001b[39m     y_pred = pipeline.predict(X_test)\n\u001b[32m     67\u001b[39m     acc = accuracy_score(y_test, y_pred)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/base.py:1365\u001b[39m, in \u001b[36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(estimator, *args, **kwargs)\u001b[39m\n\u001b[32m   1358\u001b[39m     estimator._validate_params()\n\u001b[32m   1360\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[32m   1361\u001b[39m     skip_parameter_validation=(\n\u001b[32m   1362\u001b[39m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[32m   1363\u001b[39m     )\n\u001b[32m   1364\u001b[39m ):\n\u001b[32m-> \u001b[39m\u001b[32m1365\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/imblearn/pipeline.py:526\u001b[39m, in \u001b[36mPipeline.fit\u001b[39m\u001b[34m(self, X, y, **params)\u001b[39m\n\u001b[32m    520\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m._final_estimator != \u001b[33m\"\u001b[39m\u001b[33mpassthrough\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m    521\u001b[39m         last_step_params = \u001b[38;5;28mself\u001b[39m._get_metadata_for_step(\n\u001b[32m    522\u001b[39m             step_idx=\u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m) - \u001b[32m1\u001b[39m,\n\u001b[32m    523\u001b[39m             step_params=routed_params[\u001b[38;5;28mself\u001b[39m.steps[-\u001b[32m1\u001b[39m][\u001b[32m0\u001b[39m]],\n\u001b[32m    524\u001b[39m             all_params=params,\n\u001b[32m    525\u001b[39m         )\n\u001b[32m--> \u001b[39m\u001b[32m526\u001b[39m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_final_estimator\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mXt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43myt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mlast_step_params\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mfit\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    527\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/base.py:1365\u001b[39m, in \u001b[36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(estimator, *args, **kwargs)\u001b[39m\n\u001b[32m   1358\u001b[39m     estimator._validate_params()\n\u001b[32m   1360\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[32m   1361\u001b[39m     skip_parameter_validation=(\n\u001b[32m   1362\u001b[39m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[32m   1363\u001b[39m     )\n\u001b[32m   1364\u001b[39m ):\n\u001b[32m-> \u001b[39m\u001b[32m1365\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/ensemble/_gb.py:787\u001b[39m, in \u001b[36mBaseGradientBoosting.fit\u001b[39m\u001b[34m(self, X, y, sample_weight, monitor)\u001b[39m\n\u001b[32m    784\u001b[39m     \u001b[38;5;28mself\u001b[39m._resize_state()\n\u001b[32m    786\u001b[39m \u001b[38;5;66;03m# fit the boosting stages\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m787\u001b[39m n_stages = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_fit_stages\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    788\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX_train\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    789\u001b[39m \u001b[43m    \u001b[49m\u001b[43my_train\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    790\u001b[39m \u001b[43m    \u001b[49m\u001b[43mraw_predictions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    791\u001b[39m \u001b[43m    \u001b[49m\u001b[43msample_weight_train\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    792\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_rng\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    793\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX_val\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    794\u001b[39m \u001b[43m    \u001b[49m\u001b[43my_val\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    795\u001b[39m \u001b[43m    \u001b[49m\u001b[43msample_weight_val\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    796\u001b[39m \u001b[43m    \u001b[49m\u001b[43mbegin_at_stage\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    797\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmonitor\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    798\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    800\u001b[39m \u001b[38;5;66;03m# change shape of arrays after fit (early-stopping or additional ests)\u001b[39;00m\n\u001b[32m    801\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m n_stages != \u001b[38;5;28mself\u001b[39m.estimators_.shape[\u001b[32m0\u001b[39m]:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/ensemble/_gb.py:883\u001b[39m, in \u001b[36mBaseGradientBoosting._fit_stages\u001b[39m\u001b[34m(self, X, y, raw_predictions, sample_weight, random_state, X_val, y_val, sample_weight_val, begin_at_stage, monitor)\u001b[39m\n\u001b[32m    876\u001b[39m         initial_loss = factor * \u001b[38;5;28mself\u001b[39m._loss(\n\u001b[32m    877\u001b[39m             y_true=y_oob_masked,\n\u001b[32m    878\u001b[39m             raw_prediction=raw_predictions[~sample_mask],\n\u001b[32m    879\u001b[39m             sample_weight=sample_weight_oob_masked,\n\u001b[32m    880\u001b[39m         )\n\u001b[32m    882\u001b[39m \u001b[38;5;66;03m# fit next stage of trees\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m883\u001b[39m raw_predictions = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_fit_stage\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    884\u001b[39m \u001b[43m    \u001b[49m\u001b[43mi\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    885\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    886\u001b[39m \u001b[43m    \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    887\u001b[39m \u001b[43m    \u001b[49m\u001b[43mraw_predictions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    888\u001b[39m \u001b[43m    \u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    889\u001b[39m \u001b[43m    \u001b[49m\u001b[43msample_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    890\u001b[39m \u001b[43m    \u001b[49m\u001b[43mrandom_state\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    891\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX_csc\u001b[49m\u001b[43m=\u001b[49m\u001b[43mX_csc\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    892\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX_csr\u001b[49m\u001b[43m=\u001b[49m\u001b[43mX_csr\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    893\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    895\u001b[39m \u001b[38;5;66;03m# track loss\u001b[39;00m\n\u001b[32m    896\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m do_oob:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/ensemble/_gb.py:489\u001b[39m, in \u001b[36mBaseGradientBoosting._fit_stage\u001b[39m\u001b[34m(self, i, X, y, raw_predictions, sample_weight, sample_mask, random_state, X_csc, X_csr)\u001b[39m\n\u001b[32m    486\u001b[39m     sample_weight = sample_weight * sample_mask.astype(np.float64)\n\u001b[32m    488\u001b[39m X = X_csc \u001b[38;5;28;01mif\u001b[39;00m X_csc \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m X\n\u001b[32m--> \u001b[39m\u001b[32m489\u001b[39m \u001b[43mtree\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    490\u001b[39m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneg_g_view\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m=\u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_input\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\n\u001b[32m    491\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    493\u001b[39m \u001b[38;5;66;03m# update tree leaves\u001b[39;00m\n\u001b[32m    494\u001b[39m X_for_tree_update = X_csr \u001b[38;5;28;01mif\u001b[39;00m X_csr \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m X\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/base.py:1365\u001b[39m, in \u001b[36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(estimator, *args, **kwargs)\u001b[39m\n\u001b[32m   1358\u001b[39m     estimator._validate_params()\n\u001b[32m   1360\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m config_context(\n\u001b[32m   1361\u001b[39m     skip_parameter_validation=(\n\u001b[32m   1362\u001b[39m         prefer_skip_nested_validation \u001b[38;5;129;01mor\u001b[39;00m global_skip_validation\n\u001b[32m   1363\u001b[39m     )\n\u001b[32m   1364\u001b[39m ):\n\u001b[32m-> \u001b[39m\u001b[32m1365\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfit_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mestimator\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/tree/_classes.py:1404\u001b[39m, in \u001b[36mDecisionTreeRegressor.fit\u001b[39m\u001b[34m(self, X, y, sample_weight, check_input)\u001b[39m\n\u001b[32m   1374\u001b[39m \u001b[38;5;129m@_fit_context\u001b[39m(prefer_skip_nested_validation=\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[32m   1375\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mfit\u001b[39m(\u001b[38;5;28mself\u001b[39m, X, y, sample_weight=\u001b[38;5;28;01mNone\u001b[39;00m, check_input=\u001b[38;5;28;01mTrue\u001b[39;00m):\n\u001b[32m   1376\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"Build a decision tree regressor from the training set (X, y).\u001b[39;00m\n\u001b[32m   1377\u001b[39m \n\u001b[32m   1378\u001b[39m \u001b[33;03m    Parameters\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m   1401\u001b[39m \u001b[33;03m        Fitted estimator.\u001b[39;00m\n\u001b[32m   1402\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1404\u001b[39m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_fit\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1405\u001b[39m \u001b[43m        \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1406\u001b[39m \u001b[43m        \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1407\u001b[39m \u001b[43m        \u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m=\u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1408\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcheck_input\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcheck_input\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1409\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1410\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/tree/_classes.py:472\u001b[39m, in \u001b[36mBaseDecisionTree._fit\u001b[39m\u001b[34m(self, X, y, sample_weight, check_input, missing_values_in_feature_mask)\u001b[39m\n\u001b[32m    461\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    462\u001b[39m     builder = BestFirstTreeBuilder(\n\u001b[32m    463\u001b[39m         splitter,\n\u001b[32m    464\u001b[39m         min_samples_split,\n\u001b[32m   (...)\u001b[39m\u001b[32m    469\u001b[39m         \u001b[38;5;28mself\u001b[39m.min_impurity_decrease,\n\u001b[32m    470\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m472\u001b[39m \u001b[43mbuilder\u001b[49m\u001b[43m.\u001b[49m\u001b[43mbuild\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mtree_\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msample_weight\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmissing_values_in_feature_mask\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    474\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.n_outputs_ == \u001b[32m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m is_classifier(\u001b[38;5;28mself\u001b[39m):\n\u001b[32m    475\u001b[39m     \u001b[38;5;28mself\u001b[39m.n_classes_ = \u001b[38;5;28mself\u001b[39m.n_classes_[\u001b[32m0\u001b[39m]\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from sklearn.compose import ColumnTransformer\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.model_selection import train_test_split\n",
    "from imblearn.pipeline import Pipeline as ImbPipeline\n",
    "from imblearn.over_sampling import SMOTE\n",
    "\n",
    "# Load and merge training data\n",
    "train_var = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_variants\")\n",
    "train_text = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_text\",\n",
    "                          sep=\"\\|\\|\",\n",
    "                            engine=\"python\",\n",
    "                              names=[\"ID\",\"Text\"],\n",
    "                                skiprows=1)\n",
    "df = pd.merge(train_var, train_text, on=\"ID\")\n",
    "\n",
    "# Clean categorical features\n",
    "df['Gene'] = df['Gene'].str.lower().str.replace(' ', '_')\n",
    "df['Variation'] = df['Variation'].str.lower().str.replace(' ', '_')\n",
    "\n",
    "# Clean text\n",
    "def clean_text(text):\n",
    "    text = str(text).lower()\n",
    "    text = re.sub(r'\\W+', ' ', text)\n",
    "    return text\n",
    "\n",
    "df.dropna(subset=[\"Text\"], inplace=True)\n",
    "df['Text'] = df['Text'].apply(clean_text)\n",
    "\n",
    "# Split features and target\n",
    "X = df[[\"Gene\", \"Variation\", \"Text\"]]\n",
    "y = df[\"Class\"]\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Preprocessor\n",
    "preprocessor = ColumnTransformer([\n",
    "    (\"gene_enc\", OneHotEncoder(handle_unknown=\"ignore\"), [\"Gene\"]),\n",
    "    (\"var_enc\", OneHotEncoder(handle_unknown=\"ignore\"), [\"Variation\"]),\n",
    "    (\"text_vec\", TfidfVectorizer(max_features=5000, stop_words='english'), \"Text\")\n",
    "])\n",
    "\n",
    "# Models to test\n",
    "models = {\n",
    "    \"Random Forest\": RandomForestClassifier(n_jobs=-1, class_weight='balanced'),\n",
    "    \"Gradient Boosting\": GradientBoostingClassifier(),\n",
    "    \"Logistic Regression\": LogisticRegression(max_iter=500, solver='saga', class_weight='balanced'),\n",
    "    \"Multinomial NB\": MultinomialNB(),\n",
    "    \"SVM (Linear)\": SVC(kernel='linear', class_weight='balanced')\n",
    "}\n",
    "\n",
    "# Evaluate each model with SMOTE\n",
    "for name, model in models.items():\n",
    "    pipeline = ImbPipeline([\n",
    "        (\"preprocess\", preprocessor),\n",
    "        (\"smote\", SMOTE(random_state=42)),\n",
    "        (\"model\", model)\n",
    "    ])\n",
    "    pipeline.fit(X_train, y_train)\n",
    "    y_pred = pipeline.predict(X_test)\n",
    "    acc = accuracy_score(y_test, y_pred)\n",
    "    print(f\"{name} Accuracy with SMOTE: {acc:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "9a9fa983",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting xgboost\n",
      "  Downloading xgboost-3.0.5-py3-none-macosx_12_0_arm64.whl.metadata (2.1 kB)\n",
      "Requirement already satisfied: numpy in ./venv/lib/python3.12/site-packages (from xgboost) (2.3.3)\n",
      "Requirement already satisfied: scipy in ./venv/lib/python3.12/site-packages (from xgboost) (1.16.2)\n",
      "Downloading xgboost-3.0.5-py3-none-macosx_12_0_arm64.whl (2.0 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m1.3 MB/s\u001b[0m  \u001b[33m0:00:01\u001b[0m eta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: xgboost\n",
      "Successfully installed xgboost-3.0.5\n"
     ]
    }
   ],
   "source": [
    "!pip install xgboost\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "050f6ab0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:18: SyntaxWarning: invalid escape sequence '\\|'\n",
      "<>:18: SyntaxWarning: invalid escape sequence '\\|'\n",
      "/var/folders/jf/21x3yjzs0ygd_0ttzq2sq0zm0000gn/T/ipykernel_61298/543830588.py:18: SyntaxWarning: invalid escape sequence '\\|'\n",
      "  sep=\"\\|\\|\",\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Random Forest Accuracy: 0.6205\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/venv/lib/python3.12/site-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logistic Regression Accuracy: 0.6250\n",
      "SVM Accuracy: 0.6476\n",
      "Decision Tree Accuracy: 0.5768\n",
      "✅ All models trained and saved successfully!\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.compose import ColumnTransformer\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.svm import SVC\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "# Load and merge training data\n",
    "train_var = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_variants\")\n",
    "train_text = pd.read_csv(\"/Users/anukulchandra/Python_with_machine_learning_ESPML_2405/Personalized_Cancer_Prediction-Doctor_Side-/training_text\",\n",
    "                          sep=\"\\|\\|\",\n",
    "                            engine=\"python\",\n",
    "                              names=[\"ID\",\"Text\"],\n",
    "                                skiprows=1)\n",
    "df = pd.merge(train_var, train_text, on=\"ID\")\n",
    "\n",
    "# Clean columns\n",
    "df['Text'] = df['Text'].str.lower()\n",
    "df['Gene'] = df['Gene'].str.lower().str.replace(' ', '_')\n",
    "df['Variation'] = df['Variation'].str.lower().str.replace(' ', '_')\n",
    "df.dropna(inplace=True)\n",
    "\n",
    "# Final feature set\n",
    "df = df[[\"Gene\", \"Variation\", \"Text\", \"Class\"]]\n",
    "x = df[[\"Gene\", \"Variation\", \"Text\"]]\n",
    "y = df[\"Class\"]\n",
    "\n",
    "x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Preprocessing pipeline\n",
    "preprocessor = ColumnTransformer(transformers=[\n",
    "    (\"gene_enc\", OneHotEncoder(handle_unknown=\"ignore\"), [\"Gene\"]),\n",
    "    (\"var_enc\", OneHotEncoder(handle_unknown=\"ignore\"), [\"Variation\"]),\n",
    "    (\"text_vec\", TfidfVectorizer(max_features=5000), \"Text\")\n",
    "])\n",
    "\n",
    "# Define models to evaluate\n",
    "models = {\n",
    "    \"Random Forest\": RandomForestClassifier(),\n",
    "    \"Logistic Regression\": LogisticRegression(max_iter=300, solver='saga', class_weight='balanced'),\n",
    "    \"SVM\": SVC(kernel='linear', probability=True),\n",
    "    \"Decision Tree\": DecisionTreeClassifier(),\n",
    "    \n",
    "}\n",
    "\n",
    "# Store trained models + accuracies\n",
    "trained_models = {}\n",
    "model_performance = {}\n",
    "\n",
    "for name, model in models.items():\n",
    "    pipeline = Pipeline([\n",
    "        (\"preprocess\", preprocessor),\n",
    "        (\"model\", model)\n",
    "    ])\n",
    "    pipeline.fit(x_train, y_train)\n",
    "    y_pred = pipeline.predict(x_test)\n",
    "    acc = accuracy_score(y_test, y_pred)\n",
    "    print(f\"{name} Accuracy: {acc:.4f}\")\n",
    "\n",
    "    trained_models[name] = pipeline\n",
    "    model_performance[name] = round(acc * 100, 2)\n",
    "\n",
    "# Save everything in one pickle\n",
    "with open(\"final_pipeline1.pkl\", \"wb\") as f:\n",
    "    pickle.dump({\"models\": trained_models, \"performance\": model_performance}, f)\n",
    "\n",
    "print(\"✅ All models trained and saved successfully!\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv (3.12.1)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
